// circuit-json/dist/index.mjs
// src/utils/convert-si-unit-to-number.ts
var unitMappings = {
  Hz: {
    baseUnit: "Hz",
    variants: {
      MHz: 1e6,
      kHz: 1e3,
      Hz: 1
    }
  },
  g: {
    baseUnit: "g",
    variants: {
      kg: 1e3,
      g: 1
    }
  },
  \u03A9: {
    baseUnit: "\u03A9",
    variants: {
      m\u03A9: 1e-3,
      \u03A9: 1,
      k\u03A9: 1e3,
      K\u03A9: 1e3,
      kohm: 1e3,
      M\u03A9: 1e6,
      G\u03A9: 1e9,
      T\u03A9: 1e12
    }
  },
  V: {
    baseUnit: "V",
    variants: {
      mV: 1e-3,
      V: 1,
      kV: 1e3,
      KV: 1e3,
      MV: 1e6,
      GV: 1e9,
      TV: 1e12
    }
  },
  A: {
    baseUnit: "A",
    variants: {
      \u00B5A: 1e-6,
      mA: 1e-3,
      ma: 1e-3,
      A: 1,
      kA: 1e3,
      MA: 1e6
    }
  },
  F: {
    baseUnit: "F",
    variants: {
      pF: 1e-12,
      nF: 1e-9,
      \u00B5F: 1e-6,
      uF: 1e-6,
      mF: 1e-3,
      F: 1
    }
  },
  ml: {
    baseUnit: "ml",
    variants: {
      ml: 1,
      mL: 1,
      l: 1e3,
      L: 1e3
    }
  },
  deg: {
    baseUnit: "deg",
    variants: {
      rad: 180 / Math.PI
    }
  },
  ms: {
    baseUnit: "ms",
    variants: {
      fs: 1e-12,
      ps: 1e-9,
      ns: 1e-6,
      us: 1e-3,
      \u00B5s: 1e-3,
      ms: 1,
      s: 1e3
    }
  },
  mm: {
    baseUnit: "mm",
    variants: {
      nm: 1e-6,
      \u00B5m: 1e-3,
      um: 1e-3,
      mm: 1,
      cm: 10,
      dm: 100,
      m: 1e3,
      km: 1e6,
      in: 25.4,
      ft: 304.8,
      IN: 25.4,
      FT: 304.8,
      yd: 914.4,
      mi: 1609344,
      mil: 0.0254
    }
  }
};
var unitMappingAndVariantSuffixes = /* @__PURE__ */ new Set();
for (const [baseUnit, info] of Object.entries(unitMappings)) {
  unitMappingAndVariantSuffixes.add(baseUnit);
  for (const variant of Object.keys(info.variants)) {
    unitMappingAndVariantSuffixes.add(variant);
  }
}
function getBaseTscircuitUnit(unit) {
  for (const [baseUnit, info] of Object.entries(unitMappings)) {
    if (unit in info.variants) {
      return {
        baseUnit: info.baseUnit,
        conversionFactor: info.variants[unit]
      };
    }
  }
  return {
    baseUnit: unit,
    conversionFactor: 1
  };
}
var si_prefix_multiplier = {
  tera: 1e12,
  T: 1e12,
  giga: 1e9,
  G: 1e9,
  mega: 1e6,
  M: 1e6,
  kilo: 1e3,
  k: 1e3,
  deci: 0.1,
  d: 0.1,
  centi: 0.01,
  c: 0.01,
  milli: 1e-3,
  m: 1e-3,
  micro: 1e-6,
  u: 1e-6,
  \u00B5: 1e-6,
  nano: 1e-9,
  n: 1e-9,
  pico: 1e-12,
  p: 1e-12
};
function parseAndConvertSiUnit(v) {
  if (v === void 0 || v === null)
    return { parsedUnit: null, unitOfValue: null, value: null };
  if (typeof v === "string" && v.match(/^-?[\d\.]+$/))
    return {
      value: Number.parseFloat(v),
      parsedUnit: null,
      unitOfValue: null
    };
  if (typeof v === "number")
    return { value: v, parsedUnit: null, unitOfValue: null };
  if (typeof v === "object" && "x" in v && "y" in v) {
    const { parsedUnit, unitOfValue } = parseAndConvertSiUnit(v.x);
    const xResult = parseAndConvertSiUnit(v.x);
    const yResult = parseAndConvertSiUnit(v.y);
    if (xResult.value === null || yResult.value === null) {
      return { parsedUnit: null, unitOfValue: null, value: null };
    }
    return {
      parsedUnit,
      unitOfValue,
      value: {
        x: xResult.value,
        y: yResult.value
      }
    };
  }
  const reversed_input_string = v.toString().split("").reverse().join("");
  const unit_reversed = reversed_input_string.match(/[^\d\s]+/)?.[0];
  if (!unit_reversed) {
    throw new Error(`Could not determine unit: "${v}"`);
  }
  const unit = unit_reversed.split("").reverse().join("");
  const numberPart = v.slice(0, -unit.length);
  if (unit in si_prefix_multiplier && !unitMappingAndVariantSuffixes.has(unit)) {
    const siMultiplier = si_prefix_multiplier[unit];
    return {
      parsedUnit: null,
      unitOfValue: null,
      value: Number.parseFloat(numberPart) * siMultiplier
    };
  }
  const { baseUnit, conversionFactor } = getBaseTscircuitUnit(unit);
  return {
    parsedUnit: unit,
    unitOfValue: baseUnit,
    value: conversionFactor * Number.parseFloat(numberPart)
  };
}

// src/units/index.ts
import { z } from "zod";
var resistance = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var capacitance = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value).transform((value) => {
  return Number.parseFloat(value.toPrecision(12));
});
var inductance = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var voltage = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var length = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var frequency = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var distance = length;
var current = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var duration_ms = z.string().or(z.number()).transform((v) => parseAndConvertSiUnit(v).value);
var time = duration_ms;
var ms = duration_ms;
var timestamp = z.string().datetime();
var rotation = z.string().or(z.number()).transform((arg) => {
  if (typeof arg === "number") return arg;
  if (arg.endsWith("deg")) {
    return Number.parseFloat(arg.split("deg")[0]);
  }
  if (arg.endsWith("rad")) {
    return Number.parseFloat(arg.split("rad")[0]) * 180 / Math.PI;
  }
  return Number.parseFloat(arg);
});
var battery_capacity = z.number().or(z.string().endsWith("mAh")).transform((v) => {
  if (typeof v === "string") {
    const valString = v.replace("mAh", "");
    const num = Number.parseFloat(valString);
    if (Number.isNaN(num)) {
      throw new Error("Invalid capacity");
    }
    return num;
  }
  return v;
}).describe("Battery capacity in mAh");

// src/common/point.ts
import { z as z2 } from "zod";

// src/utils/expect-types-match.ts
var expectTypesMatch = (shouldBe) => {
};
expectTypesMatch("extra props b");
expectTypesMatch("missing props b");
expectTypesMatch(true);
expectTypesMatch("mismatched prop types: a");
var expectStringUnionsMatch = (shouldBe) => {
};
expectStringUnionsMatch(true);
expectStringUnionsMatch(
  'T1 has extra: "c", T2 has extra: "d"'
);
expectStringUnionsMatch('T1 has extra: "c"');
expectStringUnionsMatch('T2 has extra: "c"');
expectStringUnionsMatch(
  'T1 has extra: "d", T2 has extra: "c"'
);
expectStringUnionsMatch(true);

// src/common/point.ts
var point = z2.object({
  x: distance,
  y: distance
});
var position = point;
expectTypesMatch(true);
expectTypesMatch(true);

// src/common/point3.ts
import { z as z3 } from "zod";
var point3 = z3.object({
  x: distance,
  y: distance,
  z: distance
});
var position3 = point3;
expectTypesMatch(true);

// src/common/size.ts
import { z as z4 } from "zod";
var size = z4.object({
  width: z4.number(),
  height: z4.number()
});
expectTypesMatch(true);

// src/common/getZodPrefixedIdWithDefault.ts
import { z as z5 } from "zod";
var randomId = (length4) => {
  const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  return Array.from(
    { length: length4 },
    () => chars[Math.floor(Math.random() * chars.length)]
  ).join("");
};
var getZodPrefixedIdWithDefault = (prefix) => {
  return z5.string().optional().default(() => `${prefix}_${randomId(10)}`);
};

// src/common/NinePointAnchor.ts
import { z as z6 } from "zod";
var ninePointAnchor = z6.enum([
  "top_left",
  "top_center",
  "top_right",
  "center_left",
  "center",
  "center_right",
  "bottom_left",
  "bottom_center",
  "bottom_right"
]);
expectTypesMatch(true);

// src/common/PcbRenderLayer.ts
import { z as z7 } from "zod";
var pcbRenderLayer = z7.enum([
  "top_silkscreen",
  "bottom_silkscreen",
  "top_copper",
  "bottom_copper",
  "top_soldermask",
  "bottom_soldermask",
  "top_fabrication_note",
  "bottom_fabrication_note",
  "top_user_note",
  "bottom_user_note",
  "top_courtyard",
  "bottom_courtyard",
  "inner1_copper",
  "inner2_copper",
  "inner3_copper",
  "inner4_copper",
  "inner5_copper",
  "inner6_copper",
  "edge_cuts",
  "drill"
]);
expectTypesMatch(true);

// src/common/asset.ts
import { z as z8 } from "zod";
var asset = z8.object({
  project_relative_path: z8.string(),
  url: z8.string(),
  mimetype: z8.string()
});
expectTypesMatch(true);

// src/common/kicadFootprintMetadata.ts
import { z as z9 } from "zod";
var kicadAt = point.extend({
  rotation: rotation.optional()
});
expectTypesMatch(true);
var kicadFont = z9.object({
  size: point.optional(),
  thickness: distance.optional()
});
expectTypesMatch(true);
var kicadEffects = z9.object({
  font: kicadFont.optional()
});
expectTypesMatch(true);
var kicadProperty = z9.object({
  value: z9.string(),
  at: kicadAt.optional(),
  layer: z9.string().optional(),
  uuid: z9.string().optional(),
  hide: z9.boolean().optional(),
  effects: kicadEffects.optional()
});
expectTypesMatch(true);
var kicadFootprintProperties = z9.object({
  Reference: kicadProperty.optional(),
  Value: kicadProperty.optional(),
  Datasheet: kicadProperty.optional(),
  Description: kicadProperty.optional()
});
expectTypesMatch(
  true
);
var kicadFootprintAttributes = z9.object({
  through_hole: z9.boolean().optional(),
  smd: z9.boolean().optional(),
  exclude_from_pos_files: z9.boolean().optional(),
  exclude_from_bom: z9.boolean().optional()
});
expectTypesMatch(
  true
);
var kicadFootprintPad = z9.object({
  name: z9.string(),
  type: z9.string(),
  shape: z9.string().optional(),
  at: kicadAt.optional(),
  size: point.optional(),
  drill: distance.optional(),
  layers: z9.array(z9.string()).optional(),
  removeUnusedLayers: z9.boolean().optional(),
  uuid: z9.string().optional()
});
expectTypesMatch(true);
var kicadFootprintModel = z9.object({
  path: z9.string(),
  offset: point3.optional(),
  scale: point3.optional(),
  rotate: point3.optional()
});
expectTypesMatch(true);
var kicadFootprintMetadata = z9.object({
  footprintName: z9.string().optional(),
  version: z9.union([z9.number(), z9.string()]).optional(),
  generator: z9.string().optional(),
  generatorVersion: z9.union([z9.number(), z9.string()]).optional(),
  layer: z9.string().optional(),
  properties: kicadFootprintProperties.optional(),
  attributes: kicadFootprintAttributes.optional(),
  pads: z9.array(kicadFootprintPad).optional(),
  embeddedFonts: z9.boolean().optional(),
  model: kicadFootprintModel.optional()
});
expectTypesMatch(true);

// src/common/kicadSymbolMetadata.ts
import { z as z10 } from "zod";
var kicadSymbolPinNumbers = z10.object({
  hide: z10.boolean().optional()
});
expectTypesMatch(true);
var kicadSymbolPinNames = z10.object({
  offset: distance.optional(),
  hide: z10.boolean().optional()
});
expectTypesMatch(true);
var kicadSymbolEffects = z10.object({
  font: kicadFont.optional(),
  justify: z10.union([z10.string(), z10.array(z10.string())]).optional(),
  hide: z10.boolean().optional()
});
expectTypesMatch(true);
var kicadSymbolProperty = z10.object({
  value: z10.string(),
  id: z10.union([z10.number(), z10.string()]).optional(),
  at: kicadAt.optional(),
  effects: kicadSymbolEffects.optional()
});
expectTypesMatch(true);
var kicadSymbolProperties = z10.object({
  Reference: kicadSymbolProperty.optional(),
  Value: kicadSymbolProperty.optional(),
  Footprint: kicadSymbolProperty.optional(),
  Datasheet: kicadSymbolProperty.optional(),
  Description: kicadSymbolProperty.optional(),
  ki_keywords: kicadSymbolProperty.optional(),
  ki_fp_filters: kicadSymbolProperty.optional()
});
expectTypesMatch(true);
var kicadSymbolMetadata = z10.object({
  symbolName: z10.string().optional(),
  extends: z10.string().optional(),
  pinNumbers: kicadSymbolPinNumbers.optional(),
  pinNames: kicadSymbolPinNames.optional(),
  excludeFromSim: z10.boolean().optional(),
  inBom: z10.boolean().optional(),
  onBoard: z10.boolean().optional(),
  properties: kicadSymbolProperties.optional(),
  embeddedFonts: z10.boolean().optional()
});
expectTypesMatch(true);

// src/base_circuit_json_error.ts
import { z as z11 } from "zod";
var base_circuit_json_error = z11.object({
  error_type: z11.string(),
  message: z11.string(),
  is_fatal: z11.boolean().optional()
});
expectTypesMatch(true);

// src/source/source_simple_capacitor.ts
import { z as z14 } from "zod";

// src/pcb/properties/supplier_name.ts
import { z as z12 } from "zod";
var supplier_name = z12.enum([
  "jlcpcb",
  "macrofab",
  "pcbway",
  "digikey",
  "mouser",
  "lcsc"
]);
expectTypesMatch(true);

// src/source/base/source_component_base.ts
import { z as z13 } from "zod";
var source_component_base = z13.object({
  type: z13.literal("source_component"),
  ftype: z13.string().optional(),
  source_component_id: z13.string(),
  name: z13.string(),
  manufacturer_part_number: z13.string().optional(),
  supplier_part_numbers: z13.record(supplier_name, z13.array(z13.string())).optional(),
  display_value: z13.string().optional(),
  display_name: z13.string().optional(),
  are_pins_interchangeable: z13.boolean().optional(),
  internally_connected_source_port_ids: z13.array(z13.array(z13.string())).optional(),
  source_group_id: z13.string().optional(),
  subcircuit_id: z13.string().optional()
});
expectTypesMatch(true);

// src/source/source_simple_capacitor.ts
var source_simple_capacitor = source_component_base.extend({
  ftype: z14.literal("simple_capacitor"),
  capacitance,
  max_voltage_rating: voltage.optional(),
  display_capacitance: z14.string().optional(),
  max_decoupling_trace_length: distance.optional()
});
expectTypesMatch(true);

// src/source/source_simple_resistor.ts
import { z as z15 } from "zod";
var source_simple_resistor = source_component_base.extend({
  ftype: z15.literal("simple_resistor"),
  resistance,
  display_resistance: z15.string().optional()
});
expectTypesMatch(true);

// src/source/source_simple_diode.ts
import { z as z16 } from "zod";
var source_simple_diode = source_component_base.extend({
  ftype: z16.literal("simple_diode")
});
expectTypesMatch(true);

// src/source/source_simple_fiducial.ts
import { z as z17 } from "zod";
var source_simple_fiducial = source_component_base.extend({
  ftype: z17.literal("simple_fiducial")
});
expectTypesMatch(true);

// src/source/source_simple_led.ts
import { z as z18 } from "zod";
var source_simple_led = source_simple_diode.extend({
  ftype: z18.literal("simple_led"),
  color: z18.string().optional(),
  wavelength: z18.string().optional()
});
expectTypesMatch(true);

// src/source/source_simple_ground.ts
import { z as z19 } from "zod";
var source_simple_ground = source_component_base.extend({
  ftype: z19.literal("simple_ground")
});
expectTypesMatch(true);

// src/source/source_simple_chip.ts
import { z as z20 } from "zod";
var source_simple_chip = source_component_base.extend({
  ftype: z20.literal("simple_chip")
});
expectTypesMatch(true);

// src/source/source_simple_power_source.ts
import { z as z21 } from "zod";
var source_simple_power_source = source_component_base.extend({
  ftype: z21.literal("simple_power_source"),
  voltage
});
expectTypesMatch(true);

// src/source/source_simple_current_source.ts
import { z as z22 } from "zod";
var source_simple_current_source = source_component_base.extend({
  ftype: z22.literal("simple_current_source"),
  current,
  frequency: frequency.optional(),
  peak_to_peak_current: current.optional(),
  wave_shape: z22.enum(["sine", "square", "triangle", "sawtooth", "dc"]).optional().default("dc"),
  phase: z22.number().optional(),
  duty_cycle: z22.number().min(0).max(1).optional()
});
expectTypesMatch(
  true
);

// src/source/properties/source_pin_attributes.ts
import { z as z23 } from "zod";
var source_pin_attributes = z23.object({
  must_be_connected: z23.boolean().optional(),
  provides_power: z23.boolean().optional(),
  requires_power: z23.boolean().optional(),
  provides_ground: z23.boolean().optional(),
  requires_ground: z23.boolean().optional(),
  provides_voltage: z23.union([z23.string(), z23.number()]).optional(),
  requires_voltage: z23.union([z23.string(), z23.number()]).optional(),
  do_not_connect: z23.boolean().optional(),
  include_in_board_pinout: z23.boolean().optional(),
  can_use_internal_pullup: z23.boolean().optional(),
  is_using_internal_pullup: z23.boolean().optional(),
  needs_external_pullup: z23.boolean().optional(),
  can_use_internal_pulldown: z23.boolean().optional(),
  is_using_internal_pulldown: z23.boolean().optional(),
  needs_external_pulldown: z23.boolean().optional(),
  can_use_open_drain: z23.boolean().optional(),
  is_using_open_drain: z23.boolean().optional(),
  can_use_push_pull: z23.boolean().optional(),
  is_using_push_pull: z23.boolean().optional(),
  should_have_decoupling_capacitor: z23.boolean().optional(),
  recommended_decoupling_capacitor_capacitance: z23.union([z23.string(), z23.number()]).optional(),
  is_configured_for_i2c_sda: z23.boolean().optional(),
  is_configured_for_i2c_scl: z23.boolean().optional(),
  is_configured_for_spi_mosi: z23.boolean().optional(),
  is_configured_for_spi_miso: z23.boolean().optional(),
  is_configured_for_spi_sck: z23.boolean().optional(),
  is_configured_for_spi_cs: z23.boolean().optional(),
  is_configured_for_uart_tx: z23.boolean().optional(),
  is_configured_for_uart_rx: z23.boolean().optional(),
  supports_i2c_sda: z23.boolean().optional(),
  supports_i2c_scl: z23.boolean().optional(),
  supports_spi_mosi: z23.boolean().optional(),
  supports_spi_miso: z23.boolean().optional(),
  supports_spi_sck: z23.boolean().optional(),
  supports_spi_cs: z23.boolean().optional(),
  supports_uart_tx: z23.boolean().optional(),
  supports_uart_rx: z23.boolean().optional()
});
expectTypesMatch(true);

// src/source/any_source_component.ts
import { z as z50 } from "zod";

// src/source/source_simple_fuse.ts
import { z as z24 } from "zod";
var source_simple_fuse = source_component_base.extend({
  ftype: z24.literal("simple_fuse"),
  current_rating_amps: z24.number().describe("Nominal current in amps the fuse is rated for"),
  voltage_rating_volts: z24.number().describe("Voltage rating in volts, e.g. \xB15V would be 5")
});
expectTypesMatch(true);

// src/source/source_simple_battery.ts
import { z as z25 } from "zod";
var source_simple_battery = source_component_base.extend({
  ftype: z25.literal("simple_battery"),
  capacity: battery_capacity
});
expectTypesMatch(true);

// src/source/source_simple_inductor.ts
import { z as z26 } from "zod";
var source_simple_inductor = source_component_base.extend({
  ftype: z26.literal("simple_inductor"),
  inductance,
  max_current_rating: z26.number().optional()
});
expectTypesMatch(true);

// src/source/source_simple_push_button.ts
import { z as z27 } from "zod";
var source_simple_push_button = source_component_base.extend({
  ftype: z27.literal("simple_push_button")
});
expectTypesMatch(true);

// src/source/source_simple_potentiometer.ts
import { z as z28 } from "zod";
var source_simple_potentiometer = source_component_base.extend({
  ftype: z28.literal("simple_potentiometer"),
  max_resistance: resistance,
  display_max_resistance: z28.string().optional()
});
expectTypesMatch(
  true
);

// src/source/source_simple_crystal.ts
import { z as z29 } from "zod";
var source_simple_crystal = source_component_base.extend({
  ftype: z29.literal("simple_crystal"),
  frequency: z29.number().describe("Frequency in Hz"),
  load_capacitance: z29.number().optional().describe("Load capacitance in pF"),
  pin_variant: z29.enum(["two_pin", "four_pin"]).optional()
});
expectTypesMatch(true);

// src/source/source_simple_pin_header.ts
import { z as z30 } from "zod";
var source_simple_pin_header = source_component_base.extend({
  ftype: z30.literal("simple_pin_header"),
  pin_count: z30.number(),
  gender: z30.enum(["male", "female"]).optional().default("male")
});
expectTypesMatch(true);

// src/source/source_simple_connector.ts
import { z as z31 } from "zod";
var source_simple_connector = source_component_base.extend({
  ftype: z31.literal("simple_connector"),
  standard: z31.enum(["usb_c", "m2"]).optional()
});
expectTypesMatch(true);

// src/source/source_simple_pinout.ts
import { z as z32 } from "zod";
var source_simple_pinout = source_component_base.extend({
  ftype: z32.literal("simple_pinout")
});
expectTypesMatch(true);

// src/source/source_simple_resonator.ts
import { z as z33 } from "zod";
var source_simple_resonator = source_component_base.extend({
  ftype: z33.literal("simple_resonator"),
  load_capacitance: capacitance,
  equivalent_series_resistance: resistance.optional(),
  frequency
});
expectTypesMatch(true);

// src/source/source_simple_transistor.ts
import { z as z34 } from "zod";
var source_simple_transistor = source_component_base.extend({
  ftype: z34.literal("simple_transistor"),
  transistor_type: z34.enum(["npn", "pnp"])
});
expectTypesMatch(true);

// src/source/source_simple_test_point.ts
import { z as z35 } from "zod";
var source_simple_test_point = source_component_base.extend({
  ftype: z35.literal("simple_test_point"),
  footprint_variant: z35.enum(["pad", "through_hole"]).optional(),
  pad_shape: z35.enum(["rect", "circle"]).optional(),
  pad_diameter: z35.union([z35.number(), z35.string()]).optional(),
  hole_diameter: z35.union([z35.number(), z35.string()]).optional(),
  width: z35.union([z35.number(), z35.string()]).optional(),
  height: z35.union([z35.number(), z35.string()]).optional()
});
expectTypesMatch(true);

// src/source/source_simple_mosfet.ts
import { z as z36 } from "zod";
var source_simple_mosfet = source_component_base.extend({
  ftype: z36.literal("simple_mosfet"),
  channel_type: z36.enum(["n", "p"]),
  mosfet_mode: z36.enum(["enhancement", "depletion"])
});
expectTypesMatch(true);

// src/source/source_simple_op_amp.ts
import { z as z37 } from "zod";
var source_simple_op_amp = source_component_base.extend({
  ftype: z37.literal("simple_op_amp")
});
expectTypesMatch(true);

// src/source/source_simple_switch.ts
import { z as z38 } from "zod";
var source_simple_switch = source_component_base.extend({
  ftype: z38.literal("simple_switch")
});
expectTypesMatch(true);

// src/source/source_project_metadata.ts
import { z as z39 } from "zod";
var source_project_metadata = z39.object({
  type: z39.literal("source_project_metadata"),
  name: z39.string().optional(),
  software_used_string: z39.string().optional(),
  project_url: z39.string().optional(),
  created_at: timestamp.optional()
});
expectTypesMatch(true);

// src/source/source_missing_property_error.ts
import { z as z40 } from "zod";
var source_missing_property_error = base_circuit_json_error.extend({
  type: z40.literal("source_missing_property_error"),
  source_missing_property_error_id: getZodPrefixedIdWithDefault(
    "source_missing_property_error"
  ),
  source_component_id: z40.string(),
  property_name: z40.string(),
  subcircuit_id: z40.string().optional(),
  error_type: z40.literal("source_missing_property_error").default("source_missing_property_error")
}).describe("The source code is missing a property");
expectTypesMatch(true);

// src/source/source_failed_to_create_component_error.ts
import { z as z41 } from "zod";
var source_failed_to_create_component_error = base_circuit_json_error.extend({
  type: z41.literal("source_failed_to_create_component_error"),
  source_failed_to_create_component_error_id: getZodPrefixedIdWithDefault(
    "source_failed_to_create_component_error"
  ),
  error_type: z41.literal("source_failed_to_create_component_error").default("source_failed_to_create_component_error"),
  component_name: z41.string().optional(),
  subcircuit_id: z41.string().optional(),
  parent_source_component_id: z41.string().optional(),
  pcb_center: z41.object({
    x: z41.number().optional(),
    y: z41.number().optional()
  }).optional(),
  schematic_center: z41.object({
    x: z41.number().optional(),
    y: z41.number().optional()
  }).optional()
}).describe("Error emitted when a component fails to be constructed");
expectTypesMatch(true);

// src/source/source_invalid_component_property_error.ts
import { z as z42 } from "zod";
var source_invalid_component_property_error = base_circuit_json_error.extend({
  type: z42.literal("source_invalid_component_property_error"),
  source_invalid_component_property_error_id: getZodPrefixedIdWithDefault(
    "source_invalid_component_property_error"
  ),
  source_component_id: z42.string(),
  property_name: z42.string(),
  property_value: z42.unknown().optional(),
  expected_format: z42.string().optional(),
  subcircuit_id: z42.string().optional(),
  error_type: z42.literal("source_invalid_component_property_error").default("source_invalid_component_property_error")
}).describe("The source component property is invalid");
expectTypesMatch(true);

// src/source/source_trace_not_connected_error.ts
import { z as z43 } from "zod";
var source_trace_not_connected_error = base_circuit_json_error.extend({
  type: z43.literal("source_trace_not_connected_error"),
  source_trace_not_connected_error_id: getZodPrefixedIdWithDefault(
    "source_trace_not_connected_error"
  ),
  error_type: z43.literal("source_trace_not_connected_error").default("source_trace_not_connected_error"),
  subcircuit_id: z43.string().optional(),
  source_group_id: z43.string().optional(),
  source_trace_id: z43.string().optional(),
  connected_source_port_ids: z43.array(z43.string()).optional(),
  selectors_not_found: z43.array(z43.string()).optional()
}).describe("Occurs when a source trace selector does not match any ports");
expectTypesMatch(true);

// src/source/source_property_ignored_warning.ts
import { z as z44 } from "zod";
var source_property_ignored_warning = z44.object({
  type: z44.literal("source_property_ignored_warning"),
  source_property_ignored_warning_id: getZodPrefixedIdWithDefault(
    "source_property_ignored_warning"
  ),
  source_component_id: z44.string(),
  property_name: z44.string(),
  subcircuit_id: z44.string().optional(),
  error_type: z44.literal("source_property_ignored_warning").default("source_property_ignored_warning"),
  message: z44.string()
}).describe("The source property was ignored");
expectTypesMatch(true);

// src/source/source_pin_missing_trace_warning.ts
import { z as z45 } from "zod";
var source_pin_missing_trace_warning = z45.object({
  type: z45.literal("source_pin_missing_trace_warning"),
  source_pin_missing_trace_warning_id: getZodPrefixedIdWithDefault(
    "source_pin_missing_trace_warning"
  ),
  warning_type: z45.literal("source_pin_missing_trace_warning").default("source_pin_missing_trace_warning"),
  message: z45.string(),
  source_component_id: z45.string(),
  source_port_id: z45.string(),
  subcircuit_id: z45.string().optional()
}).describe(
  "Warning emitted when a source component pin is missing a trace connection"
);
expectTypesMatch(true);

// src/source/source_simple_voltage_probe.ts
import { z as z46 } from "zod";
var source_simple_voltage_probe = source_component_base.extend({
  ftype: z46.literal("simple_voltage_probe")
});
expectTypesMatch(
  true
);

// src/source/source_interconnect.ts
import { z as z47 } from "zod";
var source_interconnect = source_component_base.extend({
  ftype: z47.literal("interconnect")
});
expectTypesMatch(true);

// src/source/source_i2c_misconfigured_error.ts
import { z as z48 } from "zod";
var source_i2c_misconfigured_error = base_circuit_json_error.extend({
  type: z48.literal("source_i2c_misconfigured_error"),
  source_i2c_misconfigured_error_id: getZodPrefixedIdWithDefault(
    "source_i2c_misconfigured_error"
  ),
  error_type: z48.literal("source_i2c_misconfigured_error").default("source_i2c_misconfigured_error"),
  source_port_ids: z48.array(z48.string())
}).describe(
  "Error emitted when incompatible I2C pins (e.g. SDA and SCL) are connected to the same net"
);
expectTypesMatch(true);

// src/source/source_simple_voltage_source.ts
import { z as z49 } from "zod";
var source_simple_voltage_source = source_component_base.extend({
  ftype: z49.literal("simple_voltage_source"),
  voltage,
  frequency: frequency.optional(),
  peak_to_peak_voltage: voltage.optional(),
  wave_shape: z49.enum(["sinewave", "square", "triangle", "sawtooth"]).optional(),
  phase: rotation.optional(),
  duty_cycle: z49.number().optional().describe("Duty cycle as a fraction (0 to 1)")
});
expectTypesMatch(
  true
);

// src/source/any_source_component.ts
var any_source_component = z50.union([
  source_simple_resistor,
  source_simple_capacitor,
  source_simple_diode,
  source_simple_fiducial,
  source_simple_led,
  source_simple_ground,
  source_simple_chip,
  source_simple_power_source,
  source_simple_current_source,
  source_simple_battery,
  source_simple_inductor,
  source_simple_push_button,
  source_simple_potentiometer,
  source_simple_crystal,
  source_simple_pin_header,
  source_simple_connector,
  source_simple_pinout,
  source_simple_resonator,
  source_simple_switch,
  source_simple_transistor,
  source_simple_test_point,
  source_simple_mosfet,
  source_simple_op_amp,
  source_simple_fuse,
  source_simple_voltage_probe,
  source_interconnect,
  source_simple_voltage_source,
  source_project_metadata,
  source_missing_property_error,
  source_invalid_component_property_error,
  source_failed_to_create_component_error,
  source_trace_not_connected_error,
  source_property_ignored_warning,
  source_pin_missing_trace_warning,
  source_i2c_misconfigured_error
]);
expectTypesMatch(true);

// src/source/source_port.ts
import { z as z51 } from "zod";
var source_port = z51.object({
  type: z51.literal("source_port"),
  pin_number: z51.number().optional(),
  port_hints: z51.array(z51.string()).optional(),
  name: z51.string(),
  source_port_id: z51.string(),
  source_component_id: z51.string().optional(),
  source_group_id: z51.string().optional(),
  most_frequently_referenced_by_name: z51.string().optional(),
  subcircuit_id: z51.string().optional(),
  subcircuit_connectivity_map_key: z51.string().optional()
}).merge(source_pin_attributes);
expectTypesMatch(true);

// src/source/source_component_internal_connection.ts
import { z as z52 } from "zod";
var source_component_internal_connection = z52.object({
  type: z52.literal("source_component_internal_connection"),
  source_component_internal_connection_id: z52.string(),
  source_component_id: z52.string(),
  source_port_ids: z52.array(z52.string()),
  subcircuit_id: z52.string().optional()
});
expectTypesMatch(true);

// src/source/source_trace.ts
import { z as z53 } from "zod";
var source_trace = z53.object({
  type: z53.literal("source_trace"),
  source_trace_id: z53.string(),
  connected_source_port_ids: z53.array(z53.string()),
  connected_source_net_ids: z53.array(z53.string()),
  subcircuit_id: z53.string().optional(),
  subcircuit_connectivity_map_key: z53.string().optional(),
  max_length: z53.number().optional(),
  min_trace_thickness: z53.number().optional(),
  display_name: z53.string().optional()
});
expectTypesMatch(true);

// src/source/source_group.ts
import { z as z54 } from "zod";
var source_group = z54.object({
  type: z54.literal("source_group"),
  source_group_id: z54.string(),
  subcircuit_id: z54.string().optional(),
  parent_subcircuit_id: z54.string().optional(),
  parent_source_group_id: z54.string().optional(),
  is_subcircuit: z54.boolean().optional(),
  show_as_schematic_box: z54.boolean().optional(),
  name: z54.string().optional(),
  was_automatically_named: z54.boolean().optional()
});
expectTypesMatch(true);

// src/source/source_net.ts
import { z as z55 } from "zod";
var source_net = z55.object({
  type: z55.literal("source_net"),
  source_net_id: z55.string(),
  name: z55.string(),
  member_source_group_ids: z55.array(z55.string()),
  is_power: z55.boolean().optional(),
  is_ground: z55.boolean().optional(),
  is_digital_signal: z55.boolean().optional(),
  is_analog_signal: z55.boolean().optional(),
  is_positive_voltage_source: z55.boolean().optional(),
  trace_width: z55.number().optional(),
  subcircuit_id: z55.string().optional(),
  subcircuit_connectivity_map_key: z55.string().optional()
});
expectTypesMatch(true);

// src/source/source_board.ts
import { z as z56 } from "zod";
var source_board = z56.object({
  type: z56.literal("source_board"),
  source_board_id: z56.string(),
  source_group_id: z56.string(),
  title: z56.string().optional()
}).describe("Defines a board in the source domain");
expectTypesMatch(true);

// src/source/source_ambiguous_port_reference.ts
import { z as z57 } from "zod";
var source_ambiguous_port_reference = base_circuit_json_error.extend({
  type: z57.literal("source_ambiguous_port_reference"),
  source_ambiguous_port_reference_id: getZodPrefixedIdWithDefault(
    "source_ambiguous_port_reference"
  ),
  error_type: z57.literal("source_ambiguous_port_reference").default("source_ambiguous_port_reference"),
  source_port_id: z57.string().optional(),
  source_component_id: z57.string().optional()
}).describe(
  "Error emitted when a port hint matches multiple non-overlapping pads, making the port reference ambiguous"
);
expectTypesMatch(true);

// src/source/source_pcb_ground_plane.ts
import { z as z58 } from "zod";
var source_pcb_ground_plane = z58.object({
  type: z58.literal("source_pcb_ground_plane"),
  source_pcb_ground_plane_id: z58.string(),
  source_group_id: z58.string(),
  source_net_id: z58.string(),
  subcircuit_id: z58.string().optional()
}).describe("Defines a ground plane in the source domain");
expectTypesMatch(true);

// src/source/source_manually_placed_via.ts
import { z as z60 } from "zod";

// src/pcb/properties/layer_ref.ts
import { z as z59 } from "zod";
var all_layers = [
  "top",
  "bottom",
  "inner1",
  "inner2",
  "inner3",
  "inner4",
  "inner5",
  "inner6"
];
var layer_string = z59.enum(all_layers);
var layer_ref = layer_string.or(
  z59.object({
    name: layer_string
  })
).transform((layer) => {
  if (typeof layer === "string") {
    return layer;
  }
  return layer.name;
});
expectTypesMatch(true);
var visible_layer = z59.enum(["top", "bottom"]);

// src/source/source_manually_placed_via.ts
var source_manually_placed_via = z60.object({
  type: z60.literal("source_manually_placed_via"),
  source_manually_placed_via_id: z60.string(),
  source_group_id: z60.string(),
  source_net_id: z60.string(),
  subcircuit_id: z60.string().optional(),
  source_trace_id: z60.string().optional()
}).describe("Defines a via that is manually placed in the source domain");
expectTypesMatch(true);

// src/source/source_no_power_pin_defined_warning.ts
import { z as z61 } from "zod";
var source_no_power_pin_defined_warning = z61.object({
  type: z61.literal("source_no_power_pin_defined_warning"),
  source_no_power_pin_defined_warning_id: getZodPrefixedIdWithDefault(
    "source_no_power_pin_defined_warning"
  ),
  warning_type: z61.literal("source_no_power_pin_defined_warning").default("source_no_power_pin_defined_warning"),
  message: z61.string(),
  source_component_id: z61.string(),
  source_port_ids: z61.array(z61.string()),
  subcircuit_id: z61.string().optional()
}).describe(
  "Warning emitted when a chip has no source ports with requires_power=true"
);
expectTypesMatch(true);

// src/source/source_no_ground_pin_defined_warning.ts
import { z as z62 } from "zod";
var source_no_ground_pin_defined_warning = z62.object({
  type: z62.literal("source_no_ground_pin_defined_warning"),
  source_no_ground_pin_defined_warning_id: getZodPrefixedIdWithDefault(
    "source_no_ground_pin_defined_warning"
  ),
  warning_type: z62.literal("source_no_ground_pin_defined_warning").default("source_no_ground_pin_defined_warning"),
  message: z62.string(),
  source_component_id: z62.string(),
  source_port_ids: z62.array(z62.string()),
  subcircuit_id: z62.string().optional()
}).describe(
  "Warning emitted when a chip has no source ports marked as ground pins"
);
expectTypesMatch(true);

// src/source/source_component_pins_underspecified_warning.ts
import { z as z63 } from "zod";
var source_component_pins_underspecified_warning = z63.object({
  type: z63.literal("source_component_pins_underspecified_warning"),
  source_component_pins_underspecified_warning_id: getZodPrefixedIdWithDefault(
    "source_component_pins_underspecified_warning"
  ),
  warning_type: z63.literal("source_component_pins_underspecified_warning").default("source_component_pins_underspecified_warning"),
  message: z63.string(),
  source_component_id: z63.string(),
  source_port_ids: z63.array(z63.string()),
  subcircuit_id: z63.string().optional()
}).describe(
  "Warning emitted when all ports on a source component are underspecified"
);
expectTypesMatch(true);

// src/source/source_pin_must_be_connected_error.ts
import { z as z64 } from "zod";
var source_pin_must_be_connected_error = base_circuit_json_error.extend({
  type: z64.literal("source_pin_must_be_connected_error"),
  source_pin_must_be_connected_error_id: getZodPrefixedIdWithDefault(
    "source_pin_must_be_connected_error"
  ),
  error_type: z64.literal("source_pin_must_be_connected_error").default("source_pin_must_be_connected_error"),
  source_component_id: z64.string(),
  source_port_id: z64.string(),
  subcircuit_id: z64.string().optional()
}).describe(
  "Error emitted when a pin with mustBeConnected attribute is not connected to any trace"
);
expectTypesMatch(true);

// src/source/unknown_error_finding_part.ts
import { z as z65 } from "zod";
var unknown_error_finding_part = base_circuit_json_error.extend({
  type: z65.literal("unknown_error_finding_part"),
  unknown_error_finding_part_id: getZodPrefixedIdWithDefault(
    "unknown_error_finding_part"
  ),
  error_type: z65.literal("unknown_error_finding_part").default("unknown_error_finding_part"),
  source_component_id: z65.string().optional(),
  subcircuit_id: z65.string().optional()
}).describe(
  "Error emitted when an unexpected error occurs while finding a part"
);
expectTypesMatch(true);

// src/schematic/schematic_box.ts
import { z as z66 } from "zod";
var schematic_box = z66.object({
  type: z66.literal("schematic_box"),
  schematic_component_id: z66.string().optional(),
  schematic_symbol_id: z66.string().optional(),
  width: distance,
  height: distance,
  is_dashed: z66.boolean().default(false),
  x: distance,
  y: distance,
  subcircuit_id: z66.string().optional()
}).describe("Draws a box on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_path.ts
import { z as z67 } from "zod";
var schematic_path = z67.object({
  type: z67.literal("schematic_path"),
  schematic_path_id: getZodPrefixedIdWithDefault("schematic_path"),
  schematic_component_id: z67.string().optional(),
  schematic_symbol_id: z67.string().optional(),
  fill_color: z67.string().optional(),
  is_filled: z67.boolean().optional(),
  stroke_width: distance.nullable().optional(),
  stroke_color: z67.string().optional(),
  points: z67.array(point),
  subcircuit_id: z67.string().optional()
});
expectTypesMatch(true);

// src/schematic/schematic_component.ts
import { z as z68 } from "zod";
var schematic_pin_styles = z68.record(
  z68.object({
    left_margin: length.optional(),
    right_margin: length.optional(),
    top_margin: length.optional(),
    bottom_margin: length.optional()
  })
);
var schematic_component_port_arrangement_by_size = z68.object({
  left_size: z68.number(),
  right_size: z68.number(),
  top_size: z68.number().optional(),
  bottom_size: z68.number().optional()
});
expectTypesMatch(true);
var schematic_component_port_arrangement_by_sides = z68.object({
  left_side: z68.object({
    pins: z68.array(z68.number()),
    // @ts-ignore
    direction: z68.enum(["top-to-bottom", "bottom-to-top"]).optional()
  }).optional(),
  right_side: z68.object({
    pins: z68.array(z68.number()),
    // @ts-ignore
    direction: z68.enum(["top-to-bottom", "bottom-to-top"]).optional()
  }).optional(),
  top_side: z68.object({
    pins: z68.array(z68.number()),
    // @ts-ignore
    direction: z68.enum(["left-to-right", "right-to-left"]).optional()
  }).optional(),
  bottom_side: z68.object({
    pins: z68.array(z68.number()),
    // @ts-ignore
    direction: z68.enum(["left-to-right", "right-to-left"]).optional()
  }).optional()
});
expectTypesMatch(true);
var port_arrangement = z68.union([
  schematic_component_port_arrangement_by_size,
  schematic_component_port_arrangement_by_sides
]);
var schematic_component = z68.object({
  type: z68.literal("schematic_component"),
  size,
  center: point,
  source_component_id: z68.string().optional(),
  schematic_component_id: z68.string(),
  schematic_symbol_id: z68.string().optional(),
  pin_spacing: length.optional(),
  pin_styles: schematic_pin_styles.optional(),
  box_width: length.optional(),
  symbol_name: z68.string().optional(),
  port_arrangement: port_arrangement.optional(),
  port_labels: z68.record(z68.string()).optional(),
  symbol_display_value: z68.string().optional(),
  subcircuit_id: z68.string().optional(),
  schematic_group_id: z68.string().optional(),
  is_schematic_group: z68.boolean().optional(),
  source_group_id: z68.string().optional(),
  is_box_with_pins: z68.boolean().optional().default(true)
});
expectTypesMatch(true);

// src/schematic/schematic_symbol.ts
import { z as z69 } from "zod";
var schematicSymbolMetadata = z69.object({
  kicad_symbol: kicadSymbolMetadata.optional()
}).catchall(z69.unknown());
var schematic_symbol = z69.object({
  type: z69.literal("schematic_symbol"),
  schematic_symbol_id: z69.string(),
  name: z69.string().optional(),
  metadata: schematicSymbolMetadata.optional()
}).describe(
  "Defines a named schematic symbol that can be referenced by components."
);
expectTypesMatch(true);

// src/schematic/schematic_line.ts
import { z as z70 } from "zod";
var schematic_line = z70.object({
  type: z70.literal("schematic_line"),
  schematic_line_id: getZodPrefixedIdWithDefault("schematic_line"),
  schematic_component_id: z70.string().optional(),
  schematic_symbol_id: z70.string().optional(),
  x1: distance,
  y1: distance,
  x2: distance,
  y2: distance,
  stroke_width: distance.nullable().optional(),
  color: z70.string().default("#000000"),
  is_dashed: z70.boolean().default(false),
  subcircuit_id: z70.string().optional()
}).describe("Draws a styled line on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_rect.ts
import { z as z71 } from "zod";
var schematic_rect = z71.object({
  type: z71.literal("schematic_rect"),
  schematic_rect_id: getZodPrefixedIdWithDefault("schematic_rect"),
  schematic_component_id: z71.string().optional(),
  schematic_symbol_id: z71.string().optional(),
  center: point,
  width: distance,
  height: distance,
  rotation: rotation.default(0),
  stroke_width: distance.nullable().optional(),
  color: z71.string().default("#000000"),
  is_filled: z71.boolean().default(false),
  fill_color: z71.string().optional(),
  is_dashed: z71.boolean().default(false),
  subcircuit_id: z71.string().optional()
}).describe("Draws a styled rectangle on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_circle.ts
import { z as z72 } from "zod";
var schematic_circle = z72.object({
  type: z72.literal("schematic_circle"),
  schematic_circle_id: getZodPrefixedIdWithDefault("schematic_circle"),
  schematic_component_id: z72.string().optional(),
  schematic_symbol_id: z72.string().optional(),
  center: point,
  radius: distance,
  stroke_width: distance.nullable().optional(),
  color: z72.string().default("#000000"),
  is_filled: z72.boolean().default(false),
  fill_color: z72.string().optional(),
  is_dashed: z72.boolean().default(false),
  subcircuit_id: z72.string().optional()
}).describe("Draws a styled circle on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_arc.ts
import { z as z73 } from "zod";
var schematic_arc = z73.object({
  type: z73.literal("schematic_arc"),
  schematic_arc_id: getZodPrefixedIdWithDefault("schematic_arc"),
  schematic_component_id: z73.string().optional(),
  schematic_symbol_id: z73.string().optional(),
  center: point,
  radius: distance,
  start_angle_degrees: rotation,
  end_angle_degrees: rotation,
  direction: z73.enum(["clockwise", "counterclockwise"]).default("counterclockwise"),
  stroke_width: distance.nullable().optional(),
  color: z73.string().default("#000000"),
  is_dashed: z73.boolean().default(false),
  subcircuit_id: z73.string().optional()
}).describe("Draws a styled arc on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_trace.ts
import { z as z74 } from "zod";
var schematic_trace = z74.object({
  type: z74.literal("schematic_trace"),
  schematic_trace_id: z74.string(),
  source_trace_id: z74.string().optional(),
  junctions: z74.array(
    z74.object({
      x: z74.number(),
      y: z74.number()
    })
  ),
  edges: z74.array(
    z74.object({
      from: z74.object({
        x: z74.number(),
        y: z74.number()
      }),
      to: z74.object({
        x: z74.number(),
        y: z74.number()
      }),
      is_crossing: z74.boolean().optional(),
      from_schematic_port_id: z74.string().optional(),
      to_schematic_port_id: z74.string().optional()
    })
  ),
  subcircuit_id: z74.string().optional(),
  // TODO: make required in a future release
  subcircuit_connectivity_map_key: z74.string().optional()
});
expectTypesMatch(true);

// src/schematic/schematic_text.ts
import { z as z76 } from "zod";

// src/common/FivePointAnchor.ts
import { z as z75 } from "zod";
var fivePointAnchor = z75.enum([
  "center",
  "left",
  "right",
  "top",
  "bottom"
]);
expectTypesMatch(true);

// src/schematic/schematic_text.ts
var schematic_text = z76.object({
  type: z76.literal("schematic_text"),
  schematic_component_id: z76.string().optional(),
  schematic_symbol_id: z76.string().optional(),
  schematic_text_id: z76.string(),
  text: z76.string(),
  font_size: z76.number().default(0.18),
  position: z76.object({
    x: distance,
    y: distance
  }),
  rotation: z76.number().default(0),
  anchor: z76.union([fivePointAnchor.describe("legacy"), ninePointAnchor]).default("center"),
  color: z76.string().default("#000000"),
  subcircuit_id: z76.string().optional()
});
expectTypesMatch(true);

// src/schematic/schematic_port.ts
import { z as z77 } from "zod";
var schematic_port = z77.object({
  type: z77.literal("schematic_port"),
  schematic_port_id: z77.string(),
  source_port_id: z77.string(),
  schematic_component_id: z77.string().optional(),
  center: point,
  facing_direction: z77.enum(["up", "down", "left", "right"]).optional(),
  distance_from_component_edge: z77.number().optional(),
  side_of_component: z77.enum(["top", "bottom", "left", "right"]).optional(),
  true_ccw_index: z77.number().optional(),
  pin_number: z77.number().optional(),
  display_pin_label: z77.string().optional(),
  subcircuit_id: z77.string().optional(),
  is_connected: z77.boolean().optional(),
  has_input_arrow: z77.boolean().optional(),
  has_output_arrow: z77.boolean().optional(),
  is_drawn_with_inversion_circle: z77.boolean().optional()
}).describe("Defines a port on a schematic component");
expectTypesMatch(true);

// src/schematic/schematic_net_label.ts
import { z as z78 } from "zod";
var schematic_net_label = z78.object({
  type: z78.literal("schematic_net_label"),
  schematic_net_label_id: getZodPrefixedIdWithDefault("schematic_net_label"),
  schematic_trace_id: z78.string().optional(),
  source_trace_id: z78.string().optional(),
  source_net_id: z78.string(),
  center: point,
  anchor_position: point.optional(),
  anchor_side: z78.enum(["top", "bottom", "left", "right"]),
  text: z78.string(),
  symbol_name: z78.string().optional(),
  is_movable: z78.boolean().optional(),
  subcircuit_id: z78.string().optional()
});
expectTypesMatch(true);

// src/schematic/schematic_error.ts
import { z as z79 } from "zod";
var schematic_error = base_circuit_json_error.extend({
  type: z79.literal("schematic_error"),
  schematic_error_id: z79.string(),
  // eventually each error type should be broken out into a dir of files
  error_type: z79.literal("schematic_port_not_found").default("schematic_port_not_found"),
  subcircuit_id: z79.string().optional()
}).describe("Defines a schematic error on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_layout_error.ts
import { z as z80 } from "zod";
var schematic_layout_error = base_circuit_json_error.extend({
  type: z80.literal("schematic_layout_error"),
  schematic_layout_error_id: getZodPrefixedIdWithDefault(
    "schematic_layout_error"
  ),
  error_type: z80.literal("schematic_layout_error").default("schematic_layout_error"),
  source_group_id: z80.string(),
  schematic_group_id: z80.string(),
  subcircuit_id: z80.string().optional()
}).describe("Error emitted when schematic layout fails for a group");
expectTypesMatch(true);

// src/schematic/schematic_debug_object.ts
import { z as z81 } from "zod";
var schematic_debug_object_base = z81.object({
  type: z81.literal("schematic_debug_object"),
  label: z81.string().optional(),
  subcircuit_id: z81.string().optional()
});
var schematic_debug_rect = schematic_debug_object_base.extend({
  shape: z81.literal("rect"),
  center: point,
  size
});
var schematic_debug_line = schematic_debug_object_base.extend({
  shape: z81.literal("line"),
  start: point,
  end: point
});
var schematic_debug_point = schematic_debug_object_base.extend({
  shape: z81.literal("point"),
  center: point
});
var schematic_debug_object = z81.discriminatedUnion("shape", [
  schematic_debug_rect,
  schematic_debug_line,
  schematic_debug_point
]);
expectTypesMatch(true);

// src/schematic/schematic_voltage_probe.ts
import { z as z82 } from "zod";
var schematic_voltage_probe = z82.object({
  type: z82.literal("schematic_voltage_probe"),
  schematic_voltage_probe_id: z82.string(),
  source_component_id: z82.string().optional(),
  name: z82.string().optional(),
  position: point,
  schematic_trace_id: z82.string(),
  voltage: voltage.optional(),
  subcircuit_id: z82.string().optional(),
  color: z82.string().optional(),
  label_alignment: ninePointAnchor.optional()
}).describe("Defines a voltage probe measurement point on a schematic trace");
expectTypesMatch(true);

// src/schematic/schematic_manual_edit_conflict_warning.ts
import { z as z83 } from "zod";
var schematic_manual_edit_conflict_warning = z83.object({
  type: z83.literal("schematic_manual_edit_conflict_warning"),
  schematic_manual_edit_conflict_warning_id: getZodPrefixedIdWithDefault(
    "schematic_manual_edit_conflict_warning"
  ),
  warning_type: z83.literal("schematic_manual_edit_conflict_warning").default("schematic_manual_edit_conflict_warning"),
  message: z83.string(),
  schematic_component_id: z83.string(),
  schematic_group_id: z83.string().optional(),
  subcircuit_id: z83.string().optional(),
  source_component_id: z83.string()
}).describe(
  "Warning emitted when a component has both manual placement and explicit schX/schY coordinates"
);
expectTypesMatch(true);

// src/schematic/schematic_group.ts
import { z as z84 } from "zod";
var schematic_group = z84.object({
  type: z84.literal("schematic_group"),
  schematic_group_id: getZodPrefixedIdWithDefault("schematic_group"),
  source_group_id: z84.string(),
  is_subcircuit: z84.boolean().optional(),
  subcircuit_id: z84.string().optional(),
  width: length,
  height: length,
  center: point,
  schematic_component_ids: z84.array(z84.string()),
  show_as_schematic_box: z84.boolean().optional(),
  name: z84.string().optional(),
  description: z84.string().optional()
}).describe("Defines a group of components on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_table.ts
import { z as z85 } from "zod";
var schematic_table = z85.object({
  type: z85.literal("schematic_table"),
  schematic_table_id: getZodPrefixedIdWithDefault("schematic_table"),
  anchor_position: point,
  column_widths: z85.array(distance),
  row_heights: z85.array(distance),
  cell_padding: distance.optional(),
  border_width: distance.optional(),
  subcircuit_id: z85.string().optional(),
  schematic_component_id: z85.string().optional(),
  anchor: ninePointAnchor.optional()
}).describe("Defines a table on the schematic");
expectTypesMatch(true);

// src/schematic/schematic_table_cell.ts
import { z as z86 } from "zod";
var schematic_table_cell = z86.object({
  type: z86.literal("schematic_table_cell"),
  schematic_table_cell_id: getZodPrefixedIdWithDefault(
    "schematic_table_cell"
  ),
  schematic_table_id: z86.string(),
  start_row_index: z86.number(),
  end_row_index: z86.number(),
  start_column_index: z86.number(),
  end_column_index: z86.number(),
  text: z86.string().optional(),
  center: point,
  width: distance,
  height: distance,
  horizontal_align: z86.enum(["left", "center", "right"]).optional(),
  vertical_align: z86.enum(["top", "middle", "bottom"]).optional(),
  font_size: distance.optional(),
  subcircuit_id: z86.string().optional()
}).describe("Defines a cell within a schematic_table");
expectTypesMatch(true);

// src/schematic/schematic_sheet.ts
import { z as z87 } from "zod";
var schematic_sheet = z87.object({
  type: z87.literal("schematic_sheet"),
  schematic_sheet_id: getZodPrefixedIdWithDefault("schematic_sheet"),
  name: z87.string().optional(),
  subcircuit_id: z87.string().optional()
}).describe(
  "Defines a schematic sheet or page that components can be placed on"
);
expectTypesMatch(true);

// src/pcb/properties/brep.ts
import { z as z88 } from "zod";
var point_with_bulge = z88.object({
  x: distance,
  y: distance,
  bulge: z88.number().optional()
});
expectTypesMatch(true);
var ring = z88.object({
  vertices: z88.array(point_with_bulge)
});
expectTypesMatch(true);
var brep_shape = z88.object({
  outer_ring: ring,
  inner_rings: z88.array(ring).default([])
});
expectTypesMatch(true);

// src/pcb/properties/pcb_route_hints.ts
import { z as z89 } from "zod";
var pcb_route_hint = z89.object({
  x: distance,
  y: distance,
  via: z89.boolean().optional(),
  via_to_layer: layer_ref.optional()
});
var pcb_route_hints = z89.array(pcb_route_hint);
expectTypesMatch(true);
expectTypesMatch(true);

// src/pcb/properties/route_hint_point.ts
import { z as z90 } from "zod";
var route_hint_point = z90.object({
  x: distance,
  y: distance,
  via: z90.boolean().optional(),
  to_layer: layer_ref.optional(),
  trace_width: distance.optional()
});
expectTypesMatch(true);

// src/pcb/pcb_component.ts
import { z as z91 } from "zod";
var pcb_component = z91.object({
  type: z91.literal("pcb_component"),
  pcb_component_id: getZodPrefixedIdWithDefault("pcb_component"),
  source_component_id: z91.string(),
  center: point,
  layer: layer_ref,
  rotation,
  display_offset_x: z91.string().optional().describe(
    "How to display the x offset for this part, usually corresponding with how the user specified it"
  ),
  display_offset_y: z91.string().optional().describe(
    "How to display the y offset for this part, usually corresponding with how the user specified it"
  ),
  width: length,
  height: length,
  do_not_place: z91.boolean().optional(),
  is_allowed_to_be_off_board: z91.boolean().optional(),
  subcircuit_id: z91.string().optional(),
  pcb_group_id: z91.string().optional(),
  position_mode: z91.enum([
    "packed",
    "relative_to_group_anchor",
    "relative_to_another_component",
    "none"
  ]).optional(),
  anchor_position: point.optional(),
  anchor_alignment: ninePointAnchor.optional(),
  positioned_relative_to_pcb_group_id: z91.string().optional(),
  positioned_relative_to_pcb_board_id: z91.string().optional(),
  cable_insertion_center: point.optional(),
  metadata: z91.object({
    kicad_footprint: kicadFootprintMetadata.optional()
  }).optional(),
  obstructs_within_bounds: z91.boolean().default(true).describe(
    "Does this component take up all the space within its bounds on a layer. This is generally true except for when separated pin headers are being represented by a single component (in which case, chips can be placed between the pin headers) or for tall modules where chips fit underneath"
  )
}).describe("Defines a component on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_hole.ts
import { z as z92 } from "zod";
var pcb_hole_circle = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.literal("circle"),
  hole_diameter: z92.number(),
  x: distance,
  y: distance,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_circle_shape = pcb_hole_circle.describe(
  "Defines a circular hole on the PCB"
);
expectTypesMatch(true);
var pcb_hole_rect = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.literal("rect"),
  hole_width: z92.number(),
  hole_height: z92.number(),
  x: distance,
  y: distance,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_rect_shape = pcb_hole_rect.describe(
  "Defines a rectangular (square-capable) hole on the PCB. Use equal width/height for square."
);
expectTypesMatch(true);
var pcb_hole_circle_or_square = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.enum(["circle", "square"]),
  hole_diameter: z92.number(),
  x: distance,
  y: distance,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_circle_or_square_shape = pcb_hole_circle_or_square.describe(
  "Defines a circular or square hole on the PCB"
);
expectTypesMatch(true);
var pcb_hole_oval = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.literal("oval"),
  hole_width: z92.number(),
  hole_height: z92.number(),
  x: distance,
  y: distance,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_oval_shape = pcb_hole_oval.describe(
  "Defines an oval hole on the PCB"
);
expectTypesMatch(true);
var pcb_hole_pill = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.literal("pill"),
  hole_width: z92.number(),
  hole_height: z92.number(),
  x: distance,
  y: distance,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_pill_shape = pcb_hole_pill.describe(
  "Defines a pill-shaped hole on the PCB"
);
expectTypesMatch(true);
var pcb_hole_rotated_pill = z92.object({
  type: z92.literal("pcb_hole"),
  pcb_hole_id: getZodPrefixedIdWithDefault("pcb_hole"),
  pcb_group_id: z92.string().optional(),
  subcircuit_id: z92.string().optional(),
  pcb_component_id: z92.string().optional(),
  hole_shape: z92.literal("rotated_pill"),
  hole_width: z92.number(),
  hole_height: z92.number(),
  x: distance,
  y: distance,
  ccw_rotation: rotation,
  is_covered_with_solder_mask: z92.boolean().optional(),
  soldermask_margin: z92.number().optional()
});
var pcb_hole_rotated_pill_shape = pcb_hole_rotated_pill.describe(
  "Defines a rotated pill-shaped hole on the PCB"
);
expectTypesMatch(true);
var pcb_hole = pcb_hole_circle_or_square.or(pcb_hole_oval).or(pcb_hole_pill).or(pcb_hole_rotated_pill).or(pcb_hole_circle).or(pcb_hole_rect);

// src/pcb/pcb_plated_hole.ts
import { z as z93 } from "zod";
var pcb_plated_hole_circle = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.literal("circle"),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  outer_diameter: z93.number(),
  hole_diameter: z93.number(),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional()
});
var pcb_plated_hole_oval = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.enum(["oval", "pill"]),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  outer_width: z93.number(),
  outer_height: z93.number(),
  hole_width: z93.number(),
  hole_height: z93.number(),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  ccw_rotation: rotation,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional()
});
var pcb_circular_hole_with_rect_pad = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.literal("circular_hole_with_rect_pad"),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  hole_shape: z93.literal("circle"),
  pad_shape: z93.literal("rect"),
  hole_diameter: z93.number(),
  rect_pad_width: z93.number(),
  rect_pad_height: z93.number(),
  rect_border_radius: z93.number().optional(),
  hole_offset_x: distance.default(0),
  hole_offset_y: distance.default(0),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional(),
  rect_ccw_rotation: rotation.optional()
});
var pcb_pill_hole_with_rect_pad = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.literal("pill_hole_with_rect_pad"),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  hole_shape: z93.literal("pill"),
  pad_shape: z93.literal("rect"),
  hole_width: z93.number(),
  hole_height: z93.number(),
  rect_pad_width: z93.number(),
  rect_pad_height: z93.number(),
  rect_border_radius: z93.number().optional(),
  hole_offset_x: distance.default(0),
  hole_offset_y: distance.default(0),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional()
});
var pcb_rotated_pill_hole_with_rect_pad = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.literal("rotated_pill_hole_with_rect_pad"),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  hole_shape: z93.literal("rotated_pill"),
  pad_shape: z93.literal("rect"),
  hole_width: z93.number(),
  hole_height: z93.number(),
  hole_ccw_rotation: rotation,
  rect_pad_width: z93.number(),
  rect_pad_height: z93.number(),
  rect_border_radius: z93.number().optional(),
  rect_ccw_rotation: rotation,
  hole_offset_x: distance.default(0),
  hole_offset_y: distance.default(0),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional()
});
var pcb_hole_with_polygon_pad = z93.object({
  type: z93.literal("pcb_plated_hole"),
  shape: z93.literal("hole_with_polygon_pad"),
  pcb_group_id: z93.string().optional(),
  subcircuit_id: z93.string().optional(),
  hole_shape: z93.enum(["circle", "oval", "pill", "rotated_pill"]),
  hole_diameter: z93.number().optional(),
  hole_width: z93.number().optional(),
  hole_height: z93.number().optional(),
  pad_outline: z93.array(
    z93.object({
      x: distance,
      y: distance
    })
  ).min(3),
  hole_offset_x: distance.default(0),
  hole_offset_y: distance.default(0),
  is_covered_with_solder_mask: z93.boolean().optional(),
  x: distance,
  y: distance,
  layers: z93.array(layer_ref),
  port_hints: z93.array(z93.string()).optional(),
  pcb_component_id: z93.string().optional(),
  pcb_port_id: z93.string().optional(),
  pcb_plated_hole_id: getZodPrefixedIdWithDefault("pcb_plated_hole"),
  soldermask_margin: z93.number().optional(),
  ccw_rotation: rotation.optional()
});
var pcb_plated_hole = z93.union([
  pcb_plated_hole_circle,
  pcb_plated_hole_oval,
  pcb_circular_hole_with_rect_pad,
  pcb_pill_hole_with_rect_pad,
  pcb_rotated_pill_hole_with_rect_pad,
  pcb_hole_with_polygon_pad
]);
expectTypesMatch(
  true
);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);

// src/pcb/pcb_port.ts
import { z as z94 } from "zod";
var pcb_port = z94.object({
  type: z94.literal("pcb_port"),
  pcb_port_id: getZodPrefixedIdWithDefault("pcb_port"),
  pcb_group_id: z94.string().optional(),
  subcircuit_id: z94.string().optional(),
  source_port_id: z94.string(),
  pcb_component_id: z94.string().optional(),
  x: distance,
  y: distance,
  layers: z94.array(layer_ref),
  is_board_pinout: z94.boolean().optional()
}).describe("Defines a port on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_smtpad.ts
import { z as z95 } from "zod";
var pcb_smtpad_circle = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("circle"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  x: distance,
  y: distance,
  radius: z95.number(),
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional()
});
var pcb_smtpad_rect = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("rect"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  x: distance,
  y: distance,
  width: z95.number(),
  height: z95.number(),
  rect_border_radius: z95.number().optional(),
  corner_radius: z95.number().optional(),
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional(),
  soldermask_margin_left: z95.number().optional(),
  soldermask_margin_top: z95.number().optional(),
  soldermask_margin_right: z95.number().optional(),
  soldermask_margin_bottom: z95.number().optional()
});
var pcb_smtpad_rotated_rect = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("rotated_rect"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  x: distance,
  y: distance,
  width: z95.number(),
  height: z95.number(),
  rect_border_radius: z95.number().optional(),
  corner_radius: z95.number().optional(),
  ccw_rotation: rotation,
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional(),
  soldermask_margin_left: z95.number().optional(),
  soldermask_margin_top: z95.number().optional(),
  soldermask_margin_right: z95.number().optional(),
  soldermask_margin_bottom: z95.number().optional()
});
var pcb_smtpad_pill = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("pill"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  x: distance,
  y: distance,
  width: z95.number(),
  height: z95.number(),
  radius: z95.number(),
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional()
});
var pcb_smtpad_rotated_pill = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("rotated_pill"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  x: distance,
  y: distance,
  width: z95.number(),
  height: z95.number(),
  radius: z95.number(),
  ccw_rotation: rotation,
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional()
});
var pcb_smtpad_polygon = z95.object({
  type: z95.literal("pcb_smtpad"),
  shape: z95.literal("polygon"),
  pcb_smtpad_id: getZodPrefixedIdWithDefault("pcb_smtpad"),
  pcb_group_id: z95.string().optional(),
  subcircuit_id: z95.string().optional(),
  points: z95.array(point),
  layer: layer_ref,
  port_hints: z95.array(z95.string()).optional(),
  pcb_component_id: z95.string().optional(),
  pcb_port_id: z95.string().optional(),
  is_covered_with_solder_mask: z95.boolean().optional(),
  soldermask_margin: z95.number().optional()
});
var pcb_smtpad = z95.discriminatedUnion("shape", [
  pcb_smtpad_circle,
  pcb_smtpad_rect,
  pcb_smtpad_rotated_rect,
  pcb_smtpad_rotated_pill,
  pcb_smtpad_pill,
  pcb_smtpad_polygon
]).describe("Defines an SMT pad on the PCB");
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);

// src/pcb/pcb_solder_paste.ts
import { z as z96 } from "zod";
var pcb_solder_paste_circle = z96.object({
  type: z96.literal("pcb_solder_paste"),
  shape: z96.literal("circle"),
  pcb_solder_paste_id: getZodPrefixedIdWithDefault("pcb_solder_paste"),
  pcb_group_id: z96.string().optional(),
  subcircuit_id: z96.string().optional(),
  x: distance,
  y: distance,
  radius: z96.number(),
  layer: layer_ref,
  pcb_component_id: z96.string().optional(),
  pcb_smtpad_id: z96.string().optional()
});
var pcb_solder_paste_rect = z96.object({
  type: z96.literal("pcb_solder_paste"),
  shape: z96.literal("rect"),
  pcb_solder_paste_id: getZodPrefixedIdWithDefault("pcb_solder_paste"),
  pcb_group_id: z96.string().optional(),
  subcircuit_id: z96.string().optional(),
  x: distance,
  y: distance,
  width: z96.number(),
  height: z96.number(),
  layer: layer_ref,
  pcb_component_id: z96.string().optional(),
  pcb_smtpad_id: z96.string().optional()
});
var pcb_solder_paste_pill = z96.object({
  type: z96.literal("pcb_solder_paste"),
  shape: z96.literal("pill"),
  pcb_solder_paste_id: getZodPrefixedIdWithDefault("pcb_solder_paste"),
  pcb_group_id: z96.string().optional(),
  subcircuit_id: z96.string().optional(),
  x: distance,
  y: distance,
  width: z96.number(),
  height: z96.number(),
  radius: z96.number(),
  layer: layer_ref,
  pcb_component_id: z96.string().optional(),
  pcb_smtpad_id: z96.string().optional()
});
var pcb_solder_paste_rotated_rect = z96.object({
  type: z96.literal("pcb_solder_paste"),
  shape: z96.literal("rotated_rect"),
  pcb_solder_paste_id: getZodPrefixedIdWithDefault("pcb_solder_paste"),
  pcb_group_id: z96.string().optional(),
  subcircuit_id: z96.string().optional(),
  x: distance,
  y: distance,
  width: z96.number(),
  height: z96.number(),
  ccw_rotation: distance,
  layer: layer_ref,
  pcb_component_id: z96.string().optional(),
  pcb_smtpad_id: z96.string().optional()
});
var pcb_solder_paste_oval = z96.object({
  type: z96.literal("pcb_solder_paste"),
  shape: z96.literal("oval"),
  pcb_solder_paste_id: getZodPrefixedIdWithDefault("pcb_solder_paste"),
  pcb_group_id: z96.string().optional(),
  subcircuit_id: z96.string().optional(),
  x: distance,
  y: distance,
  width: z96.number(),
  height: z96.number(),
  layer: layer_ref,
  pcb_component_id: z96.string().optional(),
  pcb_smtpad_id: z96.string().optional()
});
var pcb_solder_paste = z96.union([
  pcb_solder_paste_circle,
  pcb_solder_paste_rect,
  pcb_solder_paste_pill,
  pcb_solder_paste_rotated_rect,
  pcb_solder_paste_oval
]).describe("Defines solderpaste on the PCB");
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(
  true
);
expectTypesMatch(true);

// src/pcb/pcb_text.ts
import { z as z97 } from "zod";
var pcb_text = z97.object({
  type: z97.literal("pcb_text"),
  pcb_text_id: getZodPrefixedIdWithDefault("pcb_text"),
  pcb_group_id: z97.string().optional(),
  subcircuit_id: z97.string().optional(),
  text: z97.string(),
  center: point,
  layer: layer_ref,
  width: length,
  height: length,
  lines: z97.number(),
  // @ts-ignore
  align: z97.enum(["bottom-left"])
}).describe("Defines text on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_trace.ts
import { z as z98 } from "zod";
var pcb_trace_route_point_wire = z98.object({
  route_type: z98.literal("wire"),
  x: distance,
  y: distance,
  width: distance,
  copper_pour_id: z98.string().optional(),
  is_inside_copper_pour: z98.boolean().optional(),
  start_pcb_port_id: z98.string().optional(),
  end_pcb_port_id: z98.string().optional(),
  layer: layer_ref
});
var pcb_trace_route_point_via = z98.object({
  route_type: z98.literal("via"),
  x: distance,
  y: distance,
  copper_pour_id: z98.string().optional(),
  is_inside_copper_pour: z98.boolean().optional(),
  hole_diameter: distance.optional(),
  outer_diameter: distance.optional(),
  from_layer: z98.string(),
  to_layer: z98.string()
});
var pcb_trace_route_point = z98.union([
  pcb_trace_route_point_wire,
  pcb_trace_route_point_via
]);
var pcb_trace = z98.object({
  type: z98.literal("pcb_trace"),
  source_trace_id: z98.string().optional(),
  pcb_component_id: z98.string().optional(),
  pcb_trace_id: getZodPrefixedIdWithDefault("pcb_trace"),
  pcb_group_id: z98.string().optional(),
  subcircuit_id: z98.string().optional(),
  route_thickness_mode: z98.enum(["constant", "interpolated"]).default("constant").optional(),
  route_order_index: z98.number().optional(),
  should_round_corners: z98.boolean().optional(),
  trace_length: z98.number().optional(),
  highlight_color: z98.string().optional(),
  route: z98.array(pcb_trace_route_point)
}).describe("Defines a trace on the PCB");
expectTypesMatch(true);
expectTypesMatch(true);

// src/pcb/pcb_trace_error.ts
import { z as z99 } from "zod";
var pcb_trace_error = base_circuit_json_error.extend({
  type: z99.literal("pcb_trace_error"),
  pcb_trace_error_id: getZodPrefixedIdWithDefault("pcb_trace_error"),
  error_type: z99.literal("pcb_trace_error").default("pcb_trace_error"),
  center: point.optional(),
  pcb_trace_id: z99.string(),
  source_trace_id: z99.string(),
  pcb_component_ids: z99.array(z99.string()),
  pcb_port_ids: z99.array(z99.string()),
  subcircuit_id: z99.string().optional()
}).describe("Defines a trace error on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_trace_missing_error.ts
import { z as z100 } from "zod";
var pcb_trace_missing_error = base_circuit_json_error.extend({
  type: z100.literal("pcb_trace_missing_error"),
  pcb_trace_missing_error_id: getZodPrefixedIdWithDefault(
    "pcb_trace_missing_error"
  ),
  error_type: z100.literal("pcb_trace_missing_error").default("pcb_trace_missing_error"),
  center: point.optional(),
  source_trace_id: z100.string(),
  pcb_component_ids: z100.array(z100.string()),
  pcb_port_ids: z100.array(z100.string()),
  subcircuit_id: z100.string().optional()
}).describe(
  "Defines an error when a source trace has no corresponding PCB trace"
);
expectTypesMatch(true);

// src/pcb/pcb_port_not_matched_error.ts
import { z as z101 } from "zod";
var pcb_port_not_matched_error = base_circuit_json_error.extend({
  type: z101.literal("pcb_port_not_matched_error"),
  pcb_error_id: getZodPrefixedIdWithDefault("pcb_error"),
  error_type: z101.literal("pcb_port_not_matched_error").default("pcb_port_not_matched_error"),
  pcb_component_ids: z101.array(z101.string()),
  subcircuit_id: z101.string().optional()
}).describe("Defines a trace error on the PCB where a port is not matched");
expectTypesMatch(true);

// src/pcb/pcb_port_not_connected_error.ts
import { z as z102 } from "zod";
var pcb_port_not_connected_error = base_circuit_json_error.extend({
  type: z102.literal("pcb_port_not_connected_error"),
  pcb_port_not_connected_error_id: getZodPrefixedIdWithDefault(
    "pcb_port_not_connected_error"
  ),
  error_type: z102.literal("pcb_port_not_connected_error").default("pcb_port_not_connected_error"),
  pcb_port_ids: z102.array(z102.string()),
  pcb_component_ids: z102.array(z102.string()),
  subcircuit_id: z102.string().optional()
}).describe("Defines an error when a pcb port is not connected to any trace");
expectTypesMatch(
  true
);

// src/pcb/pcb_net.ts
import { z as z103 } from "zod";
var pcb_net = z103.object({
  type: z103.literal("pcb_net"),
  pcb_net_id: getZodPrefixedIdWithDefault("pcb_net"),
  source_net_id: z103.string().optional(),
  highlight_color: z103.string().optional()
}).describe("Defines a net on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_via.ts
import { z as z104 } from "zod";
var pcb_via = z104.object({
  type: z104.literal("pcb_via"),
  pcb_via_id: getZodPrefixedIdWithDefault("pcb_via"),
  pcb_group_id: z104.string().optional(),
  subcircuit_id: z104.string().optional(),
  subcircuit_connectivity_map_key: z104.string().optional(),
  x: distance,
  y: distance,
  outer_diameter: distance.default("0.6mm"),
  hole_diameter: distance.default("0.25mm"),
  /** @deprecated */
  from_layer: layer_ref.optional(),
  /** @deprecated */
  to_layer: layer_ref.optional(),
  layers: z104.array(layer_ref),
  pcb_trace_id: z104.string().optional(),
  net_is_assignable: z104.boolean().optional(),
  net_assigned: z104.boolean().optional(),
  is_tented: z104.boolean().optional()
}).describe("Defines a via on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_board.ts
import { z as z105 } from "zod";
var pcb_board = z105.object({
  type: z105.literal("pcb_board"),
  pcb_board_id: getZodPrefixedIdWithDefault("pcb_board"),
  pcb_panel_id: z105.string().optional(),
  carrier_pcb_board_id: z105.string().optional(),
  is_subcircuit: z105.boolean().optional(),
  subcircuit_id: z105.string().optional(),
  is_mounted_to_carrier_board: z105.boolean().optional(),
  width: length.optional(),
  height: length.optional(),
  center: point,
  display_offset_x: z105.string().optional().describe(
    "How to display the x offset for this board, usually corresponding with how the user specified it"
  ),
  display_offset_y: z105.string().optional().describe(
    "How to display the y offset for this board, usually corresponding with how the user specified it"
  ),
  thickness: length.optional().default(1.4),
  num_layers: z105.number().optional().default(4),
  outline: z105.array(point).optional(),
  shape: z105.enum(["rect", "polygon"]).optional(),
  material: z105.enum(["fr4", "fr1"]).default("fr4"),
  anchor_position: point.optional(),
  anchor_alignment: ninePointAnchor.optional(),
  position_mode: z105.enum(["relative_to_panel_anchor", "none"]).optional()
}).describe("Defines the board outline of the PCB");
expectTypesMatch(true);

// src/pcb/pcb_panel.ts
import { z as z106 } from "zod";
var pcb_panel = z106.object({
  type: z106.literal("pcb_panel"),
  pcb_panel_id: getZodPrefixedIdWithDefault("pcb_panel"),
  width: length,
  height: length,
  center: point,
  covered_with_solder_mask: z106.boolean().optional().default(true)
}).describe("Defines a PCB panel that can contain multiple boards");
expectTypesMatch(true);

// src/pcb/pcb_placement_error.ts
import { z as z107 } from "zod";
var pcb_placement_error = base_circuit_json_error.extend({
  type: z107.literal("pcb_placement_error"),
  pcb_placement_error_id: getZodPrefixedIdWithDefault("pcb_placement_error"),
  error_type: z107.literal("pcb_placement_error").default("pcb_placement_error"),
  subcircuit_id: z107.string().optional()
}).describe("Defines a placement error on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_panelization_placement_error.ts
import { z as z108 } from "zod";
var pcb_panelization_placement_error = base_circuit_json_error.extend({
  type: z108.literal("pcb_panelization_placement_error"),
  pcb_panelization_placement_error_id: getZodPrefixedIdWithDefault(
    "pcb_panelization_placement_error"
  ),
  error_type: z108.literal("pcb_panelization_placement_error").default("pcb_panelization_placement_error"),
  pcb_panel_id: z108.string().optional(),
  pcb_board_id: z108.string().optional(),
  subcircuit_id: z108.string().optional()
}).describe("Defines a panelization placement error on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_trace_hint.ts
import { z as z109 } from "zod";
var pcb_trace_hint = z109.object({
  type: z109.literal("pcb_trace_hint"),
  pcb_trace_hint_id: getZodPrefixedIdWithDefault("pcb_trace_hint"),
  pcb_port_id: z109.string(),
  pcb_component_id: z109.string(),
  route: z109.array(route_hint_point),
  subcircuit_id: z109.string().optional()
}).describe("A hint that can be used during generation of a PCB trace");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_line.ts
import { z as z110 } from "zod";
var pcb_silkscreen_line = z110.object({
  type: z110.literal("pcb_silkscreen_line"),
  pcb_silkscreen_line_id: getZodPrefixedIdWithDefault("pcb_silkscreen_line"),
  pcb_component_id: z110.string(),
  pcb_group_id: z110.string().optional(),
  subcircuit_id: z110.string().optional(),
  stroke_width: distance.default("0.1mm"),
  x1: distance,
  y1: distance,
  x2: distance,
  y2: distance,
  layer: visible_layer
}).describe("Defines a silkscreen line on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_path.ts
import { z as z111 } from "zod";
var pcb_silkscreen_path = z111.object({
  type: z111.literal("pcb_silkscreen_path"),
  pcb_silkscreen_path_id: getZodPrefixedIdWithDefault("pcb_silkscreen_path"),
  pcb_component_id: z111.string(),
  pcb_group_id: z111.string().optional(),
  subcircuit_id: z111.string().optional(),
  layer: visible_layer,
  route: z111.array(point),
  stroke_width: length
}).describe("Defines a silkscreen path on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_text.ts
import { z as z112 } from "zod";
var pcb_silkscreen_text = z112.object({
  type: z112.literal("pcb_silkscreen_text"),
  pcb_silkscreen_text_id: getZodPrefixedIdWithDefault("pcb_silkscreen_text"),
  pcb_group_id: z112.string().optional(),
  subcircuit_id: z112.string().optional(),
  font: z112.literal("tscircuit2024").default("tscircuit2024"),
  font_size: distance.default("0.2mm"),
  pcb_component_id: z112.string(),
  text: z112.string(),
  is_knockout: z112.boolean().default(false).optional(),
  knockout_padding: z112.object({
    left: length,
    top: length,
    bottom: length,
    right: length
  }).default({
    left: "0.2mm",
    top: "0.2mm",
    bottom: "0.2mm",
    right: "0.2mm"
  }).optional(),
  ccw_rotation: z112.number().optional(),
  layer: layer_ref,
  is_mirrored: z112.boolean().default(false).optional(),
  anchor_position: point.default({ x: 0, y: 0 }),
  anchor_alignment: ninePointAnchor.default("center")
}).describe("Defines silkscreen text on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_copper_text.ts
import { z as z113 } from "zod";
var pcb_copper_text = z113.object({
  type: z113.literal("pcb_copper_text"),
  pcb_copper_text_id: getZodPrefixedIdWithDefault("pcb_copper_text"),
  pcb_group_id: z113.string().optional(),
  subcircuit_id: z113.string().optional(),
  font: z113.literal("tscircuit2024").default("tscircuit2024"),
  font_size: distance.default("0.2mm"),
  pcb_component_id: z113.string(),
  text: z113.string(),
  is_knockout: z113.boolean().default(false).optional(),
  knockout_padding: z113.object({
    left: length,
    top: length,
    bottom: length,
    right: length
  }).default({
    left: "0.2mm",
    top: "0.2mm",
    bottom: "0.2mm",
    right: "0.2mm"
  }).optional(),
  ccw_rotation: z113.number().optional(),
  layer: layer_ref,
  is_mirrored: z113.boolean().default(false).optional(),
  anchor_position: point.default({ x: 0, y: 0 }),
  anchor_alignment: ninePointAnchor.default("center")
}).describe("Defines copper text on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_rect.ts
import { z as z114 } from "zod";
var pcb_silkscreen_rect = z114.object({
  type: z114.literal("pcb_silkscreen_rect"),
  pcb_silkscreen_rect_id: getZodPrefixedIdWithDefault("pcb_silkscreen_rect"),
  pcb_component_id: z114.string(),
  pcb_group_id: z114.string().optional(),
  subcircuit_id: z114.string().optional(),
  center: point,
  width: length,
  height: length,
  layer: layer_ref,
  stroke_width: length.default("1mm"),
  corner_radius: length.optional(),
  is_filled: z114.boolean().default(true).optional(),
  has_stroke: z114.boolean().optional(),
  is_stroke_dashed: z114.boolean().optional(),
  ccw_rotation: z114.number().optional()
}).describe("Defines a silkscreen rect on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_circle.ts
import { z as z115 } from "zod";
var pcb_silkscreen_circle = z115.object({
  type: z115.literal("pcb_silkscreen_circle"),
  pcb_silkscreen_circle_id: getZodPrefixedIdWithDefault(
    "pcb_silkscreen_circle"
  ),
  pcb_component_id: z115.string(),
  pcb_group_id: z115.string().optional(),
  subcircuit_id: z115.string().optional(),
  center: point,
  radius: length,
  layer: visible_layer,
  stroke_width: length.default("1mm")
}).describe("Defines a silkscreen circle on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_oval.ts
import { z as z116 } from "zod";
var pcb_silkscreen_oval = z116.object({
  type: z116.literal("pcb_silkscreen_oval"),
  pcb_silkscreen_oval_id: getZodPrefixedIdWithDefault("pcb_silkscreen_oval"),
  pcb_component_id: z116.string(),
  pcb_group_id: z116.string().optional(),
  subcircuit_id: z116.string().optional(),
  center: point,
  radius_x: distance,
  radius_y: distance,
  layer: visible_layer,
  ccw_rotation: rotation.optional()
}).describe("Defines a silkscreen oval on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_silkscreen_pill.ts
import { z as z117 } from "zod";
var pcb_silkscreen_pill = z117.object({
  type: z117.literal("pcb_silkscreen_pill"),
  pcb_silkscreen_pill_id: getZodPrefixedIdWithDefault("pcb_silkscreen_pill"),
  pcb_component_id: z117.string(),
  pcb_group_id: z117.string().optional(),
  subcircuit_id: z117.string().optional(),
  center: point,
  width: length,
  height: length,
  layer: layer_ref,
  ccw_rotation: z117.number().optional()
}).describe("Defines a silkscreen pill on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_fabrication_note_text.ts
import { z as z118 } from "zod";
var pcb_fabrication_note_text = z118.object({
  type: z118.literal("pcb_fabrication_note_text"),
  pcb_fabrication_note_text_id: getZodPrefixedIdWithDefault(
    "pcb_fabrication_note_text"
  ),
  subcircuit_id: z118.string().optional(),
  pcb_group_id: z118.string().optional(),
  font: z118.literal("tscircuit2024").default("tscircuit2024"),
  font_size: distance.default("1mm"),
  pcb_component_id: z118.string(),
  text: z118.string(),
  layer: visible_layer,
  anchor_position: point.default({ x: 0, y: 0 }),
  anchor_alignment: z118.enum(["center", "top_left", "top_right", "bottom_left", "bottom_right"]).default("center"),
  color: z118.string().optional()
}).describe(
  "Defines a fabrication note in text on the PCB, useful for leaving notes for assemblers or fabricators"
);
expectTypesMatch(true);

// src/pcb/pcb_fabrication_note_path.ts
import { z as z119 } from "zod";
var pcb_fabrication_note_path = z119.object({
  type: z119.literal("pcb_fabrication_note_path"),
  pcb_fabrication_note_path_id: getZodPrefixedIdWithDefault(
    "pcb_fabrication_note_path"
  ),
  pcb_component_id: z119.string(),
  subcircuit_id: z119.string().optional(),
  layer: layer_ref,
  route: z119.array(point),
  stroke_width: length,
  color: z119.string().optional()
}).describe(
  "Defines a fabrication path on the PCB for fabricators or assemblers"
);
expectTypesMatch(true);

// src/pcb/pcb_fabrication_note_rect.ts
import { z as z120 } from "zod";
var pcb_fabrication_note_rect = z120.object({
  type: z120.literal("pcb_fabrication_note_rect"),
  pcb_fabrication_note_rect_id: getZodPrefixedIdWithDefault(
    "pcb_fabrication_note_rect"
  ),
  pcb_component_id: z120.string(),
  pcb_group_id: z120.string().optional(),
  subcircuit_id: z120.string().optional(),
  center: point,
  width: length,
  height: length,
  layer: visible_layer,
  stroke_width: length.default("0.1mm"),
  corner_radius: length.optional(),
  is_filled: z120.boolean().optional(),
  has_stroke: z120.boolean().optional(),
  is_stroke_dashed: z120.boolean().optional(),
  color: z120.string().optional()
}).describe("Defines a fabrication note rectangle on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_fabrication_note_dimension.ts
import { z as z121 } from "zod";
var pcb_fabrication_note_dimension = z121.object({
  type: z121.literal("pcb_fabrication_note_dimension"),
  pcb_fabrication_note_dimension_id: getZodPrefixedIdWithDefault(
    "pcb_fabrication_note_dimension"
  ),
  pcb_component_id: z121.string(),
  pcb_group_id: z121.string().optional(),
  subcircuit_id: z121.string().optional(),
  layer: visible_layer,
  from: point,
  to: point,
  text: z121.string().optional(),
  text_ccw_rotation: z121.number().optional(),
  offset: length.optional(),
  offset_distance: length.optional(),
  offset_direction: z121.object({
    x: z121.number(),
    y: z121.number()
  }).optional(),
  font: z121.literal("tscircuit2024").default("tscircuit2024"),
  font_size: length.default("1mm"),
  color: z121.string().optional(),
  arrow_size: length.default("1mm")
}).describe("Defines a measurement annotation within PCB fabrication notes");
expectTypesMatch(true);

// src/pcb/pcb_note_text.ts
import { z as z122 } from "zod";
var pcb_note_text = z122.object({
  type: z122.literal("pcb_note_text"),
  pcb_note_text_id: getZodPrefixedIdWithDefault("pcb_note_text"),
  pcb_component_id: z122.string().optional(),
  pcb_group_id: z122.string().optional(),
  subcircuit_id: z122.string().optional(),
  name: z122.string().optional(),
  font: z122.literal("tscircuit2024").default("tscircuit2024"),
  font_size: distance.default("1mm"),
  text: z122.string().optional(),
  anchor_position: point.default({ x: 0, y: 0 }),
  anchor_alignment: z122.enum(["center", "top_left", "top_right", "bottom_left", "bottom_right"]).default("center"),
  color: z122.string().optional()
}).describe("Defines a documentation note in text on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_note_rect.ts
import { z as z123 } from "zod";
var pcb_note_rect = z123.object({
  type: z123.literal("pcb_note_rect"),
  pcb_note_rect_id: getZodPrefixedIdWithDefault("pcb_note_rect"),
  pcb_component_id: z123.string().optional(),
  pcb_group_id: z123.string().optional(),
  subcircuit_id: z123.string().optional(),
  name: z123.string().optional(),
  text: z123.string().optional(),
  center: point,
  width: length,
  height: length,
  stroke_width: length.default("0.1mm"),
  corner_radius: length.optional(),
  is_filled: z123.boolean().optional(),
  has_stroke: z123.boolean().optional(),
  is_stroke_dashed: z123.boolean().optional(),
  color: z123.string().optional()
}).describe("Defines a rectangular documentation note on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_note_path.ts
import { z as z124 } from "zod";
var pcb_note_path = z124.object({
  type: z124.literal("pcb_note_path"),
  pcb_note_path_id: getZodPrefixedIdWithDefault("pcb_note_path"),
  pcb_component_id: z124.string().optional(),
  pcb_group_id: z124.string().optional(),
  subcircuit_id: z124.string().optional(),
  name: z124.string().optional(),
  text: z124.string().optional(),
  route: z124.array(point),
  stroke_width: length.default("0.1mm"),
  color: z124.string().optional()
}).describe("Defines a polyline documentation note on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_note_line.ts
import { z as z125 } from "zod";
var pcb_note_line = z125.object({
  type: z125.literal("pcb_note_line"),
  pcb_note_line_id: getZodPrefixedIdWithDefault("pcb_note_line"),
  pcb_component_id: z125.string().optional(),
  pcb_group_id: z125.string().optional(),
  subcircuit_id: z125.string().optional(),
  name: z125.string().optional(),
  text: z125.string().optional(),
  x1: distance,
  y1: distance,
  x2: distance,
  y2: distance,
  stroke_width: distance.default("0.1mm"),
  color: z125.string().optional(),
  is_dashed: z125.boolean().optional()
}).describe("Defines a straight documentation note line on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_note_dimension.ts
import { z as z126 } from "zod";
var pcb_note_dimension = z126.object({
  type: z126.literal("pcb_note_dimension"),
  pcb_note_dimension_id: getZodPrefixedIdWithDefault("pcb_note_dimension"),
  pcb_component_id: z126.string().optional(),
  pcb_group_id: z126.string().optional(),
  subcircuit_id: z126.string().optional(),
  name: z126.string().optional(),
  from: point,
  to: point,
  text: z126.string().optional(),
  text_ccw_rotation: z126.number().optional(),
  offset_distance: length.optional(),
  offset_direction: z126.object({
    x: z126.number(),
    y: z126.number()
  }).optional(),
  font: z126.literal("tscircuit2024").default("tscircuit2024"),
  font_size: length.default("1mm"),
  color: z126.string().optional(),
  arrow_size: length.default("1mm")
}).describe("Defines a measurement annotation within PCB documentation notes");
expectTypesMatch(true);

// src/pcb/pcb_footprint_overlap_error.ts
import { z as z127 } from "zod";
var pcb_footprint_overlap_error = base_circuit_json_error.extend({
  type: z127.literal("pcb_footprint_overlap_error"),
  pcb_error_id: getZodPrefixedIdWithDefault("pcb_error"),
  error_type: z127.literal("pcb_footprint_overlap_error").default("pcb_footprint_overlap_error"),
  pcb_smtpad_ids: z127.array(z127.string()).optional(),
  pcb_plated_hole_ids: z127.array(z127.string()).optional(),
  pcb_hole_ids: z127.array(z127.string()).optional(),
  pcb_keepout_ids: z127.array(z127.string()).optional()
}).describe("Error emitted when a pcb footprint overlaps with another element");
expectTypesMatch(
  true
);

// src/pcb/pcb_courtyard_overlap_error.ts
import { z as z128 } from "zod";
var pcb_courtyard_overlap_error = base_circuit_json_error.extend({
  type: z128.literal("pcb_courtyard_overlap_error"),
  pcb_error_id: getZodPrefixedIdWithDefault("pcb_error"),
  error_type: z128.literal("pcb_courtyard_overlap_error").default("pcb_courtyard_overlap_error"),
  pcb_component_ids: z128.tuple([z128.string(), z128.string()])
}).describe(
  "Error emitted when the courtyard (CrtYd) of one PCB component overlaps with the courtyard of another"
);
expectTypesMatch(
  true
);

// src/pcb/pcb_keepout.ts
import { z as z129 } from "zod";
var pcb_keepout = z129.object({
  type: z129.literal("pcb_keepout"),
  shape: z129.literal("rect"),
  pcb_group_id: z129.string().optional(),
  subcircuit_id: z129.string().optional(),
  center: point,
  width: distance,
  height: distance,
  pcb_keepout_id: z129.string(),
  layers: z129.array(z129.string()),
  // Specify layers where the keepout applies
  description: z129.string().optional()
  // Optional description of the keepout
}).or(
  z129.object({
    type: z129.literal("pcb_keepout"),
    shape: z129.literal("circle"),
    pcb_group_id: z129.string().optional(),
    subcircuit_id: z129.string().optional(),
    center: point,
    radius: distance,
    pcb_keepout_id: z129.string(),
    layers: z129.array(z129.string()),
    // Specify layers where the keepout applies
    description: z129.string().optional()
    // Optional description of the keepout
  })
);
expectTypesMatch(true);

// src/pcb/pcb_cutout.ts
import { z as z130 } from "zod";
var pcb_cutout_base = z130.object({
  type: z130.literal("pcb_cutout"),
  pcb_cutout_id: getZodPrefixedIdWithDefault("pcb_cutout"),
  pcb_group_id: z130.string().optional(),
  subcircuit_id: z130.string().optional(),
  pcb_board_id: z130.string().optional(),
  pcb_panel_id: z130.string().optional()
});
var pcb_cutout_rect = pcb_cutout_base.extend({
  shape: z130.literal("rect"),
  center: point,
  width: length,
  height: length,
  rotation: rotation.optional(),
  corner_radius: length.optional()
});
expectTypesMatch(true);
var pcb_cutout_circle = pcb_cutout_base.extend({
  shape: z130.literal("circle"),
  center: point,
  radius: length
});
expectTypesMatch(true);
var pcb_cutout_polygon = pcb_cutout_base.extend({
  shape: z130.literal("polygon"),
  points: z130.array(point)
});
expectTypesMatch(true);
var pcb_cutout_path = pcb_cutout_base.extend({
  shape: z130.literal("path"),
  route: z130.array(point),
  slot_width: length,
  slot_length: length.optional(),
  space_between_slots: length.optional(),
  slot_corner_radius: length.optional()
});
expectTypesMatch(true);
var pcb_cutout = z130.discriminatedUnion("shape", [
  pcb_cutout_rect,
  pcb_cutout_circle,
  pcb_cutout_polygon,
  pcb_cutout_path
]).describe("Defines a cutout on the PCB, removing board material.");
expectTypesMatch(true);

// src/pcb/pcb_missing_footprint_error.ts
import { z as z131 } from "zod";
var pcb_missing_footprint_error = base_circuit_json_error.extend({
  type: z131.literal("pcb_missing_footprint_error"),
  pcb_missing_footprint_error_id: getZodPrefixedIdWithDefault(
    "pcb_missing_footprint_error"
  ),
  pcb_group_id: z131.string().optional(),
  subcircuit_id: z131.string().optional(),
  error_type: z131.literal("pcb_missing_footprint_error").default("pcb_missing_footprint_error"),
  source_component_id: z131.string()
}).describe("Defines a missing footprint error on the PCB");
expectTypesMatch(
  true
);

// src/pcb/external_footprint_load_error.ts
import { z as z132 } from "zod";
var external_footprint_load_error = base_circuit_json_error.extend({
  type: z132.literal("external_footprint_load_error"),
  external_footprint_load_error_id: getZodPrefixedIdWithDefault(
    "external_footprint_load_error"
  ),
  pcb_component_id: z132.string(),
  source_component_id: z132.string(),
  pcb_group_id: z132.string().optional(),
  subcircuit_id: z132.string().optional(),
  footprinter_string: z132.string().optional(),
  error_type: z132.literal("external_footprint_load_error").default("external_footprint_load_error")
}).describe("Defines an error when an external footprint fails to load");
expectTypesMatch(true);

// src/pcb/circuit_json_footprint_load_error.ts
import { z as z133 } from "zod";
var circuit_json_footprint_load_error = base_circuit_json_error.extend({
  type: z133.literal("circuit_json_footprint_load_error"),
  circuit_json_footprint_load_error_id: getZodPrefixedIdWithDefault(
    "circuit_json_footprint_load_error"
  ),
  pcb_component_id: z133.string(),
  source_component_id: z133.string(),
  pcb_group_id: z133.string().optional(),
  subcircuit_id: z133.string().optional(),
  error_type: z133.literal("circuit_json_footprint_load_error").default("circuit_json_footprint_load_error"),
  circuit_json: z133.array(z133.any()).optional()
}).describe("Defines an error when a circuit JSON footprint fails to load");
expectTypesMatch(true);

// src/pcb/pcb_group.ts
import { z as z134 } from "zod";
var pcb_group = z134.object({
  type: z134.literal("pcb_group"),
  pcb_group_id: getZodPrefixedIdWithDefault("pcb_group"),
  source_group_id: z134.string(),
  is_subcircuit: z134.boolean().optional(),
  subcircuit_id: z134.string().optional(),
  width: length.optional(),
  height: length.optional(),
  center: point,
  display_offset_x: z134.string().optional().describe(
    "How to display the x offset for this group, usually corresponding with how the user specified it"
  ),
  display_offset_y: z134.string().optional().describe(
    "How to display the y offset for this group, usually corresponding with how the user specified it"
  ),
  outline: z134.array(point).optional(),
  anchor_position: point.optional(),
  anchor_alignment: ninePointAnchor.default("center"),
  position_mode: z134.enum(["packed", "relative_to_group_anchor", "none"]).optional(),
  positioned_relative_to_pcb_group_id: z134.string().optional(),
  positioned_relative_to_pcb_board_id: z134.string().optional(),
  pcb_component_ids: z134.array(z134.string()),
  child_layout_mode: z134.enum(["packed", "none"]).optional(),
  name: z134.string().optional(),
  description: z134.string().optional(),
  layout_mode: z134.string().optional(),
  autorouter_configuration: z134.object({
    trace_clearance: length
  }).optional(),
  autorouter_used_string: z134.string().optional()
}).describe("Defines a group of components on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_autorouting_error.ts
import { z as z135 } from "zod";
var pcb_autorouting_error = base_circuit_json_error.extend({
  type: z135.literal("pcb_autorouting_error"),
  pcb_error_id: getZodPrefixedIdWithDefault("pcb_autorouting_error"),
  error_type: z135.literal("pcb_autorouting_error").default("pcb_autorouting_error"),
  subcircuit_id: z135.string().optional()
}).describe("The autorouting has failed to route a portion of the board");
expectTypesMatch(true);

// src/pcb/pcb_manual_edit_conflict_warning.ts
import { z as z136 } from "zod";
var pcb_manual_edit_conflict_warning = z136.object({
  type: z136.literal("pcb_manual_edit_conflict_warning"),
  pcb_manual_edit_conflict_warning_id: getZodPrefixedIdWithDefault(
    "pcb_manual_edit_conflict_warning"
  ),
  warning_type: z136.literal("pcb_manual_edit_conflict_warning").default("pcb_manual_edit_conflict_warning"),
  message: z136.string(),
  pcb_component_id: z136.string(),
  pcb_group_id: z136.string().optional(),
  subcircuit_id: z136.string().optional(),
  source_component_id: z136.string()
}).describe(
  "Warning emitted when a component has both manual placement and explicit pcbX/pcbY coordinates"
);
expectTypesMatch(true);

// src/pcb/pcb_connector_not_in_accessible_orientation_warning.ts
import { z as z137 } from "zod";
var connectorOrientationDirection = z137.enum(["x-", "x+", "y+", "y-"]);
var pcb_connector_not_in_accessible_orientation_warning = z137.object({
  type: z137.literal("pcb_connector_not_in_accessible_orientation_warning"),
  pcb_connector_not_in_accessible_orientation_warning_id: getZodPrefixedIdWithDefault(
    "pcb_connector_not_in_accessible_orientation_warning"
  ),
  warning_type: z137.literal("pcb_connector_not_in_accessible_orientation_warning").default("pcb_connector_not_in_accessible_orientation_warning"),
  message: z137.string(),
  pcb_component_id: z137.string(),
  source_component_id: z137.string().optional(),
  pcb_board_id: z137.string().optional(),
  facing_direction: connectorOrientationDirection,
  recommended_facing_direction: connectorOrientationDirection,
  subcircuit_id: z137.string().optional()
}).describe(
  "Warning emitted when a connector PCB component is facing inward toward the board and should be reoriented to an outward-facing direction"
);
expectTypesMatch(true);

// src/pcb/pcb_breakout_point.ts
import { z as z138 } from "zod";
var pcb_breakout_point = z138.object({
  type: z138.literal("pcb_breakout_point"),
  pcb_breakout_point_id: getZodPrefixedIdWithDefault("pcb_breakout_point"),
  pcb_group_id: z138.string(),
  subcircuit_id: z138.string().optional(),
  source_trace_id: z138.string().optional(),
  source_port_id: z138.string().optional(),
  source_net_id: z138.string().optional(),
  x: distance,
  y: distance
}).describe(
  "Defines a routing target within a pcb_group for a source_trace or source_net"
);
expectTypesMatch(true);

// src/pcb/pcb_ground_plane.ts
import { z as z139 } from "zod";
var pcb_ground_plane = z139.object({
  type: z139.literal("pcb_ground_plane"),
  pcb_ground_plane_id: getZodPrefixedIdWithDefault("pcb_ground_plane"),
  source_pcb_ground_plane_id: z139.string(),
  source_net_id: z139.string(),
  pcb_group_id: z139.string().optional(),
  subcircuit_id: z139.string().optional()
}).describe("Defines a ground plane on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_ground_plane_region.ts
import { z as z140 } from "zod";
var pcb_ground_plane_region = z140.object({
  type: z140.literal("pcb_ground_plane_region"),
  pcb_ground_plane_region_id: getZodPrefixedIdWithDefault(
    "pcb_ground_plane_region"
  ),
  pcb_ground_plane_id: z140.string(),
  pcb_group_id: z140.string().optional(),
  subcircuit_id: z140.string().optional(),
  layer: layer_ref,
  points: z140.array(point)
}).describe("Defines a polygon region of a ground plane");
expectTypesMatch(true);

// src/pcb/pcb_thermal_spoke.ts
import { z as z141 } from "zod";
var pcb_thermal_spoke = z141.object({
  type: z141.literal("pcb_thermal_spoke"),
  pcb_thermal_spoke_id: getZodPrefixedIdWithDefault("pcb_thermal_spoke"),
  pcb_ground_plane_id: z141.string(),
  shape: z141.string(),
  spoke_count: z141.number(),
  spoke_thickness: distance,
  spoke_inner_diameter: distance,
  spoke_outer_diameter: distance,
  pcb_plated_hole_id: z141.string().optional(),
  subcircuit_id: z141.string().optional()
}).describe("Pattern for connecting a ground plane to a plated hole");
expectTypesMatch(true);

// src/pcb/pcb_copper_pour.ts
import { z as z142 } from "zod";
var pcb_copper_pour_base = z142.object({
  type: z142.literal("pcb_copper_pour"),
  pcb_copper_pour_id: getZodPrefixedIdWithDefault("pcb_copper_pour"),
  pcb_group_id: z142.string().optional(),
  subcircuit_id: z142.string().optional(),
  layer: layer_ref,
  source_net_id: z142.string().optional(),
  covered_with_solder_mask: z142.boolean().optional().default(true)
});
var pcb_copper_pour_rect = pcb_copper_pour_base.extend({
  shape: z142.literal("rect"),
  center: point,
  width: length,
  height: length,
  rotation: rotation.optional()
});
expectTypesMatch(true);
var pcb_copper_pour_brep = pcb_copper_pour_base.extend({
  shape: z142.literal("brep"),
  brep_shape
});
expectTypesMatch(true);
var pcb_copper_pour_polygon = pcb_copper_pour_base.extend({
  shape: z142.literal("polygon"),
  points: z142.array(point)
});
expectTypesMatch(true);
var pcb_copper_pour = z142.discriminatedUnion("shape", [
  pcb_copper_pour_rect,
  pcb_copper_pour_brep,
  pcb_copper_pour_polygon
]).describe("Defines a copper pour on the PCB.");
expectTypesMatch(true);

// src/pcb/pcb_component_outside_board_error.ts
import { z as z143 } from "zod";
var pcb_component_outside_board_error = base_circuit_json_error.extend({
  type: z143.literal("pcb_component_outside_board_error"),
  pcb_component_outside_board_error_id: getZodPrefixedIdWithDefault(
    "pcb_component_outside_board_error"
  ),
  error_type: z143.literal("pcb_component_outside_board_error").default("pcb_component_outside_board_error"),
  pcb_component_id: z143.string(),
  pcb_board_id: z143.string(),
  component_center: point,
  component_bounds: z143.object({
    min_x: z143.number(),
    max_x: z143.number(),
    min_y: z143.number(),
    max_y: z143.number()
  }),
  subcircuit_id: z143.string().optional(),
  source_component_id: z143.string().optional()
}).describe(
  "Error emitted when a PCB component is placed outside the board boundaries"
);
expectTypesMatch(true);

// src/pcb/pcb_component_not_on_board_edge_error.ts
import { z as z144 } from "zod";
var pcb_component_not_on_board_edge_error = base_circuit_json_error.extend({
  type: z144.literal("pcb_component_not_on_board_edge_error"),
  pcb_component_not_on_board_edge_error_id: getZodPrefixedIdWithDefault(
    "pcb_component_not_on_board_edge_error"
  ),
  error_type: z144.literal("pcb_component_not_on_board_edge_error").default("pcb_component_not_on_board_edge_error"),
  pcb_component_id: z144.string(),
  pcb_board_id: z144.string(),
  component_center: point,
  pad_to_nearest_board_edge_distance: z144.number(),
  source_component_id: z144.string().optional(),
  subcircuit_id: z144.string().optional()
}).describe(
  "Error emitted when a component that must be placed on the board edge is centered away from the edge"
);
expectTypesMatch(true);

// src/pcb/pcb_component_invalid_layer_error.ts
import { z as z145 } from "zod";
var pcb_component_invalid_layer_error = base_circuit_json_error.extend({
  type: z145.literal("pcb_component_invalid_layer_error"),
  pcb_component_invalid_layer_error_id: getZodPrefixedIdWithDefault(
    "pcb_component_invalid_layer_error"
  ),
  error_type: z145.literal("pcb_component_invalid_layer_error").default("pcb_component_invalid_layer_error"),
  pcb_component_id: z145.string().optional(),
  source_component_id: z145.string(),
  layer: layer_ref,
  subcircuit_id: z145.string().optional()
}).describe(
  "Error emitted when a component is placed on an invalid layer (components can only be on 'top' or 'bottom' layers)"
);
expectTypesMatch(true);

// src/pcb/pcb_via_clearance_error.ts
import { z as z146 } from "zod";
var pcb_via_clearance_error = base_circuit_json_error.extend({
  type: z146.literal("pcb_via_clearance_error"),
  pcb_error_id: getZodPrefixedIdWithDefault("pcb_error"),
  error_type: z146.literal("pcb_via_clearance_error").default("pcb_via_clearance_error"),
  pcb_via_ids: z146.array(z146.string()).min(2),
  minimum_clearance: distance.optional(),
  actual_clearance: distance.optional(),
  pcb_center: z146.object({
    x: z146.number().optional(),
    y: z146.number().optional()
  }).optional(),
  subcircuit_id: z146.string().optional()
}).describe("Error emitted when vias are closer than the allowed clearance");
expectTypesMatch(true);

// src/pcb/pcb_courtyard_rect.ts
import { z as z147 } from "zod";
var pcb_courtyard_rect = z147.object({
  type: z147.literal("pcb_courtyard_rect"),
  pcb_courtyard_rect_id: getZodPrefixedIdWithDefault("pcb_courtyard_rect"),
  pcb_component_id: z147.string(),
  pcb_group_id: z147.string().optional(),
  subcircuit_id: z147.string().optional(),
  center: point,
  width: length,
  height: length,
  layer: visible_layer,
  color: z147.string().optional()
}).describe("Defines a courtyard rectangle on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_courtyard_outline.ts
import { z as z148 } from "zod";
var pcb_courtyard_outline = z148.object({
  type: z148.literal("pcb_courtyard_outline"),
  pcb_courtyard_outline_id: getZodPrefixedIdWithDefault(
    "pcb_courtyard_outline"
  ),
  pcb_component_id: z148.string(),
  pcb_group_id: z148.string().optional(),
  subcircuit_id: z148.string().optional(),
  layer: visible_layer,
  outline: z148.array(point).min(2)
}).describe("Defines a courtyard outline on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_courtyard_polygon.ts
import { z as z149 } from "zod";
var pcb_courtyard_polygon = z149.object({
  type: z149.literal("pcb_courtyard_polygon"),
  pcb_courtyard_polygon_id: getZodPrefixedIdWithDefault(
    "pcb_courtyard_polygon"
  ),
  pcb_component_id: z149.string(),
  pcb_group_id: z149.string().optional(),
  subcircuit_id: z149.string().optional(),
  layer: visible_layer,
  points: z149.array(point).min(3),
  color: z149.string().optional()
}).describe("Defines a courtyard polygon on the PCB");
expectTypesMatch(true);

// src/pcb/pcb_courtyard_circle.ts
import { z as z150 } from "zod";
var pcb_courtyard_circle = z150.object({
  type: z150.literal("pcb_courtyard_circle"),
  pcb_courtyard_circle_id: getZodPrefixedIdWithDefault(
    "pcb_courtyard_circle"
  ),
  pcb_component_id: z150.string(),
  pcb_group_id: z150.string().optional(),
  subcircuit_id: z150.string().optional(),
  center: point,
  radius: length,
  layer: visible_layer,
  color: z150.string().optional()
}).describe("Defines a courtyard circle on the PCB");
expectTypesMatch(true);

// src/cad/cad_component.ts
import { z as z151 } from "zod";

// src/cad/cad_model_conventions.ts
var cad_model_formats = [
  "obj",
  "stl",
  "3mf",
  "gltf",
  "glb",
  "step",
  "wrl"
];
var cad_model_axis_directions = [
  "x+",
  "x-",
  "y+",
  "y-",
  "z+",
  "z-"
];
var cadModelDefaultDirectionMap = {
  obj: "z+",
  stl: "z+",
  "3mf": "z+",
  gltf: "y+",
  glb: "y+",
  step: "z+",
  wrl: "y+"
};

// src/cad/cad_component.ts
var cad_component = z151.object({
  type: z151.literal("cad_component"),
  cad_component_id: z151.string(),
  pcb_component_id: z151.string(),
  source_component_id: z151.string(),
  position: point3,
  rotation: point3.optional(),
  size: point3.optional(),
  layer: layer_ref.optional(),
  subcircuit_id: z151.string().optional(),
  // These are all ways to generate/load the 3d model
  footprinter_string: z151.string().optional(),
  model_obj_url: z151.string().optional(),
  model_stl_url: z151.string().optional(),
  model_3mf_url: z151.string().optional(),
  model_gltf_url: z151.string().optional(),
  model_glb_url: z151.string().optional(),
  model_step_url: z151.string().optional(),
  model_wrl_url: z151.string().optional(),
  model_asset: asset.optional(),
  model_unit_to_mm_scale_factor: z151.number().optional(),
  model_board_normal_direction: z151.enum(cad_model_axis_directions).optional().describe(
    `The direction in the model's coordinate space that is considered "up" or "coming out of the board surface"`
  ),
  model_origin_position: point3.optional(),
  model_origin_alignment: z151.enum([
    "unknown",
    "center",
    "center_of_component_on_board_surface",
    "bottom_center_of_component"
  ]).optional(),
  model_object_fit: z151.enum(["contain_within_bounds", "fill_bounds"]).optional().default("contain_within_bounds"),
  model_jscad: z151.any().optional(),
  show_as_translucent_model: z151.boolean().optional(),
  anchor_alignment: z151.enum(["center", "xy_center_z_board"]).optional().default("center")
}).describe("Defines a component on the PCB");
expectTypesMatch(true);

// src/simulation/simulation_voltage_source.ts
import { z as z152 } from "zod";
var wave_shape = z152.enum(["sinewave", "square", "triangle", "sawtooth"]);
var percentage = z152.union([z152.string(), z152.number()]).transform((val) => {
  if (typeof val === "string") {
    if (val.endsWith("%")) {
      return parseFloat(val.slice(0, -1)) / 100;
    }
    return parseFloat(val);
  }
  return val;
}).pipe(
  z152.number().min(0, "Duty cycle must be non-negative").max(1, "Duty cycle cannot be greater than 100%")
);
var simulation_dc_voltage_source = z152.object({
  type: z152.literal("simulation_voltage_source"),
  simulation_voltage_source_id: getZodPrefixedIdWithDefault(
    "simulation_voltage_source"
  ),
  is_dc_source: z152.literal(true).optional().default(true),
  positive_source_port_id: z152.string().optional(),
  negative_source_port_id: z152.string().optional(),
  positive_source_net_id: z152.string().optional(),
  negative_source_net_id: z152.string().optional(),
  voltage
}).describe("Defines a DC voltage source for simulation");
var simulation_ac_voltage_source = z152.object({
  type: z152.literal("simulation_voltage_source"),
  simulation_voltage_source_id: getZodPrefixedIdWithDefault(
    "simulation_voltage_source"
  ),
  is_dc_source: z152.literal(false),
  terminal1_source_port_id: z152.string().optional(),
  terminal2_source_port_id: z152.string().optional(),
  terminal1_source_net_id: z152.string().optional(),
  terminal2_source_net_id: z152.string().optional(),
  voltage: voltage.optional(),
  frequency: frequency.optional(),
  peak_to_peak_voltage: voltage.optional(),
  wave_shape: wave_shape.optional(),
  phase: rotation.optional(),
  duty_cycle: percentage.optional()
}).describe("Defines an AC voltage source for simulation");
var simulation_voltage_source = z152.union([simulation_dc_voltage_source, simulation_ac_voltage_source]).describe("Defines a voltage source for simulation");
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);

// src/simulation/simulation_current_source.ts
import { z as z153 } from "zod";
var percentage2 = z153.union([z153.string(), z153.number()]).transform((val) => {
  if (typeof val === "string") {
    if (val.endsWith("%")) {
      return parseFloat(val.slice(0, -1)) / 100;
    }
    return parseFloat(val);
  }
  return val;
}).pipe(
  z153.number().min(0, "Duty cycle must be non-negative").max(1, "Duty cycle cannot be greater than 100%")
);
var simulation_dc_current_source = z153.object({
  type: z153.literal("simulation_current_source"),
  simulation_current_source_id: getZodPrefixedIdWithDefault(
    "simulation_current_source"
  ),
  is_dc_source: z153.literal(true).optional().default(true),
  positive_source_port_id: z153.string().optional(),
  negative_source_port_id: z153.string().optional(),
  positive_source_net_id: z153.string().optional(),
  negative_source_net_id: z153.string().optional(),
  current
}).describe("Defines a DC current source for simulation");
var simulation_ac_current_source = z153.object({
  type: z153.literal("simulation_current_source"),
  simulation_current_source_id: getZodPrefixedIdWithDefault(
    "simulation_current_source"
  ),
  is_dc_source: z153.literal(false),
  terminal1_source_port_id: z153.string().optional(),
  terminal2_source_port_id: z153.string().optional(),
  terminal1_source_net_id: z153.string().optional(),
  terminal2_source_net_id: z153.string().optional(),
  current: current.optional(),
  frequency: frequency.optional(),
  peak_to_peak_current: current.optional(),
  wave_shape: wave_shape.optional(),
  phase: rotation.optional(),
  duty_cycle: percentage2.optional()
}).describe("Defines an AC current source for simulation");
var simulation_current_source = z153.union([simulation_dc_current_source, simulation_ac_current_source]).describe("Defines a current source for simulation");
expectTypesMatch(true);
expectTypesMatch(true);
expectTypesMatch(true);

// src/simulation/simulation_experiment.ts
import { z as z154 } from "zod";
var experiment_type = z154.union([
  z154.literal("spice_dc_sweep"),
  z154.literal("spice_dc_operating_point"),
  z154.literal("spice_transient_analysis"),
  z154.literal("spice_ac_analysis")
]);
var simulation_experiment = z154.object({
  type: z154.literal("simulation_experiment"),
  simulation_experiment_id: getZodPrefixedIdWithDefault(
    "simulation_experiment"
  ),
  name: z154.string(),
  experiment_type,
  time_per_step: duration_ms.optional(),
  start_time_ms: ms.optional(),
  end_time_ms: ms.optional()
}).describe("Defines a simulation experiment configuration");
expectTypesMatch(true);

// src/simulation/simulation_transient_voltage_graph.ts
import { z as z155 } from "zod";
var simulation_transient_voltage_graph = z155.object({
  type: z155.literal("simulation_transient_voltage_graph"),
  simulation_transient_voltage_graph_id: getZodPrefixedIdWithDefault(
    "simulation_transient_voltage_graph"
  ),
  simulation_experiment_id: z155.string(),
  timestamps_ms: z155.array(z155.number()).optional(),
  voltage_levels: z155.array(z155.number()),
  source_component_id: z155.string().optional(),
  subcircuit_connectivity_map_key: z155.string().optional(),
  time_per_step: duration_ms,
  start_time_ms: ms,
  end_time_ms: ms,
  name: z155.string().optional(),
  color: z155.string().optional()
}).describe("Stores voltage measurements over time for a simulation");
expectTypesMatch(true);

// src/simulation/simulation_switch.ts
import { z as z156 } from "zod";
var simulation_switch = z156.object({
  type: z156.literal("simulation_switch"),
  simulation_switch_id: getZodPrefixedIdWithDefault("simulation_switch"),
  source_component_id: z156.string().optional(),
  closes_at: ms.optional(),
  opens_at: ms.optional(),
  starts_closed: z156.boolean().optional(),
  switching_frequency: frequency.optional()
}).describe("Defines a switch for simulation timing control");
expectTypesMatch(true);

// src/simulation/simulation_voltage_probe.ts
import { z as z157 } from "zod";
var simulation_voltage_probe = z157.object({
  type: z157.literal("simulation_voltage_probe"),
  simulation_voltage_probe_id: getZodPrefixedIdWithDefault(
    "simulation_voltage_probe"
  ),
  source_component_id: z157.string().optional(),
  name: z157.string().optional(),
  signal_input_source_port_id: z157.string().optional(),
  signal_input_source_net_id: z157.string().optional(),
  reference_input_source_port_id: z157.string().optional(),
  reference_input_source_net_id: z157.string().optional(),
  subcircuit_id: z157.string().optional(),
  color: z157.string().optional()
}).describe(
  "Defines a voltage probe for simulation. If a reference input is not provided, it measures against ground. If a reference input is provided, it measures the differential voltage between two points."
).superRefine((data, ctx) => {
  const is_differential = data.reference_input_source_port_id || data.reference_input_source_net_id;
  if (is_differential) {
    const has_ports = !!data.signal_input_source_port_id || !!data.reference_input_source_port_id;
    const has_nets = !!data.signal_input_source_net_id || !!data.reference_input_source_net_id;
    if (has_ports && has_nets) {
      ctx.addIssue({
        code: z157.ZodIssueCode.custom,
        message: "Cannot mix port and net connections in a differential probe."
      });
    } else if (has_ports) {
      if (!data.signal_input_source_port_id || !data.reference_input_source_port_id) {
        ctx.addIssue({
          code: z157.ZodIssueCode.custom,
          message: "Differential port probe requires both signal_input_source_port_id and reference_input_source_port_id."
        });
      }
    } else if (has_nets) {
      if (!data.signal_input_source_net_id || !data.reference_input_source_net_id) {
        ctx.addIssue({
          code: z157.ZodIssueCode.custom,
          message: "Differential net probe requires both signal_input_source_net_id and reference_input_source_net_id."
        });
      }
    }
  } else {
    if (!!data.signal_input_source_port_id === !!data.signal_input_source_net_id) {
      ctx.addIssue({
        code: z157.ZodIssueCode.custom,
        message: "A voltage probe must have exactly one of signal_input_source_port_id or signal_input_source_net_id."
      });
    }
  }
});
expectTypesMatch(true);

// src/simulation/simulation_unknown_experiment_error.ts
import { z as z158 } from "zod";
var simulation_unknown_experiment_error = base_circuit_json_error.extend({
  type: z158.literal("simulation_unknown_experiment_error"),
  simulation_unknown_experiment_error_id: getZodPrefixedIdWithDefault(
    "simulation_unknown_experiment_error"
  ),
  error_type: z158.literal("simulation_unknown_experiment_error").default("simulation_unknown_experiment_error"),
  simulation_experiment_id: z158.string().optional(),
  subcircuit_id: z158.string().optional()
}).describe("An unknown error occurred during the simulation experiment.");
expectTypesMatch(true);

// src/simulation/simulation_op_amp.ts
import { z as z159 } from "zod";
var simulation_op_amp = z159.object({
  type: z159.literal("simulation_op_amp"),
  simulation_op_amp_id: getZodPrefixedIdWithDefault("simulation_op_amp"),
  source_component_id: z159.string().optional(),
  inverting_input_source_port_id: z159.string(),
  non_inverting_input_source_port_id: z159.string(),
  output_source_port_id: z159.string(),
  positive_supply_source_port_id: z159.string(),
  negative_supply_source_port_id: z159.string()
}).describe("Defines a simple ideal operational amplifier for simulation");
expectTypesMatch(true);

// src/any_circuit_element.ts
import { z as z160 } from "zod";
var any_circuit_element = z160.union([
  source_trace,
  source_port,
  source_component_internal_connection,
  any_source_component,
  source_net,
  source_group,
  source_simple_chip,
  source_simple_capacitor,
  source_simple_diode,
  source_simple_led,
  source_simple_resistor,
  source_simple_power_source,
  source_simple_battery,
  source_simple_inductor,
  source_simple_pin_header,
  source_simple_pinout,
  source_simple_resonator,
  source_simple_switch,
  source_simple_transistor,
  source_simple_test_point,
  source_simple_mosfet,
  source_simple_op_amp,
  source_simple_potentiometer,
  source_simple_push_button,
  source_pcb_ground_plane,
  source_manually_placed_via,
  source_board,
  source_project_metadata,
  source_invalid_component_property_error,
  source_trace_not_connected_error,
  source_pin_missing_trace_warning,
  source_no_power_pin_defined_warning,
  source_no_ground_pin_defined_warning,
  source_component_pins_underspecified_warning,
  source_pin_must_be_connected_error,
  unknown_error_finding_part,
  source_i2c_misconfigured_error,
  source_ambiguous_port_reference,
  pcb_component,
  pcb_hole,
  pcb_missing_footprint_error,
  external_footprint_load_error,
  circuit_json_footprint_load_error,
  pcb_manual_edit_conflict_warning,
  pcb_connector_not_in_accessible_orientation_warning,
  pcb_plated_hole,
  pcb_keepout,
  pcb_port,
  pcb_net,
  pcb_text,
  pcb_trace,
  pcb_via,
  pcb_smtpad,
  pcb_solder_paste,
  pcb_board,
  pcb_panel,
  pcb_group,
  pcb_trace_hint,
  pcb_silkscreen_line,
  pcb_silkscreen_path,
  pcb_silkscreen_text,
  pcb_silkscreen_pill,
  pcb_copper_text,
  pcb_silkscreen_rect,
  pcb_silkscreen_circle,
  pcb_silkscreen_oval,
  pcb_trace_error,
  pcb_trace_missing_error,
  pcb_placement_error,
  pcb_panelization_placement_error,
  pcb_port_not_matched_error,
  pcb_port_not_connected_error,
  pcb_via_clearance_error,
  pcb_fabrication_note_path,
  pcb_fabrication_note_text,
  pcb_fabrication_note_rect,
  pcb_fabrication_note_dimension,
  pcb_note_text,
  pcb_note_rect,
  pcb_note_path,
  pcb_note_line,
  pcb_note_dimension,
  pcb_autorouting_error,
  pcb_footprint_overlap_error,
  pcb_courtyard_overlap_error,
  pcb_breakout_point,
  pcb_cutout,
  pcb_ground_plane,
  pcb_ground_plane_region,
  pcb_thermal_spoke,
  pcb_copper_pour,
  pcb_component_outside_board_error,
  pcb_component_not_on_board_edge_error,
  pcb_component_invalid_layer_error,
  pcb_courtyard_rect,
  pcb_courtyard_outline,
  pcb_courtyard_polygon,
  pcb_courtyard_circle,
  schematic_box,
  schematic_text,
  schematic_line,
  schematic_rect,
  schematic_circle,
  schematic_arc,
  schematic_component,
  schematic_symbol,
  schematic_port,
  schematic_trace,
  schematic_path,
  schematic_error,
  schematic_layout_error,
  schematic_net_label,
  schematic_debug_object,
  schematic_voltage_probe,
  schematic_manual_edit_conflict_warning,
  schematic_group,
  schematic_sheet,
  schematic_table,
  schematic_table_cell,
  cad_component,
  simulation_voltage_source,
  simulation_current_source,
  simulation_experiment,
  simulation_transient_voltage_graph,
  simulation_switch,
  simulation_voltage_probe,
  simulation_unknown_experiment_error,
  simulation_op_amp
]);
var any_soup_element = any_circuit_element;
expectTypesMatch(true);
expectStringUnionsMatch(true);
export {
  all_layers,
  any_circuit_element,
  any_soup_element,
  any_source_component,
  asset,
  base_circuit_json_error,
  battery_capacity,
  brep_shape,
  cadModelDefaultDirectionMap,
  cad_component,
  cad_model_axis_directions,
  cad_model_formats,
  capacitance,
  circuit_json_footprint_load_error,
  current,
  distance,
  duration_ms,
  experiment_type,
  external_footprint_load_error,
  frequency,
  getZodPrefixedIdWithDefault,
  inductance,
  kicadAt,
  kicadEffects,
  kicadFont,
  kicadFootprintAttributes,
  kicadFootprintMetadata,
  kicadFootprintModel,
  kicadFootprintPad,
  kicadFootprintProperties,
  kicadProperty,
  kicadSymbolEffects,
  kicadSymbolMetadata,
  kicadSymbolPinNames,
  kicadSymbolPinNumbers,
  kicadSymbolProperties,
  kicadSymbolProperty,
  layer_ref,
  layer_string,
  length,
  ms,
  ninePointAnchor,
  pcbRenderLayer,
  pcb_autorouting_error,
  pcb_board,
  pcb_breakout_point,
  pcb_component,
  pcb_component_invalid_layer_error,
  pcb_component_not_on_board_edge_error,
  pcb_component_outside_board_error,
  pcb_connector_not_in_accessible_orientation_warning,
  pcb_copper_pour,
  pcb_copper_pour_brep,
  pcb_copper_pour_polygon,
  pcb_copper_pour_rect,
  pcb_copper_text,
  pcb_courtyard_circle,
  pcb_courtyard_outline,
  pcb_courtyard_overlap_error,
  pcb_courtyard_polygon,
  pcb_courtyard_rect,
  pcb_cutout,
  pcb_cutout_circle,
  pcb_cutout_path,
  pcb_cutout_polygon,
  pcb_cutout_rect,
  pcb_fabrication_note_dimension,
  pcb_fabrication_note_path,
  pcb_fabrication_note_rect,
  pcb_fabrication_note_text,
  pcb_footprint_overlap_error,
  pcb_ground_plane,
  pcb_ground_plane_region,
  pcb_group,
  pcb_hole,
  pcb_hole_circle_or_square_shape,
  pcb_hole_circle_shape,
  pcb_hole_oval_shape,
  pcb_hole_pill_shape,
  pcb_hole_rect_shape,
  pcb_hole_rotated_pill_shape,
  pcb_keepout,
  pcb_manual_edit_conflict_warning,
  pcb_missing_footprint_error,
  pcb_net,
  pcb_note_dimension,
  pcb_note_line,
  pcb_note_path,
  pcb_note_rect,
  pcb_note_text,
  pcb_panel,
  pcb_panelization_placement_error,
  pcb_placement_error,
  pcb_plated_hole,
  pcb_port,
  pcb_port_not_connected_error,
  pcb_port_not_matched_error,
  pcb_route_hint,
  pcb_route_hints,
  pcb_silkscreen_circle,
  pcb_silkscreen_line,
  pcb_silkscreen_oval,
  pcb_silkscreen_path,
  pcb_silkscreen_pill,
  pcb_silkscreen_rect,
  pcb_silkscreen_text,
  pcb_smtpad,
  pcb_smtpad_pill,
  pcb_solder_paste,
  pcb_text,
  pcb_thermal_spoke,
  pcb_trace,
  pcb_trace_error,
  pcb_trace_hint,
  pcb_trace_missing_error,
  pcb_trace_route_point,
  pcb_trace_route_point_via,
  pcb_trace_route_point_wire,
  pcb_via,
  pcb_via_clearance_error,
  point,
  point3,
  point_with_bulge,
  port_arrangement,
  position,
  position3,
  resistance,
  ring,
  rotation,
  route_hint_point,
  schematic_arc,
  schematic_box,
  schematic_circle,
  schematic_component,
  schematic_component_port_arrangement_by_sides,
  schematic_component_port_arrangement_by_size,
  schematic_debug_line,
  schematic_debug_object,
  schematic_debug_object_base,
  schematic_debug_point,
  schematic_debug_rect,
  schematic_error,
  schematic_group,
  schematic_layout_error,
  schematic_line,
  schematic_manual_edit_conflict_warning,
  schematic_net_label,
  schematic_path,
  schematic_pin_styles,
  schematic_port,
  schematic_rect,
  schematic_sheet,
  schematic_symbol,
  schematic_table,
  schematic_table_cell,
  schematic_text,
  schematic_trace,
  schematic_voltage_probe,
  simulation_ac_current_source,
  simulation_ac_voltage_source,
  simulation_current_source,
  simulation_dc_current_source,
  simulation_dc_voltage_source,
  simulation_experiment,
  simulation_op_amp,
  simulation_switch,
  simulation_transient_voltage_graph,
  simulation_unknown_experiment_error,
  simulation_voltage_probe,
  simulation_voltage_source,
  size,
  source_ambiguous_port_reference,
  source_board,
  source_component_base,
  source_component_internal_connection,
  source_component_pins_underspecified_warning,
  source_failed_to_create_component_error,
  source_group,
  source_i2c_misconfigured_error,
  source_interconnect,
  source_invalid_component_property_error,
  source_manually_placed_via,
  source_missing_property_error,
  source_net,
  source_no_ground_pin_defined_warning,
  source_no_power_pin_defined_warning,
  source_pcb_ground_plane,
  source_pin_attributes,
  source_pin_missing_trace_warning,
  source_pin_must_be_connected_error,
  source_port,
  source_project_metadata,
  source_property_ignored_warning,
  source_simple_battery,
  source_simple_capacitor,
  source_simple_chip,
  source_simple_connector,
  source_simple_crystal,
  source_simple_current_source,
  source_simple_diode,
  source_simple_fiducial,
  source_simple_ground,
  source_simple_inductor,
  source_simple_led,
  source_simple_mosfet,
  source_simple_op_amp,
  source_simple_pin_header,
  source_simple_pinout,
  source_simple_potentiometer,
  source_simple_power_source,
  source_simple_push_button,
  source_simple_resistor,
  source_simple_resonator,
  source_simple_switch,
  source_simple_test_point,
  source_simple_transistor,
  source_simple_voltage_probe,
  source_simple_voltage_source,
  source_trace,
  source_trace_not_connected_error,
  supplier_name,
  time,
  timestamp,
  unknown_error_finding_part,
  visible_layer,
  voltage,
  wave_shape
};
