// circuit-to-canvas/dist/index.js
// lib/drawer/CircuitToCanvasDrawer.ts
import {
  compose as compose3,
  identity,
  scale,
  translate as translate3
} from "transformation-matrix";

// lib/drawer/shapes/path.ts
import { applyToPoint } from "transformation-matrix";
function drawPath(params) {
  const {
    ctx,
    points,
    fill,
    stroke,
    strokeWidth = 1,
    minStrokePx,
    realToCanvasMat,
    closePath = false
  } = params;
  if (points.length < 2) return;
  ctx.beginPath();
  const canvasPoints = points.map(
    (p) => applyToPoint(realToCanvasMat, [p.x, p.y])
  );
  const firstPoint = canvasPoints[0];
  if (!firstPoint) return;
  const [firstX, firstY] = firstPoint;
  ctx.moveTo(firstX, firstY);
  for (let i = 1; i < canvasPoints.length; i++) {
    const point = canvasPoints[i];
    if (!point) continue;
    const [x, y] = point;
    ctx.lineTo(x, y);
  }
  if (closePath) {
    ctx.closePath();
  }
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke) {
    const scaledStrokeWidth = strokeWidth * Math.abs(realToCanvasMat.a);
    const clampedStrokeWidth = minStrokePx === void 0 ? scaledStrokeWidth : Math.max(scaledStrokeWidth, minStrokePx);
    ctx.strokeStyle = stroke;
    ctx.lineWidth = clampedStrokeWidth;
    ctx.stroke();
  }
}

// lib/drawer/shapes/rect.ts
import { applyToPoint as applyToPoint2 } from "transformation-matrix";
function drawRect(params) {
  const {
    ctx,
    center,
    width,
    height,
    fill,
    realToCanvasMat,
    borderRadius = 0,
    ccwRotationDegrees = 0,
    stroke,
    strokeWidth,
    isStrokeDashed = false
  } = params;
  const [cx, cy] = applyToPoint2(realToCanvasMat, [center.x, center.y]);
  const scaledWidth = width * Math.abs(realToCanvasMat.a);
  const scaledHeight = height * Math.abs(realToCanvasMat.a);
  const scaledRadius = borderRadius * Math.abs(realToCanvasMat.a);
  const scaledStrokeWidth = strokeWidth ? strokeWidth * Math.abs(realToCanvasMat.a) : void 0;
  ctx.save();
  ctx.translate(cx, cy);
  if (ccwRotationDegrees !== 0) {
    ctx.rotate(-ccwRotationDegrees * (Math.PI / 180));
  }
  ctx.beginPath();
  if (scaledRadius > 0) {
    const x = -scaledWidth / 2;
    const y = -scaledHeight / 2;
    const r = Math.min(scaledRadius, scaledWidth / 2, scaledHeight / 2);
    ctx.moveTo(x + r, y);
    ctx.lineTo(x + scaledWidth - r, y);
    ctx.arcTo(x + scaledWidth, y, x + scaledWidth, y + r, r);
    ctx.lineTo(x + scaledWidth, y + scaledHeight - r);
    ctx.arcTo(
      x + scaledWidth,
      y + scaledHeight,
      x + scaledWidth - r,
      y + scaledHeight,
      r
    );
    ctx.lineTo(x + r, y + scaledHeight);
    ctx.arcTo(x, y + scaledHeight, x, y + scaledHeight - r, r);
    ctx.lineTo(x, y + r);
    ctx.arcTo(x, y, x + r, y, r);
  } else {
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
  }
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke && scaledStrokeWidth) {
    if (isStrokeDashed) {
      ctx.setLineDash([scaledStrokeWidth * 3, scaledStrokeWidth * 2]);
    } else {
      ctx.setLineDash([]);
    }
    ctx.strokeStyle = stroke;
    ctx.lineWidth = scaledStrokeWidth;
    ctx.stroke();
  }
  ctx.restore();
}

// lib/drawer/elements/pcb-board.ts
function drawPcbBoard(params) {
  const {
    ctx,
    board,
    realToCanvasMat,
    colorMap,
    drawBoardMaterial,
    minBoardOutlineStrokePx
  } = params;
  const { width, height, center, outline } = board;
  if (outline && Array.isArray(outline) && outline.length >= 3) {
    if (drawBoardMaterial) {
      drawPath({
        ctx,
        points: outline.map((p) => ({ x: p.x, y: p.y })),
        fill: colorMap.substrate,
        realToCanvasMat,
        closePath: true
      });
    }
    drawPath({
      ctx,
      points: outline.map((p) => ({ x: p.x, y: p.y })),
      stroke: colorMap.boardOutline,
      strokeWidth: 0.1,
      realToCanvasMat,
      closePath: true,
      minStrokePx: minBoardOutlineStrokePx
    });
    return;
  }
  if (width !== void 0 && height !== void 0 && center) {
    if (drawBoardMaterial) {
      drawRect({
        ctx,
        center,
        width,
        height,
        fill: colorMap.substrate,
        realToCanvasMat
      });
    }
    const halfWidth = width / 2;
    const halfHeight = height / 2;
    const corners = [
      { x: center.x - halfWidth, y: center.y - halfHeight },
      { x: center.x + halfWidth, y: center.y - halfHeight },
      { x: center.x + halfWidth, y: center.y + halfHeight },
      { x: center.x - halfWidth, y: center.y + halfHeight }
    ];
    drawPath({
      ctx,
      points: corners,
      stroke: colorMap.boardOutline,
      strokeWidth: 0.1,
      realToCanvasMat,
      closePath: true,
      minStrokePx: minBoardOutlineStrokePx
    });
  }
}

// lib/drawer/elements/pcb-copper-pour.ts
import { applyToPoint as applyToPoint4 } from "transformation-matrix";

// lib/drawer/shapes/polygon.ts
import { applyToPoint as applyToPoint3 } from "transformation-matrix";
function drawPolygon(params) {
  const { ctx, points, fill, realToCanvasMat } = params;
  if (points.length < 3) return;
  ctx.beginPath();
  const canvasPoints = points.map(
    (p) => applyToPoint3(realToCanvasMat, [p.x, p.y])
  );
  const firstPoint = canvasPoints[0];
  if (!firstPoint) return;
  const [firstX, firstY] = firstPoint;
  ctx.moveTo(firstX, firstY);
  for (let i = 1; i < canvasPoints.length; i++) {
    const point = canvasPoints[i];
    if (!point) continue;
    const [x, y] = point;
    ctx.lineTo(x, y);
  }
  ctx.closePath();
  ctx.fillStyle = fill;
  ctx.fill();
}

// lib/drawer/elements/pcb-copper-pour.ts
function layerToColor(layer, colorMap) {
  return colorMap.copper[layer] ?? colorMap.copper.top;
}
function layerToPourOpacity(layer) {
  if (layer === "top" || layer === "bottom") return 0.5;
  return 1;
}
function computeArcFromBulge(startX, startY, endX, endY, bulge) {
  if (Math.abs(bulge) < 1e-10) {
    return null;
  }
  const chordX = endX - startX;
  const chordY = endY - startY;
  const chordLength = Math.hypot(chordX, chordY);
  if (chordLength < 1e-10) {
    return null;
  }
  const sagitta = Math.abs(bulge) * (chordLength / 2);
  const halfChord = chordLength / 2;
  const radius = (sagitta * sagitta + halfChord * halfChord) / (2 * sagitta);
  const distToCenter = radius - sagitta;
  const midX = (startX + endX) / 2;
  const midY = (startY + endY) / 2;
  const perpX = -chordY / chordLength;
  const perpY = chordX / chordLength;
  const sign = bulge > 0 ? -1 : 1;
  const centerX = midX + sign * perpX * distToCenter;
  const centerY = midY + sign * perpY * distToCenter;
  return { centerX, centerY, radius };
}
function drawArcFromBulge(ctx, realStartX, realStartY, realEndX, realEndY, bulge, realToCanvasMat) {
  if (Math.abs(bulge) < 1e-10) {
    const [endX, endY] = applyToPoint4(realToCanvasMat, [realEndX, realEndY]);
    ctx.lineTo(endX, endY);
    return;
  }
  const arc = computeArcFromBulge(
    realStartX,
    realStartY,
    realEndX,
    realEndY,
    bulge
  );
  if (!arc) {
    const [endX, endY] = applyToPoint4(realToCanvasMat, [realEndX, realEndY]);
    ctx.lineTo(endX, endY);
    return;
  }
  const [canvasStartX, canvasStartY] = applyToPoint4(realToCanvasMat, [
    realStartX,
    realStartY
  ]);
  const [canvasEndX, canvasEndY] = applyToPoint4(realToCanvasMat, [
    realEndX,
    realEndY
  ]);
  const [canvasCenterX, canvasCenterY] = applyToPoint4(realToCanvasMat, [
    arc.centerX,
    arc.centerY
  ]);
  const canvasRadius = Math.hypot(
    canvasStartX - canvasCenterX,
    canvasStartY - canvasCenterY
  );
  const startAngle = Math.atan2(
    canvasStartY - canvasCenterY,
    canvasStartX - canvasCenterX
  );
  const endAngle = Math.atan2(
    canvasEndY - canvasCenterY,
    canvasEndX - canvasCenterX
  );
  const det = realToCanvasMat.a * realToCanvasMat.d - realToCanvasMat.b * realToCanvasMat.c;
  const isFlipped = det < 0;
  const counterclockwise = bulge > 0 ? !isFlipped : isFlipped;
  ctx.arc(
    canvasCenterX,
    canvasCenterY,
    canvasRadius,
    startAngle,
    endAngle,
    counterclockwise
  );
}
function drawRing(ctx, ring, realToCanvasMat) {
  if (ring.vertices.length < 2) return;
  if (ring.vertices.length === 2) {
    const v0 = ring.vertices[0];
    const v1 = ring.vertices[1];
    if (v0 && v1 && Math.abs((v0.bulge ?? 0) - 1) < 1e-10 && Math.abs((v1.bulge ?? 0) - 1) < 1e-10) {
      const [x0, y0] = applyToPoint4(realToCanvasMat, [v0.x, v0.y]);
      ctx.moveTo(x0, y0);
      drawArcFromBulge(ctx, v0.x, v0.y, v1.x, v1.y, 1, realToCanvasMat);
      drawArcFromBulge(ctx, v1.x, v1.y, v0.x, v0.y, 1, realToCanvasMat);
      return;
    }
  }
  const firstVertex = ring.vertices[0];
  if (!firstVertex) return;
  const [firstX, firstY] = applyToPoint4(realToCanvasMat, [
    firstVertex.x,
    firstVertex.y
  ]);
  ctx.moveTo(firstX, firstY);
  for (let i = 0; i < ring.vertices.length; i++) {
    const currentVertex = ring.vertices[i];
    const nextIndex = (i + 1) % ring.vertices.length;
    const nextVertex = ring.vertices[nextIndex];
    if (!currentVertex || !nextVertex) continue;
    const bulge = currentVertex.bulge ?? 0;
    if (Math.abs(bulge) < 1e-10) {
      const [nextX, nextY] = applyToPoint4(realToCanvasMat, [
        nextVertex.x,
        nextVertex.y
      ]);
      ctx.lineTo(nextX, nextY);
    } else {
      drawArcFromBulge(
        ctx,
        currentVertex.x,
        currentVertex.y,
        nextVertex.x,
        nextVertex.y,
        bulge,
        realToCanvasMat
      );
    }
  }
}
function drawPcbCopperPour(params) {
  const { ctx, pour, realToCanvasMat, colorMap } = params;
  const color = layerToColor(pour.layer, colorMap);
  const opacity = layerToPourOpacity(pour.layer);
  ctx.save();
  ctx.globalAlpha = opacity;
  if (pour.shape === "rect") {
    const [cx, cy] = applyToPoint4(realToCanvasMat, [
      pour.center.x,
      pour.center.y
    ]);
    const scaledWidth = pour.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = pour.height * Math.abs(realToCanvasMat.a);
    ctx.translate(cx, cy);
    if (pour.rotation) {
      ctx.rotate(-pour.rotation * (Math.PI / 180));
    }
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    ctx.fillStyle = color;
    ctx.fill();
    ctx.restore();
    return;
  }
  if (pour.shape === "polygon") {
    if (pour.points && pour.points.length >= 3) {
      const canvasPoints = pour.points.map(
        (p) => applyToPoint4(realToCanvasMat, [p.x, p.y])
      );
      const firstPoint = canvasPoints[0];
      if (!firstPoint) {
        ctx.restore();
        return;
      }
      ctx.beginPath();
      const [firstX, firstY] = firstPoint;
      ctx.moveTo(firstX, firstY);
      for (let i = 1; i < canvasPoints.length; i++) {
        const point = canvasPoints[i];
        if (!point) continue;
        const [x, y] = point;
        ctx.lineTo(x, y);
      }
      ctx.closePath();
      ctx.fillStyle = color;
      ctx.fill();
    }
    ctx.restore();
    return;
  }
  if (pour.shape === "brep") {
    ctx.beginPath();
    drawRing(ctx, pour.brep_shape.outer_ring, realToCanvasMat);
    if (pour.brep_shape.inner_rings) {
      for (const innerRing of pour.brep_shape.inner_rings) {
        drawRing(ctx, innerRing, realToCanvasMat);
      }
    }
    ctx.fillStyle = color;
    ctx.fill("evenodd");
    ctx.restore();
    return;
  }
  ctx.restore();
}

// lib/drawer/elements/pcb-copper-text.ts
import { applyToPoint as applyToPoint6 } from "transformation-matrix";

// lib/drawer/shapes/text/text.ts
import { glyphLineAlphabet } from "@tscircuit/alphabet";
import { applyToPoint as applyToPoint5 } from "transformation-matrix";

// lib/drawer/shapes/text/getAlphabetLayout.ts
import {
  glyphAdvanceRatio,
  kerningRatio,
  textMetrics
} from "@tscircuit/alphabet";
var getAdvanceRatio = (char) => glyphAdvanceRatio[char] ?? (char === " " ? textMetrics.spaceWidthRatio : textMetrics.glyphWidthRatio);
function getAlphabetAdvanceWidth(char, nextChar, fontSize) {
  const advanceRatio = getAdvanceRatio(char);
  const letterSpacingRatio = nextChar ? textMetrics.letterSpacingRatio : 0;
  const kerningAdjustmentRatio = nextChar ? kerningRatio[char]?.[nextChar] ?? 0 : 0;
  return fontSize * (advanceRatio + letterSpacingRatio + kerningAdjustmentRatio);
}
function getAlphabetLayout(text, fontSize) {
  const glyphWidth = fontSize * textMetrics.glyphWidthRatio;
  const letterSpacing = fontSize * textMetrics.letterSpacingRatio;
  const spaceWidth = fontSize * textMetrics.spaceWidthRatio;
  const strokeWidth = fontSize * textMetrics.strokeWidthRatio;
  const lineHeight = fontSize * textMetrics.lineHeightRatio;
  const lines = text.replace(/\\n/g, "\n").split("\n");
  const lineWidths = lines.map((line) => {
    const characters = Array.from(line);
    return characters.reduce(
      (sum, char, index) => sum + getAlphabetAdvanceWidth(char, characters[index + 1], fontSize),
      0
    );
  });
  const width = lineWidths.reduce(
    (maxWidth, lineWidth) => Math.max(maxWidth, lineWidth),
    0
  );
  const height = lines.length > 1 ? fontSize + (lines.length - 1) * lineHeight : fontSize;
  return {
    width,
    height,
    glyphWidth,
    letterSpacing,
    spaceWidth,
    strokeWidth,
    lineHeight,
    lines,
    lineWidths
  };
}

// lib/drawer/shapes/text/getTextStartPosition.ts
function getTextStartPosition(alignment, layout) {
  const totalWidth = layout.width + layout.strokeWidth;
  const totalHeight = layout.height + layout.strokeWidth;
  let x = 0;
  let y = 0;
  if (alignment === "center" || alignment === "top_center" || alignment === "bottom_center") {
    x = -totalWidth / 2;
  } else if (alignment === "top_left" || alignment === "bottom_left" || alignment === "center_left") {
    x = 0;
  } else if (alignment === "top_right" || alignment === "bottom_right" || alignment === "center_right") {
    x = -totalWidth;
  } else if (alignment === "top_center" || alignment === "bottom_center") {
    x = -totalWidth / 2;
  }
  if (alignment === "center" || alignment === "center_left" || alignment === "center_right") {
    y = -totalHeight / 2;
  } else if (alignment === "top_left" || alignment === "top_right" || alignment === "top_center") {
    y = 0;
  } else if (alignment === "bottom_left" || alignment === "bottom_right" || alignment === "bottom_center") {
    y = -totalHeight;
  }
  return { x, y };
}
function getLineStartX(params) {
  const { alignment, lineWidth, maxWidth, strokeWidth } = params;
  const totalLineWidth = lineWidth + strokeWidth;
  const totalMaxWidth = maxWidth + strokeWidth;
  if (alignment === "top_left" || alignment === "bottom_left" || alignment === "center_left") {
    return 0;
  }
  if (alignment === "top_right" || alignment === "bottom_right" || alignment === "center_right") {
    return totalMaxWidth - totalLineWidth;
  }
  return (totalMaxWidth - totalLineWidth) / 2;
}

// lib/drawer/shapes/text/text.ts
var getGlyphLines = (char) => glyphLineAlphabet[char];
function strokeAlphabetLine(params) {
  const { ctx, line, fontSize, startX, startY, layout } = params;
  const { strokeWidth } = layout;
  const height = fontSize;
  const glyphScaleX = fontSize;
  const topY = startY;
  const characters = Array.from(line);
  let cursor = startX + strokeWidth / 2;
  characters.forEach((char, index) => {
    const glyphLines = getGlyphLines(char);
    if (glyphLines?.length) {
      ctx.beginPath();
      for (const glyph of glyphLines) {
        const x1 = cursor + glyph.x1 * glyphScaleX;
        const y1 = topY + (1 - glyph.y1) * height;
        const x2 = cursor + glyph.x2 * glyphScaleX;
        const y2 = topY + (1 - glyph.y2) * height;
        ctx.moveTo(x1, y1);
        ctx.lineTo(x2, y2);
      }
      ctx.stroke();
    }
    cursor += getAlphabetAdvanceWidth(char, characters[index + 1], fontSize);
  });
}
function strokeAlphabetText(params) {
  const {
    ctx,
    text,
    fontSize,
    startX,
    startY,
    anchorAlignment = "center"
  } = params;
  const layout = getAlphabetLayout(text, fontSize);
  const { lines, lineWidths, lineHeight, width, strokeWidth } = layout;
  lines.forEach((line, lineIndex) => {
    const lineStartX = startX + getLineStartX({
      alignment: anchorAlignment,
      lineWidth: lineWidths[lineIndex],
      maxWidth: width,
      strokeWidth
    });
    const lineStartY = startY + lineIndex * lineHeight;
    strokeAlphabetLine({
      ctx,
      line,
      fontSize,
      startX: lineStartX,
      startY: lineStartY,
      layout
    });
  });
}
var DEFAULT_KNOCKOUT_PADDING = {
  left: 0.2,
  right: 0.2,
  top: 0.2,
  bottom: 0.2
};
function drawText(params) {
  const {
    ctx,
    text,
    x,
    y,
    fontSize,
    color,
    realToCanvasMat,
    anchorAlignment,
    rotation = 0,
    mirrorX = false,
    knockout = false,
    knockoutPadding = {}
  } = params;
  if (!text) return;
  const [canvasX, canvasY] = applyToPoint5(realToCanvasMat, [x, y]);
  const scale2 = Math.abs(realToCanvasMat.a);
  const scaledFontSize = fontSize * scale2;
  const layout = getAlphabetLayout(text, scaledFontSize);
  const startPos = getTextStartPosition(anchorAlignment, layout);
  ctx.save();
  ctx.translate(canvasX, canvasY);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  if (mirrorX) {
    ctx.scale(-1, 1);
  }
  if (knockout) {
    const padding = { ...DEFAULT_KNOCKOUT_PADDING, ...knockoutPadding };
    const paddingLeft = padding.left * scale2;
    const paddingRight = padding.right * scale2;
    const paddingTop = padding.top * scale2;
    const paddingBottom = padding.bottom * scale2;
    const totalWidth = layout.width + layout.strokeWidth;
    const rectX = startPos.x - paddingLeft * 4;
    const rectY = startPos.y - paddingTop * 4;
    const rectWidth = totalWidth + paddingLeft * 2 + paddingRight * 2;
    const rectHeight = layout.height + layout.strokeWidth + paddingTop * 2 + paddingBottom * 2;
    ctx.fillStyle = color;
    ctx.fillRect(rectX, rectY, rectWidth, rectHeight);
  } else {
    ctx.strokeStyle = color;
  }
  ctx.lineWidth = layout.strokeWidth;
  ctx.lineCap = "round";
  ctx.lineJoin = "round";
  const { lines, lineWidths, lineHeight, width, strokeWidth } = layout;
  lines.forEach((line, lineIndex) => {
    const lineStartX = startPos.x + getLineStartX({
      alignment: anchorAlignment,
      lineWidth: lineWidths[lineIndex],
      maxWidth: width,
      strokeWidth
    });
    const lineStartY = startPos.y + lineIndex * lineHeight;
    strokeAlphabetLine({
      ctx,
      line,
      fontSize: scaledFontSize,
      startX: lineStartX,
      startY: lineStartY,
      layout
    });
  });
  ctx.restore();
}

// lib/drawer/elements/pcb-copper-text.ts
var DEFAULT_PADDING = { left: 0.2, right: 0.2, top: 0.2, bottom: 0.2 };
function mapAnchorAlignment(alignment) {
  if (!alignment) return "center";
  if (alignment.includes("left")) return "center_left";
  if (alignment.includes("right")) return "center_right";
  return "center";
}
function drawPcbCopperText(params) {
  const { ctx, text, realToCanvasMat, colorMap } = params;
  const content = text.text ?? "";
  if (!content) return;
  const [x, y] = applyToPoint6(realToCanvasMat, [
    text.anchor_position.x,
    text.anchor_position.y
  ]);
  const scale2 = Math.abs(realToCanvasMat.a);
  const fontSize = (text.font_size ?? 1) * scale2;
  const rotation = text.ccw_rotation ?? 0;
  const padding = {
    ...DEFAULT_PADDING,
    ...text.knockout_padding
  };
  const textColor = colorMap.copper[text.layer] ?? colorMap.copper.top;
  const layout = getAlphabetLayout(content, fontSize);
  const totalWidth = layout.width + layout.strokeWidth;
  const alignment = mapAnchorAlignment(text.anchor_alignment);
  const startPos = getTextStartPosition(alignment, layout);
  ctx.save();
  ctx.translate(x, y);
  if (text.is_mirrored) ctx.scale(-1, 1);
  if (rotation !== 0) ctx.rotate(-rotation * (Math.PI / 180));
  ctx.lineWidth = layout.strokeWidth;
  ctx.lineCap = "round";
  ctx.lineJoin = "round";
  if (text.is_knockout) {
    const paddingLeft = padding.left * scale2;
    const paddingRight = padding.right * scale2;
    const paddingTop = padding.top * scale2;
    const paddingBottom = padding.bottom * scale2;
    const rectX = startPos.x - paddingLeft * 4;
    const rectY = startPos.y - paddingTop * 4;
    const rectWidth = totalWidth + paddingLeft * 2 + paddingRight * 2;
    const rectHeight = layout.height + layout.strokeWidth + paddingTop * 2 + paddingBottom * 2;
    ctx.fillStyle = textColor;
    ctx.fillRect(rectX, rectY, rectWidth, rectHeight);
  } else {
    ctx.strokeStyle = textColor;
  }
  strokeAlphabetText({
    ctx,
    text: content,
    fontSize,
    startX: startPos.x,
    startY: startPos.y,
    anchorAlignment: alignment
  });
  ctx.restore();
}

// lib/drawer/shapes/circle.ts
import { applyToPoint as applyToPoint7 } from "transformation-matrix";
function drawCircle(params) {
  const {
    ctx,
    center,
    radius,
    fill,
    realToCanvasMat,
    stroke,
    strokeWidth,
    isStrokeDashed = false
  } = params;
  const [cx, cy] = applyToPoint7(realToCanvasMat, [center.x, center.y]);
  const scaledRadius = radius * Math.abs(realToCanvasMat.a);
  const scaledStrokeWidth = strokeWidth ? strokeWidth * Math.abs(realToCanvasMat.a) : void 0;
  ctx.beginPath();
  ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke && scaledStrokeWidth) {
    if (isStrokeDashed) {
      ctx.setLineDash([scaledStrokeWidth * 3, scaledStrokeWidth * 2]);
    } else {
      ctx.setLineDash([]);
    }
    ctx.strokeStyle = stroke;
    ctx.lineWidth = scaledStrokeWidth;
    ctx.stroke();
  }
}

// lib/drawer/elements/pcb-courtyard-circle.ts
function layerToCourtyardColor(layer, colorMap) {
  return layer === "bottom" ? colorMap.courtyard.bottom : colorMap.courtyard.top;
}
function drawPcbCourtyardCircle(params) {
  const { ctx, circle, realToCanvasMat, colorMap } = params;
  drawCircle({
    ctx,
    center: circle.center,
    radius: circle.radius,
    stroke: layerToCourtyardColor(circle.layer, colorMap),
    strokeWidth: 0.05,
    // Default thin line for courtyard info
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-courtyard-outline.ts
import { applyToPoint as applyToPoint8 } from "transformation-matrix";
function layerToCourtyardColor2(layer, colorMap) {
  return layer === "bottom" ? colorMap.courtyard.bottom : colorMap.courtyard.top;
}
function drawPcbCourtyardOutline(params) {
  const { ctx, outline, realToCanvasMat, colorMap } = params;
  if (!outline.outline || outline.outline.length < 2) return;
  ctx.beginPath();
  const startPoint = applyToPoint8(
    realToCanvasMat,
    outline.outline[0]
  );
  ctx.moveTo(startPoint.x, startPoint.y);
  for (let i = 1; i < outline.outline.length; i++) {
    const point = applyToPoint8(
      realToCanvasMat,
      outline.outline[i]
    );
    ctx.lineTo(point.x, point.y);
  }
  ctx.closePath();
  ctx.lineWidth = 0.05 * Math.abs(realToCanvasMat.a);
  ctx.strokeStyle = layerToCourtyardColor2(outline.layer, colorMap);
  ctx.stroke();
}

// lib/drawer/elements/pcb-courtyard-rect.ts
function layerToCourtyardColor3(layer, colorMap) {
  return layer === "bottom" ? colorMap.courtyard.bottom : colorMap.courtyard.top;
}
function drawPcbCourtyardRect(params) {
  const { ctx, rect, realToCanvasMat, colorMap } = params;
  drawRect({
    ctx,
    center: rect.center,
    width: rect.width,
    height: rect.height,
    stroke: layerToCourtyardColor3(rect.layer, colorMap),
    strokeWidth: 0.05,
    // Default thin line for courtyard info
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-cutout.ts
function drawPcbCutout(params) {
  const { ctx, cutout, realToCanvasMat, colorMap } = params;
  if (cutout.shape === "rect") {
    drawRect({
      ctx,
      center: cutout.center,
      width: cutout.width,
      height: cutout.height,
      fill: colorMap.drill,
      realToCanvasMat,
      ccwRotationDegrees: cutout.rotation ?? 0,
      borderRadius: cutout.corner_radius ?? 0
    });
    return;
  }
  if (cutout.shape === "circle") {
    drawCircle({
      ctx,
      center: cutout.center,
      radius: cutout.radius,
      fill: colorMap.drill,
      realToCanvasMat
    });
    return;
  }
  if (cutout.shape === "polygon") {
    if (cutout.points && cutout.points.length >= 3) {
      drawPolygon({
        ctx,
        points: cutout.points,
        fill: colorMap.drill,
        realToCanvasMat
      });
    }
    return;
  }
}

// lib/drawer/shapes/dimension-line.ts
import { applyToPoint as applyToPoint9 } from "transformation-matrix";
var TEXT_OFFSET_MULTIPLIER = 1.5;
var CHARACTER_WIDTH_MULTIPLIER = 0.6;
var TEXT_INTERSECTION_PADDING_MULTIPLIER = 0.3;
function normalize(v) {
  const len = Math.hypot(v.x, v.y) || 1;
  return { x: v.x / len, y: v.y / len };
}
function drawDimensionLine(params) {
  const {
    ctx,
    from,
    to,
    realToCanvasMat,
    color,
    fontSize,
    arrowSize = 1,
    strokeWidth: manualStrokeWidth,
    text,
    textRotation,
    offset
  } = params;
  const direction = normalize({ x: to.x - from.x, y: to.y - from.y });
  const perpendicular = { x: -direction.y, y: direction.x };
  const hasOffsetDirection = offset?.direction && typeof offset.direction.x === "number" && typeof offset.direction.y === "number";
  const normalizedOffsetDirection = hasOffsetDirection ? normalize(offset.direction) : { x: 0, y: 0 };
  const offsetMagnitude = offset?.distance ?? 0;
  const offsetVector = {
    x: normalizedOffsetDirection.x * offsetMagnitude,
    y: normalizedOffsetDirection.y * offsetMagnitude
  };
  const fromOffset = { x: from.x + offsetVector.x, y: from.y + offsetVector.y };
  const toOffset = { x: to.x + offsetVector.x, y: to.y + offsetVector.y };
  const fromBase = {
    x: fromOffset.x + direction.x * arrowSize,
    y: fromOffset.y + direction.y * arrowSize
  };
  const toBase = {
    x: toOffset.x - direction.x * arrowSize,
    y: toOffset.y - direction.y * arrowSize
  };
  const scaleValue = Math.abs(realToCanvasMat.a);
  const strokeWidth = manualStrokeWidth ?? arrowSize / 5;
  const lineColor = color || "rgba(255,255,255,0.5)";
  const extensionDirection = hasOffsetDirection && (Math.abs(normalizedOffsetDirection.x) > Number.EPSILON || Math.abs(normalizedOffsetDirection.y) > Number.EPSILON) ? normalizedOffsetDirection : perpendicular;
  const extensionLength = offsetMagnitude + 0.5;
  const allPoints = [];
  const getExtensionPoints = (anchor) => {
    const endPoint = {
      x: anchor.x + extensionDirection.x * extensionLength,
      y: anchor.y + extensionDirection.y * extensionLength
    };
    const halfWidth = strokeWidth / 2;
    const extPerpendicular = {
      x: -extensionDirection.y,
      y: extensionDirection.x
    };
    return [
      {
        x: anchor.x + extPerpendicular.x * halfWidth,
        y: anchor.y + extPerpendicular.y * halfWidth
      },
      {
        x: anchor.x - extPerpendicular.x * halfWidth,
        y: anchor.y - extPerpendicular.y * halfWidth
      },
      {
        x: endPoint.x - extPerpendicular.x * halfWidth,
        y: endPoint.y - extPerpendicular.y * halfWidth
      },
      {
        x: endPoint.x + extPerpendicular.x * halfWidth,
        y: endPoint.y + extPerpendicular.y * halfWidth
      }
    ];
  };
  allPoints.push(fromOffset);
  allPoints.push({
    x: fromBase.x + perpendicular.x * (arrowSize / 2),
    y: fromBase.y + perpendicular.y * (arrowSize / 2)
  });
  allPoints.push({
    x: fromBase.x + perpendicular.x * (strokeWidth / 2),
    y: fromBase.y + perpendicular.y * (strokeWidth / 2)
  });
  allPoints.push({
    x: toBase.x + perpendicular.x * (strokeWidth / 2),
    y: toBase.y + perpendicular.y * (strokeWidth / 2)
  });
  allPoints.push({
    x: toBase.x + perpendicular.x * (arrowSize / 2),
    y: toBase.y + perpendicular.y * (arrowSize / 2)
  });
  allPoints.push(toOffset);
  allPoints.push({
    x: toBase.x - perpendicular.x * (arrowSize / 2),
    y: toBase.y - perpendicular.y * (arrowSize / 2)
  });
  allPoints.push({
    x: toBase.x - perpendicular.x * (strokeWidth / 2),
    y: toBase.y - perpendicular.y * (strokeWidth / 2)
  });
  allPoints.push({
    x: fromBase.x - perpendicular.x * (strokeWidth / 2),
    y: fromBase.y - perpendicular.y * (strokeWidth / 2)
  });
  allPoints.push({
    x: fromBase.x - perpendicular.x * (arrowSize / 2),
    y: fromBase.y - perpendicular.y * (arrowSize / 2)
  });
  allPoints.push(fromOffset);
  const startPoint = allPoints[0];
  const addTick = (anchor) => {
    const pts = getExtensionPoints(anchor);
    allPoints.push(startPoint);
    allPoints.push(pts[0]);
    allPoints.push(...pts);
    allPoints.push(pts[0]);
    allPoints.push(startPoint);
  };
  addTick(from);
  addTick(to);
  drawPolygon({
    ctx,
    points: allPoints,
    fill: lineColor,
    realToCanvasMat
  });
  if (text) {
    const midPoint = {
      x: (from.x + to.x) / 2 + offsetVector.x,
      y: (from.y + to.y) / 2 + offsetVector.y
    };
    const [screenFromX, screenFromY] = applyToPoint9(realToCanvasMat, [
      fromOffset.x,
      fromOffset.y
    ]);
    const [screenToX, screenToY] = applyToPoint9(realToCanvasMat, [
      toOffset.x,
      toOffset.y
    ]);
    const screenDirection = normalize({
      x: screenToX - screenFromX,
      y: screenToY - screenFromY
    });
    let textAngle = Math.atan2(screenDirection.y, screenDirection.x) * 180 / Math.PI;
    if (textAngle > 90 || textAngle < -90) {
      textAngle += 180;
    }
    const finalTextAngle = typeof textRotation === "number" && Number.isFinite(textRotation) ? textAngle - textRotation : textAngle;
    let additionalOffset = 0;
    if (text && typeof textRotation === "number" && Number.isFinite(textRotation)) {
      const textWidth = text.length * fontSize * CHARACTER_WIDTH_MULTIPLIER;
      const textHeight = fontSize;
      const rotationRad = textRotation * Math.PI / 180;
      const sinRot = Math.abs(Math.sin(rotationRad));
      const cosRot = Math.abs(Math.cos(rotationRad));
      const halfWidth = textWidth / 2;
      const halfHeight = textHeight / 2;
      const maxExtension = halfWidth * sinRot + halfHeight * cosRot;
      additionalOffset = maxExtension + fontSize * TEXT_INTERSECTION_PADDING_MULTIPLIER;
    }
    const textOffset = arrowSize * TEXT_OFFSET_MULTIPLIER + additionalOffset;
    const textPoint = {
      x: midPoint.x + perpendicular.x * textOffset,
      y: midPoint.y + perpendicular.y * textOffset
    };
    drawText({
      ctx,
      text,
      x: textPoint.x,
      y: textPoint.y,
      fontSize,
      color: lineColor,
      realToCanvasMat,
      anchorAlignment: "center",
      rotation: -finalTextAngle
      // drawText expects CCW rotation in degrees
    });
  }
}

// lib/drawer/elements/pcb-fabrication-note-dimension.ts
function drawPcbFabricationNoteDimension(params) {
  const { ctx, pcbFabricationNoteDimension, realToCanvasMat, colorMap } = params;
  const color = pcbFabricationNoteDimension.color ?? colorMap.fabricationNote;
  drawDimensionLine({
    ctx,
    from: pcbFabricationNoteDimension.from,
    to: pcbFabricationNoteDimension.to,
    realToCanvasMat,
    color,
    fontSize: pcbFabricationNoteDimension.font_size ?? 1,
    arrowSize: pcbFabricationNoteDimension.arrow_size ?? 1,
    text: pcbFabricationNoteDimension.text,
    textRotation: pcbFabricationNoteDimension.text_ccw_rotation,
    offset: pcbFabricationNoteDimension.offset_distance && pcbFabricationNoteDimension.offset_direction ? {
      distance: pcbFabricationNoteDimension.offset_distance,
      direction: pcbFabricationNoteDimension.offset_direction
    } : void 0
  });
}

// lib/drawer/shapes/line.ts
import { applyToPoint as applyToPoint10 } from "transformation-matrix";
function drawLine(params) {
  const {
    ctx,
    start,
    end,
    strokeWidth,
    stroke,
    realToCanvasMat,
    lineCap = "round"
  } = params;
  const [x1, y1] = applyToPoint10(realToCanvasMat, [start.x, start.y]);
  const [x2, y2] = applyToPoint10(realToCanvasMat, [end.x, end.y]);
  const scaledStrokeWidth = strokeWidth * Math.abs(realToCanvasMat.a);
  ctx.beginPath();
  ctx.moveTo(x1, y1);
  ctx.lineTo(x2, y2);
  ctx.lineWidth = scaledStrokeWidth;
  ctx.strokeStyle = stroke;
  ctx.lineCap = lineCap;
  ctx.stroke();
}

// lib/drawer/elements/pcb-fabrication-note-path.ts
function drawPcbFabricationNotePath(params) {
  const { ctx, path, realToCanvasMat, colorMap } = params;
  const defaultColor = colorMap.fabricationNote;
  const color = path.color ?? defaultColor;
  if (!path.route || path.route.length < 2) return;
  for (let i = 0; i < path.route.length - 1; i++) {
    const start = path.route[i];
    const end = path.route[i + 1];
    if (!start || !end) continue;
    drawLine({
      ctx,
      start: { x: start.x, y: start.y },
      end: { x: end.x, y: end.y },
      strokeWidth: path.stroke_width ?? 0.1,
      stroke: color,
      realToCanvasMat
    });
  }
}

// lib/drawer/elements/pcb-fabrication-note-rect.ts
function drawPcbFabricationNoteRect(params) {
  const { ctx, rect, realToCanvasMat, colorMap } = params;
  const defaultColor = colorMap.fabricationNote;
  const color = rect.color ?? defaultColor;
  const isFilled = rect.is_filled ?? false;
  const hasStroke = rect.has_stroke ?? true;
  const isStrokeDashed = rect.is_stroke_dashed ?? false;
  drawRect({
    ctx,
    center: rect.center,
    width: rect.width,
    height: rect.height,
    fill: isFilled ? color : void 0,
    stroke: hasStroke ? color : void 0,
    strokeWidth: hasStroke ? rect.stroke_width : void 0,
    borderRadius: rect.corner_radius,
    realToCanvasMat,
    isStrokeDashed
  });
}

// lib/drawer/elements/pcb-fabrication-note-text.ts
function layerToColor2(layer, colorMap) {
  return colorMap.fabricationNote;
}
function drawPcbFabricationNoteText(params) {
  const { ctx, text, realToCanvasMat, colorMap } = params;
  const defaultColor = layerToColor2(text.layer, colorMap);
  const color = text.color ?? defaultColor;
  const fontSize = text.font_size;
  drawText({
    ctx,
    text: text.text,
    x: text.anchor_position.x,
    y: text.anchor_position.y,
    fontSize,
    color,
    realToCanvasMat,
    anchorAlignment: text.anchor_alignment
  });
}

// lib/drawer/shapes/oval.ts
import { applyToPoint as applyToPoint11 } from "transformation-matrix";
function drawOval(params) {
  const {
    ctx,
    center,
    radius_x,
    radius_y,
    fill,
    stroke,
    strokeWidth = 0.1,
    realToCanvasMat,
    rotation = 0
  } = params;
  const [cx, cy] = applyToPoint11(realToCanvasMat, [center.x, center.y]);
  const scaledRadiusX = radius_x * Math.abs(realToCanvasMat.a);
  const scaledRadiusY = radius_y * Math.abs(realToCanvasMat.a);
  const scaledStrokeWidth = strokeWidth * Math.abs(realToCanvasMat.a);
  ctx.save();
  ctx.translate(cx, cy);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  ctx.beginPath();
  ctx.ellipse(0, 0, scaledRadiusX, scaledRadiusY, 0, 0, Math.PI * 2);
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke) {
    ctx.strokeStyle = stroke;
    ctx.lineWidth = scaledStrokeWidth;
    ctx.stroke();
  }
  ctx.restore();
}

// lib/drawer/shapes/pill.ts
import { applyToPoint as applyToPoint12 } from "transformation-matrix";
function drawPill(params) {
  const {
    ctx,
    center,
    width,
    height,
    fill,
    realToCanvasMat,
    rotation = 0,
    stroke,
    strokeWidth
  } = params;
  const [cx, cy] = applyToPoint12(realToCanvasMat, [center.x, center.y]);
  const scaledWidth = width * Math.abs(realToCanvasMat.a);
  const scaledHeight = height * Math.abs(realToCanvasMat.a);
  const scaledStrokeWidth = strokeWidth ? strokeWidth * Math.abs(realToCanvasMat.a) : void 0;
  ctx.save();
  ctx.translate(cx, cy);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  ctx.beginPath();
  if (scaledWidth > scaledHeight) {
    const radius = scaledHeight / 2;
    const straightLength = scaledWidth - scaledHeight;
    ctx.moveTo(-straightLength / 2, -radius);
    ctx.lineTo(straightLength / 2, -radius);
    ctx.arc(straightLength / 2, 0, radius, -Math.PI / 2, Math.PI / 2);
    ctx.lineTo(-straightLength / 2, radius);
    ctx.arc(-straightLength / 2, 0, radius, Math.PI / 2, -Math.PI / 2);
  } else if (scaledHeight > scaledWidth) {
    const radius = scaledWidth / 2;
    const straightLength = scaledHeight - scaledWidth;
    ctx.moveTo(radius, -straightLength / 2);
    ctx.lineTo(radius, straightLength / 2);
    ctx.arc(0, straightLength / 2, radius, 0, Math.PI);
    ctx.lineTo(-radius, -straightLength / 2);
    ctx.arc(0, -straightLength / 2, radius, Math.PI, 0);
  } else {
    ctx.arc(0, 0, scaledWidth / 2, 0, Math.PI * 2);
  }
  ctx.closePath();
  if (fill) {
    ctx.fillStyle = fill;
    ctx.fill();
  }
  if (stroke && scaledStrokeWidth) {
    ctx.strokeStyle = stroke;
    ctx.lineWidth = scaledStrokeWidth;
    ctx.stroke();
  }
  ctx.restore();
}

// lib/drawer/elements/pcb-hole.ts
function getRotation(hole) {
  if ("ccw_rotation" in hole && typeof hole.ccw_rotation === "number") {
    return hole.ccw_rotation;
  }
  return 0;
}
function drawPcbHole(params) {
  const {
    ctx,
    hole,
    realToCanvasMat,
    colorMap,
    soldermaskMargin = 0,
    drawSoldermask = false
  } = params;
  if (drawSoldermask && hole.is_covered_with_solder_mask === true) {
    return;
  }
  const holeInset = soldermaskMargin < 0 ? Math.abs(soldermaskMargin) : 0;
  if (hole.hole_shape === "circle") {
    drawCircle({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius: hole.hole_diameter / 2 - holeInset,
      fill: colorMap.drill,
      realToCanvasMat
    });
    return;
  }
  if (hole.hole_shape === "square") {
    const ccwRotationDegrees = getRotation(hole);
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_diameter - holeInset * 2,
      height: hole.hole_diameter - holeInset * 2,
      fill: colorMap.drill,
      realToCanvasMat,
      ccwRotationDegrees
    });
    return;
  }
  if (hole.hole_shape === "oval") {
    const rotation = getRotation(hole);
    drawOval({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius_x: hole.hole_width / 2 - holeInset,
      radius_y: hole.hole_height / 2 - holeInset,
      fill: colorMap.drill,
      realToCanvasMat,
      rotation
    });
    return;
  }
  if (hole.hole_shape === "rect") {
    const ccwRotationDegrees = getRotation(hole);
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_width - holeInset * 2,
      height: hole.hole_height - holeInset * 2,
      fill: colorMap.drill,
      realToCanvasMat,
      ccwRotationDegrees
    });
    return;
  }
  if (hole.hole_shape === "pill" || hole.hole_shape === "rotated_pill") {
    const rotation = getRotation(hole);
    drawPill({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_width - holeInset * 2,
      height: hole.hole_height - holeInset * 2,
      fill: colorMap.drill,
      realToCanvasMat,
      rotation
    });
    return;
  }
}

// lib/drawer/elements/pcb-keepout.ts
import { applyToPoint as applyToPoint13 } from "transformation-matrix";
function hexToRgba(hex, alpha) {
  const r = parseInt(hex.slice(1, 3), 16);
  const g = parseInt(hex.slice(3, 5), 16);
  const b = parseInt(hex.slice(5, 7), 16);
  return `rgba(${r}, ${g}, ${b}, ${alpha})`;
}
function drawPcbKeepout(params) {
  const { ctx, keepout, realToCanvasMat, colorMap } = params;
  const strokeColor = colorMap.keepout;
  const fillColor = hexToRgba(colorMap.keepout, 0.2);
  const hatchSpacing = 1;
  if (keepout.shape === "rect") {
    const [cx, cy] = applyToPoint13(realToCanvasMat, [
      keepout.center.x,
      keepout.center.y
    ]);
    const scaledWidth = keepout.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = keepout.height * Math.abs(realToCanvasMat.a);
    const rotation = keepout.rotation ?? 0;
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * (Math.PI / 180));
    }
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    ctx.fillStyle = fillColor;
    ctx.fill();
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    ctx.clip();
    const scaledSpacing = hatchSpacing * Math.abs(realToCanvasMat.a);
    const diagonal = Math.sqrt(
      scaledWidth * scaledWidth + scaledHeight * scaledHeight
    );
    const halfWidth = scaledWidth / 2;
    const halfHeight = scaledHeight / 2;
    ctx.strokeStyle = strokeColor;
    ctx.lineWidth = 0.15 * Math.abs(realToCanvasMat.a);
    ctx.setLineDash([]);
    for (let offset = -diagonal; offset < diagonal * 2; offset += scaledSpacing) {
      ctx.beginPath();
      const startX = -halfWidth + offset;
      const startY = -halfHeight;
      const endX = -halfWidth + offset + diagonal;
      const endY = -halfHeight + diagonal;
      ctx.moveTo(startX, startY);
      ctx.lineTo(endX, endY);
      ctx.stroke();
    }
    ctx.restore();
    return;
  }
  if (keepout.shape === "circle") {
    const [cx, cy] = applyToPoint13(realToCanvasMat, [
      keepout.center.x,
      keepout.center.y
    ]);
    const scaledRadius = keepout.radius * Math.abs(realToCanvasMat.a);
    const scaledSpacing = hatchSpacing * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    ctx.beginPath();
    ctx.arc(0, 0, scaledRadius, 0, Math.PI * 2);
    ctx.fillStyle = fillColor;
    ctx.fill();
    ctx.beginPath();
    ctx.arc(0, 0, scaledRadius, 0, Math.PI * 2);
    ctx.clip();
    const diagonal = scaledRadius * 2;
    ctx.strokeStyle = strokeColor;
    ctx.lineWidth = 0.15 * Math.abs(realToCanvasMat.a);
    ctx.setLineDash([]);
    for (let offset = -diagonal; offset < diagonal * 2; offset += scaledSpacing) {
      ctx.beginPath();
      ctx.moveTo(offset - diagonal, -diagonal);
      ctx.lineTo(offset + diagonal, diagonal);
      ctx.stroke();
    }
    ctx.restore();
    return;
  }
}

// lib/drawer/elements/pcb-note-dimension.ts
var DEFAULT_NOTE_COLOR = "rgba(255,255,255,0.5)";
function drawPcbNoteDimension(params) {
  const { ctx, pcbNoteDimension, realToCanvasMat } = params;
  const color = pcbNoteDimension.color ?? DEFAULT_NOTE_COLOR;
  drawDimensionLine({
    ctx,
    from: pcbNoteDimension.from,
    to: pcbNoteDimension.to,
    realToCanvasMat,
    color,
    fontSize: pcbNoteDimension.font_size,
    arrowSize: pcbNoteDimension.arrow_size,
    text: pcbNoteDimension.text,
    textRotation: pcbNoteDimension.text_ccw_rotation,
    offset: pcbNoteDimension.offset_distance && pcbNoteDimension.offset_direction ? {
      distance: pcbNoteDimension.offset_distance,
      direction: pcbNoteDimension.offset_direction
    } : void 0
  });
}

// lib/drawer/elements/pcb-note-line.ts
import { applyToPoint as applyToPoint14 } from "transformation-matrix";
function drawPcbNoteLine(params) {
  const { ctx, line, realToCanvasMat, colorMap } = params;
  const defaultColor = "rgb(89, 148, 220)";
  const color = line.color ?? defaultColor;
  const strokeWidth = line.stroke_width ?? 0.1;
  const isDashed = line.is_dashed ?? false;
  const [x1, y1] = applyToPoint14(realToCanvasMat, [line.x1, line.y1]);
  const [x2, y2] = applyToPoint14(realToCanvasMat, [line.x2, line.y2]);
  const scaledStrokeWidth = strokeWidth * Math.abs(realToCanvasMat.a);
  ctx.save();
  if (isDashed) {
    ctx.setLineDash([scaledStrokeWidth * 2, scaledStrokeWidth * 2]);
  } else {
    ctx.setLineDash([]);
  }
  ctx.beginPath();
  ctx.moveTo(x1, y1);
  ctx.lineTo(x2, y2);
  ctx.lineWidth = scaledStrokeWidth;
  ctx.strokeStyle = color;
  ctx.lineCap = "round";
  ctx.stroke();
  ctx.restore();
}

// lib/drawer/elements/pcb-note-path.ts
function drawPcbNotePath(params) {
  const { ctx, path, realToCanvasMat, colorMap } = params;
  const defaultColor = "rgb(89, 148, 220)";
  const color = path.color ?? defaultColor;
  if (!path.route || path.route.length < 2) return;
  for (let i = 0; i < path.route.length - 1; i++) {
    const start = path.route[i];
    const end = path.route[i + 1];
    if (!start || !end) continue;
    drawLine({
      ctx,
      start: { x: start.x, y: start.y },
      end: { x: end.x, y: end.y },
      strokeWidth: path.stroke_width ?? 0.1,
      stroke: color,
      realToCanvasMat
    });
  }
}

// lib/drawer/elements/pcb-note-rect.ts
function drawPcbNoteRect(params) {
  const { ctx, rect, realToCanvasMat, colorMap } = params;
  const defaultColor = "rgb(89, 148, 220)";
  const color = rect.color ?? defaultColor;
  const isFilled = rect.is_filled ?? false;
  const hasStroke = rect.has_stroke ?? true;
  const isStrokeDashed = rect.is_stroke_dashed ?? false;
  drawRect({
    ctx,
    center: rect.center,
    width: rect.width,
    height: rect.height,
    fill: isFilled ? color : void 0,
    stroke: hasStroke ? color : void 0,
    strokeWidth: hasStroke ? rect.stroke_width : void 0,
    borderRadius: rect.corner_radius,
    realToCanvasMat,
    isStrokeDashed
  });
}

// lib/drawer/elements/pcb-note-text.ts
var DEFAULT_NOTE_TEXT_COLOR = "rgb(89, 148, 220)";
function drawPcbNoteText(params) {
  const { ctx, text, realToCanvasMat, colorMap } = params;
  const defaultColor = DEFAULT_NOTE_TEXT_COLOR;
  const color = text.color ?? defaultColor;
  const fontSize = text.font_size ?? 1;
  drawText({
    ctx,
    text: text.text ?? "",
    x: text.anchor_position.x,
    y: text.anchor_position.y,
    fontSize,
    color,
    realToCanvasMat,
    anchorAlignment: text.anchor_alignment ?? "center"
  });
}

// lib/drawer/elements/pcb-panel.ts
function drawPcbPanelElement(params) {
  const { ctx, panel, realToCanvasMat, colorMap, drawBoardMaterial } = params;
  const { width, height, center } = panel;
  if (width !== void 0 && height !== void 0 && center) {
    if (drawBoardMaterial) {
      drawRect({
        ctx,
        center,
        width,
        height,
        fill: colorMap.substrate,
        realToCanvasMat
      });
    }
    const halfWidth = width / 2;
    const halfHeight = height / 2;
    const corners = [
      { x: center.x - halfWidth, y: center.y - halfHeight },
      { x: center.x + halfWidth, y: center.y - halfHeight },
      { x: center.x + halfWidth, y: center.y + halfHeight },
      { x: center.x - halfWidth, y: center.y + halfHeight }
    ];
    drawPath({
      ctx,
      points: corners,
      stroke: colorMap.boardOutline,
      strokeWidth: 0.5,
      realToCanvasMat,
      closePath: true
    });
  }
}

// lib/drawer/elements/soldermask-margin.ts
import { applyToPoint as applyToPoint15 } from "transformation-matrix";
function drawSoldermaskRingForRect(ctx, center, width, height, margin, borderRadius, rotation, realToCanvasMat, soldermaskColor, padColor, asymmetricMargins) {
  const [cx, cy] = applyToPoint15(realToCanvasMat, [center.x, center.y]);
  const scaledWidth = width * Math.abs(realToCanvasMat.a);
  const scaledHeight = height * Math.abs(realToCanvasMat.a);
  const scaledRadius = borderRadius * Math.abs(realToCanvasMat.a);
  const ml = asymmetricMargins?.left ?? margin;
  const mr = asymmetricMargins?.right ?? margin;
  const mt = asymmetricMargins?.top ?? margin;
  const mb = asymmetricMargins?.bottom ?? margin;
  const scaledThicknessL = Math.max(0, -ml) * Math.abs(realToCanvasMat.a);
  const scaledThicknessR = Math.max(0, -mr) * Math.abs(realToCanvasMat.a);
  const scaledThicknessT = Math.max(0, -mt) * Math.abs(realToCanvasMat.a);
  const scaledThicknessB = Math.max(0, -mb) * Math.abs(realToCanvasMat.a);
  ctx.save();
  ctx.translate(cx, cy);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  const prevCompositeOp = ctx.globalCompositeOperation;
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = "source-atop";
  }
  const outerWidth = scaledWidth;
  const outerHeight = scaledHeight;
  const outerRadius = scaledRadius;
  ctx.beginPath();
  if (outerRadius > 0) {
    const x = -outerWidth / 2;
    const y = -outerHeight / 2;
    const r = Math.min(outerRadius, outerWidth / 2, outerHeight / 2);
    ctx.moveTo(x + r, y);
    ctx.lineTo(x + outerWidth - r, y);
    ctx.arcTo(x + outerWidth, y, x + outerWidth, y + r, r);
    ctx.lineTo(x + outerWidth, y + outerHeight - r);
    ctx.arcTo(
      x + outerWidth,
      y + outerHeight,
      x + outerWidth - r,
      y + outerHeight,
      r
    );
    ctx.lineTo(x + r, y + outerHeight);
    ctx.arcTo(x, y + outerHeight, x, y + outerHeight - r, r);
    ctx.lineTo(x, y + r);
    ctx.arcTo(x, y, x + r, y, r);
  } else {
    ctx.rect(-outerWidth / 2, -outerHeight / 2, outerWidth, outerHeight);
  }
  ctx.fillStyle = soldermaskColor;
  ctx.fill();
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = prevCompositeOp || "source-over";
  }
  const innerWidth = Math.max(
    0,
    scaledWidth - (scaledThicknessL + scaledThicknessR)
  );
  const innerHeight = Math.max(
    0,
    scaledHeight - (scaledThicknessT + scaledThicknessB)
  );
  const innerRadius = Math.max(
    0,
    scaledRadius - (scaledThicknessL + scaledThicknessR + scaledThicknessT + scaledThicknessB) / 4
  );
  if (innerWidth > 0 && innerHeight > 0) {
    ctx.beginPath();
    const x = -scaledWidth / 2 + scaledThicknessL;
    const y = -scaledHeight / 2 + scaledThicknessT;
    if (innerRadius > 0) {
      const r = Math.min(innerRadius, innerWidth / 2, innerHeight / 2);
      ctx.moveTo(x + r, y);
      ctx.lineTo(x + innerWidth - r, y);
      ctx.arcTo(x + innerWidth, y, x + innerWidth, y + r, r);
      ctx.lineTo(x + innerWidth, y + innerHeight - r);
      ctx.arcTo(
        x + innerWidth,
        y + innerHeight,
        x + innerWidth - r,
        y + innerHeight,
        r
      );
      ctx.lineTo(x + r, y + innerHeight);
      ctx.arcTo(x, y + innerHeight, x, y + innerHeight - r, r);
      ctx.lineTo(x, y + r);
      ctx.arcTo(x, y, x + r, y, r);
    } else {
      ctx.rect(x, y, innerWidth, innerHeight);
    }
    ctx.fillStyle = padColor;
    ctx.fill();
  }
  ctx.restore();
}
function drawSoldermaskRingForCircle(ctx, center, radius, margin, realToCanvasMat, soldermaskColor, padColor) {
  const [cx, cy] = applyToPoint15(realToCanvasMat, [center.x, center.y]);
  const scaledRadius = radius * Math.abs(realToCanvasMat.a);
  const scaledMargin = Math.abs(margin) * Math.abs(realToCanvasMat.a);
  ctx.save();
  const prevCompositeOp = ctx.globalCompositeOperation;
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = "source-atop";
  }
  ctx.beginPath();
  ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
  ctx.fillStyle = soldermaskColor;
  ctx.fill();
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = prevCompositeOp || "source-over";
  }
  const innerRadius = Math.max(0, scaledRadius - scaledMargin);
  if (innerRadius > 0) {
    ctx.beginPath();
    ctx.arc(cx, cy, innerRadius, 0, Math.PI * 2);
    ctx.fillStyle = padColor;
    ctx.fill();
  }
  ctx.restore();
}
function drawSoldermaskRingForPill(ctx, center, width, height, margin, rotation, realToCanvasMat, soldermaskColor, padColor) {
  const [cx, cy] = applyToPoint15(realToCanvasMat, [center.x, center.y]);
  const scaledWidth = width * Math.abs(realToCanvasMat.a);
  const scaledHeight = height * Math.abs(realToCanvasMat.a);
  const scaledMargin = Math.abs(margin) * Math.abs(realToCanvasMat.a);
  ctx.save();
  ctx.translate(cx, cy);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  const prevCompositeOp = ctx.globalCompositeOperation;
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = "source-atop";
  }
  const outerWidth = scaledWidth;
  const outerHeight = scaledHeight;
  ctx.beginPath();
  if (outerWidth > outerHeight) {
    const radius = outerHeight / 2;
    const straightLength = outerWidth - outerHeight;
    ctx.moveTo(-straightLength / 2, -radius);
    ctx.lineTo(straightLength / 2, -radius);
    ctx.arc(straightLength / 2, 0, radius, -Math.PI / 2, Math.PI / 2);
    ctx.lineTo(-straightLength / 2, radius);
    ctx.arc(-straightLength / 2, 0, radius, Math.PI / 2, -Math.PI / 2);
  } else if (outerHeight > outerWidth) {
    const radius = outerWidth / 2;
    const straightLength = outerHeight - outerWidth;
    ctx.moveTo(radius, -straightLength / 2);
    ctx.lineTo(radius, straightLength / 2);
    ctx.arc(0, straightLength / 2, radius, 0, Math.PI);
    ctx.lineTo(-radius, -straightLength / 2);
    ctx.arc(0, -straightLength / 2, radius, Math.PI, 0);
  } else {
    ctx.arc(0, 0, outerWidth / 2, 0, Math.PI * 2);
  }
  ctx.fillStyle = soldermaskColor;
  ctx.fill();
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = prevCompositeOp || "source-over";
  }
  const innerWidth = scaledWidth - scaledMargin * 2;
  const innerHeight = scaledHeight - scaledMargin * 2;
  if (innerWidth > 0 && innerHeight > 0) {
    ctx.beginPath();
    if (innerWidth > innerHeight) {
      const radius = innerHeight / 2;
      const straightLength = innerWidth - innerHeight;
      ctx.moveTo(-straightLength / 2, -radius);
      ctx.lineTo(straightLength / 2, -radius);
      ctx.arc(straightLength / 2, 0, radius, -Math.PI / 2, Math.PI / 2);
      ctx.lineTo(-straightLength / 2, radius);
      ctx.arc(-straightLength / 2, 0, radius, Math.PI / 2, -Math.PI / 2);
    } else if (innerHeight > innerWidth) {
      const radius = innerWidth / 2;
      const straightLength = innerHeight - innerWidth;
      ctx.moveTo(radius, -straightLength / 2);
      ctx.lineTo(radius, straightLength / 2);
      ctx.arc(0, straightLength / 2, radius, 0, Math.PI);
      ctx.lineTo(-radius, -straightLength / 2);
      ctx.arc(0, -straightLength / 2, radius, Math.PI, 0);
    } else {
      ctx.arc(0, 0, innerWidth / 2, 0, Math.PI * 2);
    }
    ctx.fillStyle = padColor;
    ctx.fill();
  }
  ctx.restore();
}
function drawSoldermaskRingForOval(ctx, center, radius_x, radius_y, margin, rotation, realToCanvasMat, soldermaskColor, holeColor) {
  const [cx, cy] = applyToPoint15(realToCanvasMat, [center.x, center.y]);
  const scaledRadiusX = radius_x * Math.abs(realToCanvasMat.a);
  const scaledRadiusY = radius_y * Math.abs(realToCanvasMat.a);
  const scaledMargin = Math.abs(margin) * Math.abs(realToCanvasMat.a);
  ctx.save();
  ctx.translate(cx, cy);
  if (rotation !== 0) {
    ctx.rotate(-rotation * (Math.PI / 180));
  }
  const prevCompositeOp = ctx.globalCompositeOperation;
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = "source-atop";
  }
  ctx.beginPath();
  ctx.ellipse(0, 0, scaledRadiusX, scaledRadiusY, 0, 0, Math.PI * 2);
  ctx.fillStyle = soldermaskColor;
  ctx.fill();
  if (ctx.globalCompositeOperation !== void 0) {
    ctx.globalCompositeOperation = prevCompositeOp || "source-over";
  }
  const innerRadiusX = Math.max(0, scaledRadiusX - scaledMargin);
  const innerRadiusY = Math.max(0, scaledRadiusY - scaledMargin);
  if (innerRadiusX > 0 && innerRadiusY > 0) {
    ctx.beginPath();
    ctx.ellipse(0, 0, innerRadiusX, innerRadiusY, 0, 0, Math.PI * 2);
    ctx.fillStyle = holeColor;
    ctx.fill();
  }
  ctx.restore();
}
function offsetPolygonPoints(points, offset) {
  if (points.length < 3 || offset === 0) return points;
  let centerX = 0;
  let centerY = 0;
  for (const point of points) {
    centerX += point.x;
    centerY += point.y;
  }
  centerX /= points.length;
  centerY /= points.length;
  const result = [];
  for (const point of points) {
    const vectorX = point.x - centerX;
    const vectorY = point.y - centerY;
    const distance = Math.sqrt(vectorX * vectorX + vectorY * vectorY);
    if (distance > 0) {
      const normalizedX = vectorX / distance;
      const normalizedY = vectorY / distance;
      result.push({
        x: point.x + normalizedX * offset,
        y: point.y + normalizedY * offset
      });
    } else {
      result.push({
        x: point.x + offset,
        y: point.y
      });
    }
  }
  return result;
}

// lib/drawer/elements/pcb-plated-hole.ts
function drawPcbPlatedHole(params) {
  const {
    ctx,
    hole,
    realToCanvasMat,
    colorMap,
    soldermaskMargin = 0,
    drawSoldermask
  } = params;
  if (hole.is_covered_with_solder_mask === true && drawSoldermask) {
    return;
  }
  const copperColor = colorMap.copper.top;
  const copperInset = soldermaskMargin < 0 ? Math.abs(soldermaskMargin) : 0;
  if (hole.shape === "circle") {
    drawCircle({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius: hole.outer_diameter / 2 - copperInset,
      fill: copperColor,
      realToCanvasMat
    });
    cutCurrentLayer(ctx, () => {
      drawCircle({
        ctx,
        center: { x: hole.x, y: hole.y },
        radius: hole.hole_diameter / 2,
        fill: "#000",
        realToCanvasMat
      });
    });
    drawCircle({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius: hole.hole_diameter / 2,
      fill: colorMap.drill,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "oval") {
    drawOval({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius_x: hole.outer_width / 2 - copperInset,
      radius_y: hole.outer_height / 2 - copperInset,
      fill: copperColor,
      realToCanvasMat,
      rotation: hole.ccw_rotation
    });
    cutCurrentLayer(ctx, () => {
      drawOval({
        ctx,
        center: { x: hole.x, y: hole.y },
        radius_x: hole.hole_width / 2,
        radius_y: hole.hole_height / 2,
        fill: "#000",
        realToCanvasMat,
        rotation: hole.ccw_rotation
      });
    });
    drawOval({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius_x: hole.hole_width / 2,
      radius_y: hole.hole_height / 2,
      fill: colorMap.drill,
      realToCanvasMat,
      rotation: hole.ccw_rotation
    });
    return;
  }
  if (hole.shape === "pill") {
    drawPill({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.outer_width - copperInset * 2,
      height: hole.outer_height - copperInset * 2,
      fill: copperColor,
      realToCanvasMat,
      rotation: hole.ccw_rotation
    });
    cutCurrentLayer(ctx, () => {
      drawPill({
        ctx,
        center: { x: hole.x, y: hole.y },
        width: hole.hole_width,
        height: hole.hole_height,
        fill: "#000",
        realToCanvasMat,
        rotation: hole.ccw_rotation
      });
    });
    drawPill({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_width,
      height: hole.hole_height,
      fill: colorMap.drill,
      realToCanvasMat,
      rotation: hole.ccw_rotation
    });
    return;
  }
  if (hole.shape === "circular_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width - copperInset * 2,
      height: hole.rect_pad_height - copperInset * 2,
      fill: copperColor,
      realToCanvasMat,
      borderRadius: hole.rect_border_radius ? Math.max(0, hole.rect_border_radius - copperInset) : 0,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
    const holeX = hole.x + (hole.hole_offset_x ?? 0);
    const holeY = hole.y + (hole.hole_offset_y ?? 0);
    cutCurrentLayer(ctx, () => {
      drawCircle({
        ctx,
        center: { x: holeX, y: holeY },
        radius: hole.hole_diameter / 2,
        fill: "#000",
        realToCanvasMat
      });
    });
    drawCircle({
      ctx,
      center: { x: holeX, y: holeY },
      radius: hole.hole_diameter / 2,
      fill: colorMap.drill,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "pill_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width - copperInset * 2,
      height: hole.rect_pad_height - copperInset * 2,
      fill: copperColor,
      realToCanvasMat,
      borderRadius: hole.rect_border_radius ? Math.max(0, hole.rect_border_radius - copperInset) : 0
    });
    const holeX = hole.x + (hole.hole_offset_x ?? 0);
    const holeY = hole.y + (hole.hole_offset_y ?? 0);
    cutCurrentLayer(ctx, () => {
      drawPill({
        ctx,
        center: { x: holeX, y: holeY },
        width: hole.hole_width,
        height: hole.hole_height,
        fill: "#000",
        realToCanvasMat
      });
    });
    drawPill({
      ctx,
      center: { x: holeX, y: holeY },
      width: hole.hole_width,
      height: hole.hole_height,
      fill: colorMap.drill,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "rotated_pill_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width - copperInset * 2,
      height: hole.rect_pad_height - copperInset * 2,
      fill: copperColor,
      realToCanvasMat,
      borderRadius: hole.rect_border_radius ? Math.max(0, hole.rect_border_radius - copperInset) : 0,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
    const holeX = hole.x + (hole.hole_offset_x ?? 0);
    const holeY = hole.y + (hole.hole_offset_y ?? 0);
    cutCurrentLayer(ctx, () => {
      drawPill({
        ctx,
        center: { x: holeX, y: holeY },
        width: hole.hole_width,
        height: hole.hole_height,
        fill: "#000",
        realToCanvasMat,
        rotation: hole.hole_ccw_rotation
      });
    });
    drawPill({
      ctx,
      center: { x: holeX, y: holeY },
      width: hole.hole_width,
      height: hole.hole_height,
      fill: colorMap.drill,
      realToCanvasMat,
      rotation: hole.hole_ccw_rotation
    });
    return;
  }
  if (hole.shape === "hole_with_polygon_pad") {
    const padOutline = hole.pad_outline;
    if (padOutline && padOutline.length >= 3) {
      const padPoints = padOutline.map((point) => ({
        x: hole.x + point.x,
        y: hole.y + point.y
      }));
      const copperPoints = copperInset > 0 ? offsetPolygonPoints(padPoints, -copperInset) : padPoints;
      if (copperPoints.length >= 3) {
        drawPolygon({
          ctx,
          points: copperPoints,
          fill: copperColor,
          realToCanvasMat
        });
      }
    }
    const holeX = hole.x + (hole.hole_offset_x ?? 0);
    const holeY = hole.y + (hole.hole_offset_y ?? 0);
    const holeShape = hole.hole_shape;
    if (holeShape === "circle") {
      cutCurrentLayer(ctx, () => {
        drawCircle({
          ctx,
          center: { x: holeX, y: holeY },
          radius: (hole.hole_diameter ?? 0) / 2,
          fill: "#000",
          realToCanvasMat
        });
      });
      drawCircle({
        ctx,
        center: { x: holeX, y: holeY },
        radius: (hole.hole_diameter ?? 0) / 2,
        fill: colorMap.drill,
        realToCanvasMat
      });
    } else if (holeShape === "oval") {
      cutCurrentLayer(ctx, () => {
        drawOval({
          ctx,
          center: { x: holeX, y: holeY },
          radius_x: (hole.hole_width ?? 0) / 2,
          radius_y: (hole.hole_height ?? 0) / 2,
          fill: "#000",
          realToCanvasMat
        });
      });
      drawOval({
        ctx,
        center: { x: holeX, y: holeY },
        radius_x: (hole.hole_width ?? 0) / 2,
        radius_y: (hole.hole_height ?? 0) / 2,
        fill: colorMap.drill,
        realToCanvasMat
      });
    } else if (holeShape === "pill") {
      cutCurrentLayer(ctx, () => {
        drawPill({
          ctx,
          center: { x: holeX, y: holeY },
          width: hole.hole_width ?? 0,
          height: hole.hole_height ?? 0,
          fill: "#000",
          realToCanvasMat
        });
      });
      drawPill({
        ctx,
        center: { x: holeX, y: holeY },
        width: hole.hole_width ?? 0,
        height: hole.hole_height ?? 0,
        fill: colorMap.drill,
        realToCanvasMat
      });
    } else if (holeShape === "rotated_pill") {
      cutCurrentLayer(ctx, () => {
        drawPill({
          ctx,
          center: { x: holeX, y: holeY },
          width: hole.hole_width ?? 0,
          height: hole.hole_height ?? 0,
          fill: "#000",
          realToCanvasMat
        });
      });
      drawPill({
        ctx,
        center: { x: holeX, y: holeY },
        width: hole.hole_width ?? 0,
        height: hole.hole_height ?? 0,
        fill: colorMap.drill,
        realToCanvasMat
      });
    }
    return;
  }
}
function cutCurrentLayer(ctx, drawCutPath) {
  ctx.save();
  ctx.globalCompositeOperation = "destination-out";
  drawCutPath();
  ctx.restore();
}

// lib/drawer/elements/pcb-silkscreen-circle.ts
function layerToSilkscreenColor(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenCircle(params) {
  const { ctx, circle, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor(circle.layer, colorMap);
  drawCircle({
    ctx,
    center: circle.center,
    radius: circle.radius,
    stroke: color,
    strokeWidth: circle.stroke_width,
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-silkscreen-line.ts
function layerToSilkscreenColor2(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenLine(params) {
  const { ctx, line, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor2(line.layer, colorMap);
  drawLine({
    ctx,
    start: { x: line.x1, y: line.y1 },
    end: { x: line.x2, y: line.y2 },
    strokeWidth: line.stroke_width ?? 0.1,
    stroke: color,
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-silkscreen-oval.ts
function layerToSilkscreenColor3(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenOval(params) {
  const { ctx, oval, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor3(oval.layer, colorMap);
  drawOval({
    ctx,
    center: oval.center,
    radius_x: oval.radius_x,
    radius_y: oval.radius_y,
    stroke: color,
    strokeWidth: 0.1,
    realToCanvasMat,
    rotation: oval.ccw_rotation ?? 0
  });
}

// lib/drawer/elements/pcb-silkscreen-path.ts
function layerToSilkscreenColor4(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenPath(params) {
  const { ctx, path, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor4(path.layer, colorMap);
  if (!path.route || path.route.length < 2) return;
  for (let i = 0; i < path.route.length - 1; i++) {
    const start = path.route[i];
    const end = path.route[i + 1];
    if (!start || !end) continue;
    drawLine({
      ctx,
      start: { x: start.x, y: start.y },
      end: { x: end.x, y: end.y },
      strokeWidth: path.stroke_width ?? 0.1,
      stroke: color,
      realToCanvasMat
    });
  }
}

// lib/drawer/elements/pcb-silkscreen-pill.ts
function layerToSilkscreenColor5(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenPill(params) {
  const { ctx, pill, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor5(pill.layer, colorMap);
  const strokeWidth = 0.2;
  drawPill({
    ctx,
    center: pill.center,
    width: pill.width,
    height: pill.height,
    stroke: color,
    strokeWidth,
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-silkscreen-rect.ts
function layerToSilkscreenColor6(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function drawPcbSilkscreenRect(params) {
  const { ctx, rect, realToCanvasMat, colorMap } = params;
  const color = layerToSilkscreenColor6(rect.layer, colorMap);
  drawRect({
    ctx,
    center: rect.center,
    width: rect.width,
    height: rect.height,
    fill: rect.is_filled ? color : "transparent",
    stroke: color,
    strokeWidth: rect.stroke_width,
    realToCanvasMat
  });
}

// lib/drawer/elements/pcb-silkscreen-text.ts
function layerToSilkscreenColor7(layer, colorMap) {
  return layer === "bottom" ? colorMap.silkscreen.bottom : colorMap.silkscreen.top;
}
function mapAnchorAlignment2(alignment) {
  if (!alignment) return "center";
  return alignment;
}
function drawPcbSilkscreenText(params) {
  const { ctx, text, realToCanvasMat, colorMap } = params;
  const content = text.text ?? "";
  if (!content) return;
  const color = layerToSilkscreenColor7(text.layer, colorMap);
  const alignment = mapAnchorAlignment2(text.anchor_alignment);
  const fontSize = text.font_size ?? 1;
  drawText({
    ctx,
    text: content,
    x: text.anchor_position.x,
    y: text.anchor_position.y,
    fontSize,
    color,
    realToCanvasMat,
    anchorAlignment: alignment,
    rotation: text.ccw_rotation ?? 0,
    mirrorX: text.layer === "bottom",
    knockout: text.is_knockout,
    knockoutPadding: text.is_knockout ? text.knockout_padding : void 0
  });
}

// lib/drawer/elements/pcb-smtpad.ts
function layerToColor3(layer, colorMap) {
  return colorMap.copper[layer] ?? colorMap.copper.top;
}
function getBorderRadius(pad) {
  if (pad.shape === "rect" || pad.shape === "rotated_rect") {
    return pad.corner_radius ?? pad.rect_border_radius ?? 0;
  }
  return 0;
}
function drawPcbSmtPad(params) {
  const { ctx, pad, realToCanvasMat, colorMap } = params;
  const color = layerToColor3(pad.layer, colorMap);
  if (pad.shape === "rect") {
    drawRect({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      fill: color,
      realToCanvasMat,
      borderRadius: getBorderRadius(pad)
    });
    return;
  }
  if (pad.shape === "rotated_rect") {
    drawRect({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      fill: color,
      realToCanvasMat,
      borderRadius: getBorderRadius(pad),
      ccwRotationDegrees: pad.ccw_rotation ?? 0
    });
    return;
  }
  if (pad.shape === "circle") {
    drawCircle({
      ctx,
      center: { x: pad.x, y: pad.y },
      radius: pad.radius,
      fill: color,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "pill") {
    drawPill({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      fill: color,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "rotated_pill") {
    drawPill({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      fill: color,
      realToCanvasMat,
      rotation: pad.ccw_rotation ?? 0
    });
    return;
  }
  if (pad.shape === "polygon") {
    if (pad.points && pad.points.length >= 3) {
      drawPolygon({
        ctx,
        points: pad.points,
        fill: color,
        realToCanvasMat
      });
    }
    return;
  }
}

// lib/drawer/elements/pcb-soldermask/board.ts
import { applyToPoint as applyToPoint16 } from "transformation-matrix";

// lib/drawer/elements/helper-functions/draw-polygon.ts
function drawPolygonPath(params) {
  const { ctx, points } = params;
  if (points.length < 3) return;
  const firstPoint = points[0];
  if (firstPoint) {
    ctx.moveTo(firstPoint.x, firstPoint.y);
    for (let i = 1; i < points.length; i++) {
      const point = points[i];
      if (point) {
        ctx.lineTo(point.x, point.y);
      }
    }
    ctx.closePath();
  }
}

// lib/drawer/elements/pcb-soldermask/board.ts
function drawBoardSoldermask(params) {
  const { ctx, board, realToCanvasMat, soldermaskColor } = params;
  const { width, height, center, outline } = board;
  if (outline && Array.isArray(outline) && outline.length >= 3) {
    const canvasPoints = outline.map((p) => {
      const [x, y] = applyToPoint16(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    ctx.beginPath();
    drawPolygonPath({ ctx, points: canvasPoints });
    ctx.fillStyle = soldermaskColor;
    ctx.fill();
  } else if (width !== void 0 && height !== void 0 && center) {
    const [cx, cy] = applyToPoint16(realToCanvasMat, [center.x, center.y]);
    const scaledWidth = width * Math.abs(realToCanvasMat.a);
    const scaledHeight = height * Math.abs(realToCanvasMat.a);
    ctx.fillStyle = soldermaskColor;
    ctx.fillRect(
      cx - scaledWidth / 2,
      cy - scaledHeight / 2,
      scaledWidth,
      scaledHeight
    );
  }
}

// lib/drawer/elements/pcb-soldermask/merge-soldermask-layer.ts
function isPatternCapableContext(ctx) {
  return "createPattern" in ctx && typeof Reflect.get(ctx, "createPattern") === "function";
}
function mergeSoldermaskLayer(baseCtx, soldermaskCtx) {
  if (baseCtx === soldermaskCtx) return;
  if (soldermaskCtx.canvas.width <= 0 || soldermaskCtx.canvas.height <= 0)
    return;
  if (!isPatternCapableContext(baseCtx)) return;
  let pattern = null;
  try {
    pattern = baseCtx.createPattern(
      soldermaskCtx.canvas,
      "no-repeat"
    );
  } catch {
    return;
  }
  if (!pattern) return;
  baseCtx.save();
  baseCtx.globalCompositeOperation = "source-over";
  baseCtx.fillStyle = pattern;
  baseCtx.fillRect(
    0,
    0,
    soldermaskCtx.canvas.width,
    soldermaskCtx.canvas.height
  );
  baseCtx.restore();
}

// lib/drawer/elements/pcb-soldermask/create-soldermask-layer-context.ts
function isCanvasConstructor(value) {
  return typeof value === "function";
}
function createSoldermaskLayerContext(baseCtx, width, height) {
  if (width <= 0 || height <= 0) return null;
  const g = globalThis;
  let layerCanvas = null;
  if ("OffscreenCanvas" in g && typeof g.OffscreenCanvas === "function") {
    layerCanvas = new g.OffscreenCanvas(width, height);
  } else if ("document" in g && g.document && typeof g.document.createElement === "function") {
    const canvas = g.document.createElement("canvas");
    canvas.width = width;
    canvas.height = height;
    layerCanvas = canvas;
  } else {
    const CanvasCtor = baseCtx.canvas.constructor;
    if (isCanvasConstructor(CanvasCtor)) {
      try {
        layerCanvas = new CanvasCtor(width, height);
      } catch {
        return null;
      }
    }
  }
  if (!layerCanvas) return null;
  return layerCanvas.getContext("2d");
}

// lib/drawer/elements/pcb-soldermask/panel.ts
import { applyToPoint as applyToPoint17 } from "transformation-matrix";
function drawPanelSoldermask(params) {
  const { ctx, panel, realToCanvasMat, soldermaskColor } = params;
  const { width, height, center } = panel;
  if (width === void 0 || height === void 0 || !center) return;
  const [cx, cy] = applyToPoint17(realToCanvasMat, [center.x, center.y]);
  const scaledWidth = width * Math.abs(realToCanvasMat.a);
  const scaledHeight = height * Math.abs(realToCanvasMat.a);
  ctx.fillStyle = soldermaskColor;
  ctx.fillRect(
    cx - scaledWidth / 2,
    cy - scaledHeight / 2,
    scaledWidth,
    scaledHeight
  );
}

// lib/drawer/elements/pcb-soldermask/cutout.ts
import { applyToPoint as applyToPoint19 } from "transformation-matrix";

// lib/drawer/elements/helper-functions/draw-rounded-rect.ts
import { applyToPoint as applyToPoint18, compose, rotate, translate } from "transformation-matrix";
function drawRoundedRectPath(params) {
  const {
    ctx,
    cx: centerX,
    cy: centerY,
    width,
    height,
    radius,
    ccwRotationDegrees = 0
  } = params;
  const ccwRotationRadians = ccwRotationDegrees * Math.PI / 180;
  const realToPxTransform = compose(
    translate(centerX, centerY),
    rotate(-ccwRotationRadians)
  );
  const rectHalfWidth = width / 2;
  const rectHalfHeight = height / 2;
  const rectCornerRadius = Math.min(radius, rectHalfWidth, rectHalfHeight);
  if (rectCornerRadius > 0) {
    const rectTopLeftCorner = applyToPoint18(realToPxTransform, {
      x: -rectHalfWidth,
      y: -rectHalfHeight
    });
    const rectTopRightCorner = applyToPoint18(realToPxTransform, {
      x: rectHalfWidth,
      y: -rectHalfHeight
    });
    const rectBottomRightCorner = applyToPoint18(realToPxTransform, {
      x: rectHalfWidth,
      y: rectHalfHeight
    });
    const rectBottomLeftCorner = applyToPoint18(realToPxTransform, {
      x: -rectHalfWidth,
      y: rectHalfHeight
    });
    const leftEdgeMidpoint = applyToPoint18(realToPxTransform, {
      x: -rectHalfWidth,
      y: 0
    });
    ctx.moveTo(leftEdgeMidpoint.x, leftEdgeMidpoint.y);
    ctx.arcTo(
      rectTopLeftCorner.x,
      rectTopLeftCorner.y,
      rectTopRightCorner.x,
      rectTopRightCorner.y,
      rectCornerRadius
    );
    ctx.arcTo(
      rectTopRightCorner.x,
      rectTopRightCorner.y,
      rectBottomRightCorner.x,
      rectBottomRightCorner.y,
      rectCornerRadius
    );
    ctx.arcTo(
      rectBottomRightCorner.x,
      rectBottomRightCorner.y,
      rectBottomLeftCorner.x,
      rectBottomLeftCorner.y,
      rectCornerRadius
    );
    ctx.arcTo(
      rectBottomLeftCorner.x,
      rectBottomLeftCorner.y,
      rectTopLeftCorner.x,
      rectTopLeftCorner.y,
      rectCornerRadius
    );
  } else {
    const rectTopLeftCorner = applyToPoint18(realToPxTransform, {
      x: -rectHalfWidth,
      y: -rectHalfHeight
    });
    const rectTopRightCorner = applyToPoint18(realToPxTransform, {
      x: rectHalfWidth,
      y: -rectHalfHeight
    });
    const rectBottomRightCorner = applyToPoint18(realToPxTransform, {
      x: rectHalfWidth,
      y: rectHalfHeight
    });
    const rectBottomLeftCorner = applyToPoint18(realToPxTransform, {
      x: -rectHalfWidth,
      y: rectHalfHeight
    });
    ctx.moveTo(rectTopLeftCorner.x, rectTopLeftCorner.y);
    ctx.lineTo(rectTopRightCorner.x, rectTopRightCorner.y);
    ctx.lineTo(rectBottomRightCorner.x, rectBottomRightCorner.y);
    ctx.lineTo(rectBottomLeftCorner.x, rectBottomLeftCorner.y);
  }
  ctx.closePath();
}

// lib/drawer/elements/pcb-soldermask/cut-path-from-soldermask.ts
function cutPathFromSoldermask(ctx) {
  ctx.save();
  ctx.globalCompositeOperation = "destination-out";
  ctx.fill();
  ctx.restore();
}

// lib/drawer/elements/pcb-soldermask/cutout.ts
function processCutoutSoldermask(params) {
  const { ctx, cutout, realToCanvasMat } = params;
  if (cutout.shape === "rect") {
    const [cx, cy] = applyToPoint19(realToCanvasMat, [
      cutout.center?.x ?? 0,
      cutout.center?.y ?? 0
    ]);
    const scaledWidth = cutout.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = cutout.height * Math.abs(realToCanvasMat.a);
    const scaledRadius = (cutout.corner_radius ?? 0) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (cutout.rotation) {
      ctx.rotate(-cutout.rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight,
      radius: scaledRadius
    });
    ctx.restore();
    cutPathFromSoldermask(ctx);
  } else if (cutout.shape === "circle") {
    const [cx, cy] = applyToPoint19(realToCanvasMat, [
      cutout.center?.x ?? 0,
      cutout.center?.y ?? 0
    ]);
    const scaledRadius = cutout.radius * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
    ctx.closePath();
    cutPathFromSoldermask(ctx);
  } else if (cutout.shape === "polygon" && cutout.points && cutout.points.length >= 3) {
    const canvasPoints = cutout.points.map((p) => {
      const [x, y] = applyToPoint19(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    ctx.beginPath();
    drawPolygonPath({ ctx, points: canvasPoints });
    cutPathFromSoldermask(ctx);
  }
}

// lib/drawer/elements/pcb-soldermask/copper-pour/copper-pour.ts
import { applyToPoint as applyToPoint22 } from "transformation-matrix";

// lib/drawer/elements/pcb-soldermask/copper-pour/cut-covered-pour-openings.ts
var CUTOUT_MASK_COLOR = "rgba(255, 255, 255, 1)";
function cutCoveredPourOpenings(params) {
  const { ctx, elements, layer, realToCanvasMat } = params;
  ctx.save();
  ctx.globalCompositeOperation = "destination-out";
  for (const element of elements) {
    if (element.type === "pcb_smtpad") {
      if (element.layer !== layer) continue;
      if (element.is_covered_with_solder_mask === true) continue;
      cutSmtPad({ ctx, pad: element, realToCanvasMat });
      continue;
    }
    if (element.type === "pcb_plated_hole") {
      if (element.layers && !element.layers.includes(layer)) continue;
      if (element.is_covered_with_solder_mask === true) continue;
      cutPlatedHole({ ctx, hole: element, realToCanvasMat });
      continue;
    }
    if (element.type === "pcb_via") {
      if (element.layers && !element.layers.includes(layer)) continue;
      drawCircle({
        ctx,
        center: { x: element.x, y: element.y },
        radius: element.outer_diameter / 2,
        fill: CUTOUT_MASK_COLOR,
        realToCanvasMat
      });
      continue;
    }
    if (element.type === "pcb_hole") {
      if (element.is_covered_with_solder_mask === true) continue;
      cutHole({ ctx, hole: element, realToCanvasMat });
    }
  }
  ctx.restore();
}
function cutSmtPad(params) {
  const { ctx, pad, realToCanvasMat } = params;
  if (pad.shape === "rect") {
    drawRect({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      borderRadius: pad.corner_radius ?? pad.rect_border_radius ?? 0,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "rotated_rect") {
    drawRect({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      borderRadius: pad.corner_radius ?? pad.rect_border_radius ?? 0,
      ccwRotationDegrees: pad.ccw_rotation ?? 0,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "circle") {
    drawCircle({
      ctx,
      center: { x: pad.x, y: pad.y },
      radius: pad.radius,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "pill" || pad.shape === "rotated_pill") {
    drawPill({
      ctx,
      center: { x: pad.x, y: pad.y },
      width: pad.width,
      height: pad.height,
      rotation: pad.shape === "rotated_pill" ? pad.ccw_rotation ?? 0 : 0,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (pad.shape === "polygon" && pad.points && pad.points.length >= 3) {
    drawPolygon({
      ctx,
      points: pad.points,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
  }
}
function cutPlatedHole(params) {
  const { ctx, hole, realToCanvasMat } = params;
  if (hole.shape === "circle") {
    drawCircle({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius: hole.outer_diameter / 2,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "oval") {
    drawOval({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius_x: hole.outer_width / 2,
      radius_y: hole.outer_height / 2,
      rotation: hole.ccw_rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "pill") {
    drawPill({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.outer_width,
      height: hole.outer_height,
      rotation: hole.ccw_rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "circular_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width,
      height: hole.rect_pad_height,
      borderRadius: hole.rect_border_radius ?? 0,
      ccwRotationDegrees: hole.rect_ccw_rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "pill_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width,
      height: hole.rect_pad_height,
      borderRadius: hole.rect_border_radius ?? 0,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "rotated_pill_hole_with_rect_pad") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.rect_pad_width,
      height: hole.rect_pad_height,
      borderRadius: hole.rect_border_radius ?? 0,
      ccwRotationDegrees: hole.rect_ccw_rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.shape === "hole_with_polygon_pad" && hole.pad_outline && hole.pad_outline.length >= 3) {
    drawPolygon({
      ctx,
      points: hole.pad_outline.map((point) => ({
        x: hole.x + point.x,
        y: hole.y + point.y
      })),
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
  }
}
function cutHole(params) {
  const { ctx, hole, realToCanvasMat } = params;
  const rotation = "ccw_rotation" in hole && typeof hole.ccw_rotation === "number" ? hole.ccw_rotation : 0;
  if (hole.hole_shape === "circle") {
    drawCircle({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius: hole.hole_diameter / 2,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.hole_shape === "square") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_diameter,
      height: hole.hole_diameter,
      ccwRotationDegrees: rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.hole_shape === "oval") {
    drawOval({
      ctx,
      center: { x: hole.x, y: hole.y },
      radius_x: hole.hole_width / 2,
      radius_y: hole.hole_height / 2,
      rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.hole_shape === "rect") {
    drawRect({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_width,
      height: hole.hole_height,
      ccwRotationDegrees: rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
    return;
  }
  if (hole.hole_shape === "pill" || hole.hole_shape === "rotated_pill") {
    drawPill({
      ctx,
      center: { x: hole.x, y: hole.y },
      width: hole.hole_width,
      height: hole.hole_height,
      rotation,
      fill: CUTOUT_MASK_COLOR,
      realToCanvasMat
    });
  }
}

// lib/drawer/elements/pcb-soldermask/copper-pour/draw-ring.ts
import { applyToPoint as applyToPoint21 } from "transformation-matrix";

// lib/drawer/elements/pcb-soldermask/copper-pour/arc-from-bulge.ts
import { applyToPoint as applyToPoint20 } from "transformation-matrix";
function computeArcFromBulge2(startX, startY, endX, endY, bulge) {
  if (Math.abs(bulge) < 1e-10) return null;
  const chordX = endX - startX;
  const chordY = endY - startY;
  const chordLength = Math.hypot(chordX, chordY);
  if (chordLength < 1e-10) return null;
  const sagitta = Math.abs(bulge) * (chordLength / 2);
  const halfChord = chordLength / 2;
  const radius = (sagitta * sagitta + halfChord * halfChord) / (2 * sagitta);
  const distToCenter = radius - sagitta;
  const midX = (startX + endX) / 2;
  const midY = (startY + endY) / 2;
  const perpX = -chordY / chordLength;
  const perpY = chordX / chordLength;
  const sign = bulge > 0 ? -1 : 1;
  return {
    centerX: midX + sign * perpX * distToCenter,
    centerY: midY + sign * perpY * distToCenter,
    radius
  };
}
function drawArcFromBulge2(ctx, realStartX, realStartY, realEndX, realEndY, bulge, realToCanvasMat) {
  if (Math.abs(bulge) < 1e-10) {
    const [endX, endY] = applyToPoint20(realToCanvasMat, [realEndX, realEndY]);
    ctx.lineTo(endX, endY);
    return;
  }
  const arc = computeArcFromBulge2(
    realStartX,
    realStartY,
    realEndX,
    realEndY,
    bulge
  );
  if (!arc) {
    const [endX, endY] = applyToPoint20(realToCanvasMat, [realEndX, realEndY]);
    ctx.lineTo(endX, endY);
    return;
  }
  const [canvasStartX, canvasStartY] = applyToPoint20(realToCanvasMat, [
    realStartX,
    realStartY
  ]);
  const [canvasEndX, canvasEndY] = applyToPoint20(realToCanvasMat, [
    realEndX,
    realEndY
  ]);
  const [canvasCenterX, canvasCenterY] = applyToPoint20(realToCanvasMat, [
    arc.centerX,
    arc.centerY
  ]);
  const canvasRadius = Math.hypot(
    canvasStartX - canvasCenterX,
    canvasStartY - canvasCenterY
  );
  const startAngle = Math.atan2(
    canvasStartY - canvasCenterY,
    canvasStartX - canvasCenterX
  );
  const endAngle = Math.atan2(
    canvasEndY - canvasCenterY,
    canvasEndX - canvasCenterX
  );
  const det = realToCanvasMat.a * realToCanvasMat.d - realToCanvasMat.b * realToCanvasMat.c;
  const isFlipped = det < 0;
  const counterclockwise = bulge > 0 ? !isFlipped : isFlipped;
  ctx.arc(
    canvasCenterX,
    canvasCenterY,
    canvasRadius,
    startAngle,
    endAngle,
    counterclockwise
  );
}

// lib/drawer/elements/pcb-soldermask/copper-pour/draw-ring.ts
function drawRing2(ctx, ring, realToCanvasMat) {
  if (ring.vertices.length < 2) return;
  if (ring.vertices.length === 2) {
    const v0 = ring.vertices[0];
    const v1 = ring.vertices[1];
    if (v0 && v1 && Math.abs((v0.bulge ?? 0) - 1) < 1e-10 && Math.abs((v1.bulge ?? 0) - 1) < 1e-10) {
      const [x0, y0] = applyToPoint21(realToCanvasMat, [v0.x, v0.y]);
      ctx.moveTo(x0, y0);
      drawArcFromBulge2(ctx, v0.x, v0.y, v1.x, v1.y, 1, realToCanvasMat);
      drawArcFromBulge2(ctx, v1.x, v1.y, v0.x, v0.y, 1, realToCanvasMat);
      return;
    }
  }
  const firstVertex = ring.vertices[0];
  if (!firstVertex) return;
  const [firstX, firstY] = applyToPoint21(realToCanvasMat, [
    firstVertex.x,
    firstVertex.y
  ]);
  ctx.moveTo(firstX, firstY);
  for (let i = 0; i < ring.vertices.length; i++) {
    const currentVertex = ring.vertices[i];
    const nextVertex = ring.vertices[(i + 1) % ring.vertices.length];
    if (!currentVertex || !nextVertex) continue;
    const bulge = currentVertex.bulge ?? 0;
    if (Math.abs(bulge) < 1e-10) {
      const [nextX, nextY] = applyToPoint21(realToCanvasMat, [
        nextVertex.x,
        nextVertex.y
      ]);
      ctx.lineTo(nextX, nextY);
    } else {
      drawArcFromBulge2(
        ctx,
        currentVertex.x,
        currentVertex.y,
        nextVertex.x,
        nextVertex.y,
        bulge,
        realToCanvasMat
      );
    }
  }
}

// lib/drawer/elements/pcb-soldermask/copper-pour/copper-pour.ts
function processCopperPourSoldermask(params) {
  const {
    ctx,
    pour,
    realToCanvasMat,
    soldermaskOverCopperColor,
    layer,
    elements
  } = params;
  if (pour.layer !== layer) return;
  if (pour.covered_with_solder_mask !== true) return;
  ctx.save();
  ctx.fillStyle = soldermaskOverCopperColor;
  if (pour.shape === "rect") {
    const [cx, cy] = applyToPoint22(realToCanvasMat, [
      pour.center.x,
      pour.center.y
    ]);
    const scaledWidth = pour.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = pour.height * Math.abs(realToCanvasMat.a);
    ctx.translate(cx, cy);
    if (pour.rotation) {
      ctx.rotate(-pour.rotation * (Math.PI / 180));
    }
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    ctx.fill();
    ctx.restore();
    cutCoveredPourOpenings({
      ctx,
      elements,
      layer,
      realToCanvasMat
    });
    return;
  }
  if (pour.shape === "polygon") {
    if (pour.points && pour.points.length >= 3) {
      const canvasPoints = pour.points.map(
        (p) => applyToPoint22(realToCanvasMat, [p.x, p.y])
      );
      const firstPoint = canvasPoints[0];
      if (firstPoint) {
        ctx.beginPath();
        const [firstX, firstY] = firstPoint;
        ctx.moveTo(firstX, firstY);
        for (let i = 1; i < canvasPoints.length; i++) {
          const point = canvasPoints[i];
          if (!point) continue;
          const [x, y] = point;
          ctx.lineTo(x, y);
        }
        ctx.closePath();
        ctx.fill();
        cutCoveredPourOpenings({
          ctx,
          elements,
          layer,
          realToCanvasMat
        });
      }
    }
    ctx.restore();
    return;
  }
  if (pour.shape === "brep") {
    ctx.beginPath();
    drawRing2(ctx, pour.brep_shape.outer_ring, realToCanvasMat);
    if (pour.brep_shape.inner_rings) {
      for (const innerRing of pour.brep_shape.inner_rings) {
        drawRing2(ctx, innerRing, realToCanvasMat);
      }
    }
    ctx.fill("evenodd");
    cutCoveredPourOpenings({
      ctx,
      elements,
      layer,
      realToCanvasMat
    });
    ctx.restore();
    return;
  }
  ctx.restore();
}

// lib/drawer/elements/pcb-soldermask/hole.ts
import { applyToPoint as applyToPoint24 } from "transformation-matrix";

// lib/drawer/elements/helper-functions/draw-pill.ts
import { applyToPoint as applyToPoint23, compose as compose2, rotate as rotate2, translate as translate2 } from "transformation-matrix";
function drawPillPath(params) {
  const {
    ctx,
    cx: centerX,
    cy: centerY,
    width,
    height,
    ccwRotationDegrees = 0
  } = params;
  const ccwRotationRadians = ccwRotationDegrees * Math.PI / 180;
  const realToPxTransform = compose2(
    translate2(centerX, centerY),
    rotate2(-ccwRotationRadians)
  );
  if (width > height) {
    const semicircleRadius = height / 2;
    const semicircleCenterOffset = (width - height) / 2;
    const leftSemicircleCenter = applyToPoint23(realToPxTransform, {
      x: -semicircleCenterOffset,
      y: 0
    });
    const rightSemicircleCenter = applyToPoint23(realToPxTransform, {
      x: semicircleCenterOffset,
      y: 0
    });
    const startPoint = applyToPoint23(realToPxTransform, {
      x: semicircleCenterOffset,
      y: -semicircleRadius
    });
    ctx.moveTo(startPoint.x, startPoint.y);
    ctx.arc(
      rightSemicircleCenter.x,
      rightSemicircleCenter.y,
      semicircleRadius,
      -ccwRotationRadians - Math.PI / 2,
      -ccwRotationRadians + Math.PI / 2
    );
    ctx.arc(
      leftSemicircleCenter.x,
      leftSemicircleCenter.y,
      semicircleRadius,
      -ccwRotationRadians + Math.PI / 2,
      -ccwRotationRadians - Math.PI / 2
    );
  } else if (height > width) {
    const semicircleRadius = width / 2;
    const semicircleCenterOffset = (height - width) / 2;
    const topSemicircleCenter = applyToPoint23(realToPxTransform, {
      x: 0,
      y: -semicircleCenterOffset
    });
    const bottomSemicircleCenter = applyToPoint23(realToPxTransform, {
      x: 0,
      y: semicircleCenterOffset
    });
    const startPoint = applyToPoint23(realToPxTransform, {
      x: semicircleRadius,
      y: semicircleCenterOffset
    });
    ctx.moveTo(startPoint.x, startPoint.y);
    ctx.arc(
      bottomSemicircleCenter.x,
      bottomSemicircleCenter.y,
      semicircleRadius,
      -ccwRotationRadians,
      -ccwRotationRadians + Math.PI
    );
    ctx.arc(
      topSemicircleCenter.x,
      topSemicircleCenter.y,
      semicircleRadius,
      -ccwRotationRadians + Math.PI,
      -ccwRotationRadians
    );
  } else {
    const radius = width / 2;
    ctx.moveTo(centerX + radius, centerY);
    ctx.arc(centerX, centerY, radius, 0, Math.PI * 2);
  }
  ctx.closePath();
}

// lib/drawer/elements/pcb-soldermask/hole.ts
function processHoleSoldermask(params) {
  const { ctx, hole, realToCanvasMat, soldermaskOverCopperColor } = params;
  const isCoveredWithSoldermask = hole.is_covered_with_solder_mask === true;
  const margin = hole.soldermask_margin ?? 0;
  if (isCoveredWithSoldermask) {
    ctx.fillStyle = soldermaskOverCopperColor;
    drawHoleShapePath({ ctx, hole, realToCanvasMat, margin: 0 });
    ctx.fill();
  } else if (margin < 0) {
    drawHoleShapePath({ ctx, hole, realToCanvasMat, margin });
    cutPathFromSoldermask(ctx);
    drawNegativeMarginRingForHole({
      ctx,
      hole,
      realToCanvasMat,
      soldermaskOverCopperColor,
      margin
    });
  } else if (margin > 0) {
    drawHoleShapePath({ ctx, hole, realToCanvasMat, margin });
    cutPathFromSoldermask(ctx);
  } else {
    drawHoleShapePath({ ctx, hole, realToCanvasMat, margin: 0 });
    cutPathFromSoldermask(ctx);
  }
}
function getHoleRotation(hole) {
  if ("ccw_rotation" in hole && typeof hole.ccw_rotation === "number") {
    return hole.ccw_rotation;
  }
  return 0;
}
function drawHoleShapePath(params) {
  const { ctx, hole, realToCanvasMat, margin } = params;
  const rotation = getHoleRotation(hole);
  if (hole.hole_shape === "circle") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledRadius = (hole.hole_diameter / 2 + margin) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
    ctx.closePath();
  } else if (hole.hole_shape === "square") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledSize = (hole.hole_diameter + margin * 2) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.rect(-scaledSize / 2, -scaledSize / 2, scaledSize, scaledSize);
    ctx.closePath();
    ctx.restore();
  } else if (hole.hole_shape === "oval") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledRadiusX = (hole.hole_width / 2 + margin) * Math.abs(realToCanvasMat.a);
    const scaledRadiusY = (hole.hole_height / 2 + margin) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.ellipse(0, 0, scaledRadiusX, scaledRadiusY, 0, 0, Math.PI * 2);
    ctx.closePath();
    ctx.restore();
  } else if (hole.hole_shape === "rect") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledWidth = (hole.hole_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const scaledHeight = (hole.hole_height + margin * 2) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    ctx.closePath();
    ctx.restore();
  } else if (hole.hole_shape === "pill" || hole.hole_shape === "rotated_pill") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledWidth = (hole.hole_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const scaledHeight = (hole.hole_height + margin * 2) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight
    });
    ctx.restore();
  }
}
function drawNegativeMarginRingForHole(params) {
  const { ctx, hole, realToCanvasMat, soldermaskOverCopperColor, margin } = params;
  const thickness = Math.abs(margin);
  const rotation = getHoleRotation(hole);
  ctx.fillStyle = soldermaskOverCopperColor;
  if (hole.hole_shape === "circle") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledRadius = hole.hole_diameter / 2 * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    const innerRadius = Math.max(0, scaledRadius - scaledThickness);
    ctx.save();
    ctx.beginPath();
    ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
    if (innerRadius > 0) {
      ctx.arc(cx, cy, innerRadius, 0, Math.PI * 2, true);
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (hole.hole_shape === "square") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledSize = hole.hole_diameter * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    const innerSize = Math.max(0, scaledSize - scaledThickness * 2);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.rect(-scaledSize / 2, -scaledSize / 2, scaledSize, scaledSize);
    if (innerSize > 0) {
      ctx.rect(-innerSize / 2, -innerSize / 2, innerSize, innerSize);
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (hole.hole_shape === "oval") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledRadiusX = hole.hole_width / 2 * Math.abs(realToCanvasMat.a);
    const scaledRadiusY = hole.hole_height / 2 * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    const innerRadiusX = Math.max(0, scaledRadiusX - scaledThickness);
    const innerRadiusY = Math.max(0, scaledRadiusY - scaledThickness);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.ellipse(0, 0, scaledRadiusX, scaledRadiusY, 0, 0, Math.PI * 2);
    if (innerRadiusX > 0 && innerRadiusY > 0) {
      ctx.moveTo(innerRadiusX, 0);
      ctx.ellipse(0, 0, innerRadiusX, innerRadiusY, 0, 0, Math.PI * 2);
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (hole.hole_shape === "rect") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledWidth = hole.hole_width * Math.abs(realToCanvasMat.a);
    const scaledHeight = hole.hole_height * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    const innerWidth = Math.max(0, scaledWidth - scaledThickness * 2);
    const innerHeight = Math.max(0, scaledHeight - scaledThickness * 2);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    ctx.rect(-scaledWidth / 2, -scaledHeight / 2, scaledWidth, scaledHeight);
    if (innerWidth > 0 && innerHeight > 0) {
      ctx.rect(-innerWidth / 2, -innerHeight / 2, innerWidth, innerHeight);
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (hole.hole_shape === "pill" || hole.hole_shape === "rotated_pill") {
    const [cx, cy] = applyToPoint24(realToCanvasMat, [hole.x, hole.y]);
    const scaledWidth = hole.hole_width * Math.abs(realToCanvasMat.a);
    const scaledHeight = hole.hole_height * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight
    });
    const innerWidth = scaledWidth - scaledThickness * 2;
    const innerHeight = scaledHeight - scaledThickness * 2;
    if (innerWidth > 0 && innerHeight > 0) {
      drawPillPath({
        ctx,
        cx: 0,
        cy: 0,
        width: innerWidth,
        height: innerHeight
      });
    }
    ctx.fill("evenodd");
    ctx.restore();
  }
}

// lib/drawer/elements/pcb-soldermask/plated-hole.ts
import { applyToPoint as applyToPoint25 } from "transformation-matrix";
function processPlatedHoleSoldermask(params) {
  const { ctx, hole, realToCanvasMat, soldermaskOverCopperColor, layer } = params;
  if (hole.layers && !hole.layers.includes(layer)) return;
  const isCoveredWithSoldermask = hole.is_covered_with_solder_mask === true;
  const margin = hole.soldermask_margin ?? 0;
  if (isCoveredWithSoldermask) {
    ctx.fillStyle = soldermaskOverCopperColor;
    drawPlatedHoleShapePath({ ctx, hole, realToCanvasMat, margin: 0 });
    ctx.fill();
  } else if (margin < 0) {
    drawPlatedHoleShapePath({ ctx, hole, realToCanvasMat, margin });
    cutPathFromSoldermask(ctx);
    drawNegativeMarginRingForPlatedHole({
      ctx,
      hole,
      realToCanvasMat,
      soldermaskOverCopperColor,
      margin
    });
  } else if (margin > 0) {
    drawPlatedHoleShapePath({ ctx, hole, realToCanvasMat, margin });
    cutPathFromSoldermask(ctx);
  } else {
    drawPlatedHoleShapePath({ ctx, hole, realToCanvasMat, margin: 0 });
    cutPathFromSoldermask(ctx);
  }
}
function drawPlatedHoleShapePath(params) {
  const { ctx, hole, realToCanvasMat, margin } = params;
  if (hole.shape === "circle") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const radius = (hole.outer_diameter / 2 + margin) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    ctx.arc(centerX, centerY, radius, 0, Math.PI * 2);
    ctx.closePath();
  } else if (hole.shape === "oval") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const ovalRadiusX = (hole.outer_width / 2 + margin) * Math.abs(realToCanvasMat.a);
    const ovalRadiusY = (hole.outer_height / 2 + margin) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    ctx.ellipse(
      centerX,
      centerY,
      ovalRadiusX,
      ovalRadiusY,
      -((hole.ccw_rotation ?? 0) * Math.PI) / 180,
      0,
      Math.PI * 2
    );
    ctx.closePath();
  } else if (hole.shape === "pill") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const pillWidth = (hole.outer_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const pillHeight = (hole.outer_height + margin * 2) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: pillWidth,
      height: pillHeight,
      ccwRotationDegrees: hole.ccw_rotation
    });
  } else if (hole.shape === "circular_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const rectPadWidth = (hole.rect_pad_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadHeight = (hole.rect_pad_height + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadCornerRadius = (hole.rect_border_radius ?? 0) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: rectPadWidth,
      height: rectPadHeight,
      radius: rectPadCornerRadius,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
  } else if (hole.shape === "pill_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const rectPadWidth = (hole.rect_pad_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadHeight = (hole.rect_pad_height + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadCornerRadius = (hole.rect_border_radius ?? 0) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: rectPadWidth,
      height: rectPadHeight,
      radius: rectPadCornerRadius
    });
  } else if (hole.shape === "rotated_pill_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const rectPadWidth = (hole.rect_pad_width + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadHeight = (hole.rect_pad_height + margin * 2) * Math.abs(realToCanvasMat.a);
    const rectPadCornerRadius = (hole.rect_border_radius ?? 0) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: rectPadWidth,
      height: rectPadHeight,
      radius: rectPadCornerRadius,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
  } else if (hole.shape === "hole_with_polygon_pad" && hole.pad_outline && hole.pad_outline.length >= 3) {
    const padPoints = hole.pad_outline.map(
      (point) => ({
        x: hole.x + point.x,
        y: hole.y + point.y
      })
    );
    const marginPoints = margin !== 0 ? offsetPolygonPoints(padPoints, margin) : padPoints;
    const canvasPoints = marginPoints.map((p) => {
      const [x, y] = applyToPoint25(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    ctx.beginPath();
    drawPolygonPath({ ctx, points: canvasPoints });
  }
}
function drawNegativeMarginRingForPlatedHole(params) {
  const { ctx, hole, realToCanvasMat, soldermaskOverCopperColor, margin } = params;
  const marginThickness = Math.abs(margin);
  ctx.fillStyle = soldermaskOverCopperColor;
  if (hole.shape === "circle") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const outerRadius = hole.outer_diameter / 2 * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    const innerRadius = Math.max(0, outerRadius - marginMagnitude);
    ctx.beginPath();
    ctx.arc(centerX, centerY, outerRadius, 0, Math.PI * 2);
    if (innerRadius > 0) {
      ctx.arc(centerX, centerY, innerRadius, 0, Math.PI * 2, true);
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "oval") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const padRadiusX = hole.outer_width / 2 * Math.abs(realToCanvasMat.a);
    const padRadiusY = hole.outer_height / 2 * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    const innerRadiusX = Math.max(0, padRadiusX - marginMagnitude);
    const innerRadiusY = Math.max(0, padRadiusY - marginMagnitude);
    const ccwRotation = -((hole.ccw_rotation ?? 0) * Math.PI) / 180;
    ctx.beginPath();
    ctx.ellipse(
      centerX,
      centerY,
      padRadiusX,
      padRadiusY,
      ccwRotation,
      0,
      Math.PI * 2
    );
    if (innerRadiusX > 0 && innerRadiusY > 0) {
      const soldermaskEllipseStartX = centerX + innerRadiusX * Math.cos(ccwRotation);
      const soldermaskEllipseStartY = centerY + innerRadiusX * Math.sin(ccwRotation);
      ctx.moveTo(soldermaskEllipseStartX, soldermaskEllipseStartY);
      ctx.ellipse(
        centerX,
        centerY,
        innerRadiusX,
        innerRadiusY,
        ccwRotation,
        0,
        Math.PI * 2
      );
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "pill") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const padWidth = hole.outer_width * Math.abs(realToCanvasMat.a);
    const padHeight = hole.outer_height * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: padWidth,
      height: padHeight,
      ccwRotationDegrees: hole.ccw_rotation
    });
    const soldermaskPadWidth = padWidth - marginMagnitude * 2;
    const soldermaskPadHeight = padHeight - marginMagnitude * 2;
    if (soldermaskPadWidth > 0 && soldermaskPadHeight > 0) {
      drawPillPath({
        ctx,
        cx: centerX,
        cy: centerY,
        width: soldermaskPadWidth,
        height: soldermaskPadHeight,
        ccwRotationDegrees: hole.ccw_rotation
      });
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "circular_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const padWidth = hole.rect_pad_width * Math.abs(realToCanvasMat.a);
    const padHeight = hole.rect_pad_height * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    const padCornerRadius = hole.rect_border_radius ? hole.rect_border_radius * Math.abs(realToCanvasMat.a) : 0;
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: padWidth,
      height: padHeight,
      radius: padCornerRadius,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
    const soldermaskPadWidth = padWidth - marginMagnitude * 2;
    const soldermaskPadHeight = padHeight - marginMagnitude * 2;
    if (soldermaskPadWidth > 0 && soldermaskPadHeight > 0) {
      const soldermaskPadCornerRadius = Math.max(
        0,
        padCornerRadius - marginMagnitude
      );
      drawRoundedRectPath({
        ctx,
        cx: centerX,
        cy: centerY,
        width: soldermaskPadWidth,
        height: soldermaskPadHeight,
        radius: soldermaskPadCornerRadius,
        ccwRotationDegrees: hole.rect_ccw_rotation
      });
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "pill_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const padWidth = hole.rect_pad_width * Math.abs(realToCanvasMat.a);
    const padHeight = hole.rect_pad_height * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    const padCornerRadius = hole.rect_border_radius ? hole.rect_border_radius * Math.abs(realToCanvasMat.a) : 0;
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: padWidth,
      height: padHeight,
      radius: padCornerRadius
    });
    const soldermaskPadWidth = padWidth - marginMagnitude * 2;
    const soldermaskPadHeight = padHeight - marginMagnitude * 2;
    if (soldermaskPadWidth > 0 && soldermaskPadHeight > 0) {
      const soldermaskPadCornerRadius = Math.max(
        0,
        padCornerRadius - marginMagnitude
      );
      drawRoundedRectPath({
        ctx,
        cx: centerX,
        cy: centerY,
        width: soldermaskPadWidth,
        height: soldermaskPadHeight,
        radius: soldermaskPadCornerRadius
      });
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "rotated_pill_hole_with_rect_pad") {
    const [centerX, centerY] = applyToPoint25(realToCanvasMat, [hole.x, hole.y]);
    const padWidth = hole.rect_pad_width * Math.abs(realToCanvasMat.a);
    const padHeight = hole.rect_pad_height * Math.abs(realToCanvasMat.a);
    const marginMagnitude = marginThickness * Math.abs(realToCanvasMat.a);
    const padCornerRadius = hole.rect_border_radius ? hole.rect_border_radius * Math.abs(realToCanvasMat.a) : 0;
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: centerX,
      cy: centerY,
      width: padWidth,
      height: padHeight,
      radius: padCornerRadius,
      ccwRotationDegrees: hole.rect_ccw_rotation
    });
    const soldermaskPadWidth = padWidth - marginMagnitude * 2;
    const soldermaskPadHeight = padHeight - marginMagnitude * 2;
    if (soldermaskPadWidth > 0 && soldermaskPadHeight > 0) {
      const soldermaskPadCornerRadius = Math.max(
        0,
        padCornerRadius - marginMagnitude
      );
      drawRoundedRectPath({
        ctx,
        cx: centerX,
        cy: centerY,
        width: soldermaskPadWidth,
        height: soldermaskPadHeight,
        radius: soldermaskPadCornerRadius,
        ccwRotationDegrees: hole.rect_ccw_rotation
      });
    }
    ctx.fill("evenodd");
  } else if (hole.shape === "hole_with_polygon_pad" && hole.pad_outline && hole.pad_outline.length >= 3) {
    const padPoints = hole.pad_outline.map(
      (point) => ({
        x: hole.x + point.x,
        y: hole.y + point.y
      })
    );
    ctx.beginPath();
    const outerPoints = padPoints.map((p) => {
      const [x, y] = applyToPoint25(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    drawPolygonPath({ ctx, points: outerPoints });
    const innerMarginPoints = offsetPolygonPoints(padPoints, -marginThickness);
    if (innerMarginPoints.length >= 3) {
      const innerPoints = innerMarginPoints.map((p) => {
        const [x, y] = applyToPoint25(realToCanvasMat, [p.x, p.y]);
        return { x, y };
      });
      drawPolygonPath({ ctx, points: innerPoints });
    }
    ctx.fill("evenodd");
  }
}

// lib/drawer/elements/pcb-soldermask/smt-pad.ts
import { applyToPoint as applyToPoint26 } from "transformation-matrix";
function processSmtPadSoldermask(params) {
  const { ctx, pad, realToCanvasMat, soldermaskOverCopperColor, layer } = params;
  if (pad.layer !== layer) return;
  const isCoveredWithSoldermask = pad.is_covered_with_solder_mask === true;
  const margin = pad.soldermask_margin ?? 0;
  let ml = margin;
  let mr = margin;
  let mt = margin;
  let mb = margin;
  if (pad.shape === "rect" || pad.shape === "rotated_rect") {
    ml = pad.soldermask_margin_left ?? margin;
    mr = pad.soldermask_margin_right ?? margin;
    mt = pad.soldermask_margin_top ?? margin;
    mb = pad.soldermask_margin_bottom ?? margin;
  }
  if (isCoveredWithSoldermask) {
    ctx.fillStyle = soldermaskOverCopperColor;
    drawPadShapePath({ ctx, pad, realToCanvasMat, ml: 0, mr: 0, mt: 0, mb: 0 });
    ctx.fill();
  } else if (ml < 0 || mr < 0 || mt < 0 || mb < 0) {
    drawPadShapePath({ ctx, pad, realToCanvasMat, ml, mr, mt, mb });
    cutPathFromSoldermask(ctx);
    drawNegativeMarginRingForPad({
      ctx,
      pad,
      realToCanvasMat,
      soldermaskOverCopperColor,
      ml,
      mr,
      mt,
      mb
    });
  } else if (ml > 0 || mr > 0 || mt > 0 || mb > 0) {
    drawPadShapePath({ ctx, pad, realToCanvasMat, ml, mr, mt, mb });
    cutPathFromSoldermask(ctx);
  } else {
    drawPadShapePath({ ctx, pad, realToCanvasMat, ml: 0, mr: 0, mt: 0, mb: 0 });
    cutPathFromSoldermask(ctx);
  }
}
function drawPadShapePath(params) {
  const { ctx, pad, realToCanvasMat, ml, mr, mt, mb } = params;
  const rotation = pad.shape === "rotated_rect" || pad.shape === "rotated_pill" ? pad.ccw_rotation ?? 0 : 0;
  if (pad.shape === "rect" || pad.shape === "rotated_rect") {
    const borderRadius = pad.corner_radius ?? pad.rect_border_radius ?? 0;
    const radians = rotation * Math.PI / 180;
    const dxLocal = (mr - ml) / 2;
    const dyLocal = (mt - mb) / 2;
    const dxGlobal = dxLocal * Math.cos(radians) - dyLocal * Math.sin(radians);
    const dyGlobal = dxLocal * Math.sin(radians) + dyLocal * Math.cos(radians);
    const [cx, cy] = applyToPoint26(realToCanvasMat, [
      pad.x + dxGlobal,
      pad.y + dyGlobal
    ]);
    const scaledWidth = (pad.width + ml + mr) * Math.abs(realToCanvasMat.a);
    const scaledHeight = (pad.height + mt + mb) * Math.abs(realToCanvasMat.a);
    const scaledRadius = borderRadius * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight,
      radius: scaledRadius
    });
    ctx.restore();
  } else if (pad.shape === "circle") {
    const avgMargin = (ml + mr + mt + mb) / 4;
    const [cx, cy] = applyToPoint26(realToCanvasMat, [pad.x, pad.y]);
    const scaledRadius = (pad.radius + avgMargin) * Math.abs(realToCanvasMat.a);
    ctx.beginPath();
    ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
    ctx.closePath();
  } else if (pad.shape === "pill" || pad.shape === "rotated_pill") {
    const avgMargin = (ml + mr) / 2;
    const [cx, cy] = applyToPoint26(realToCanvasMat, [pad.x, pad.y]);
    const scaledWidth = (pad.width + avgMargin * 2) * Math.abs(realToCanvasMat.a);
    const scaledHeight = (pad.height + avgMargin * 2) * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight
    });
    ctx.restore();
  } else if (pad.shape === "polygon" && pad.points && pad.points.length >= 3) {
    const avgMargin = (ml + mr + mt + mb) / 4;
    const points = avgMargin !== 0 ? offsetPolygonPoints(pad.points, avgMargin) : pad.points;
    const canvasPoints = points.map((p) => {
      const [x, y] = applyToPoint26(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    ctx.beginPath();
    drawPolygonPath({ ctx, points: canvasPoints });
  }
}
function drawNegativeMarginRingForPad(params) {
  const {
    ctx,
    pad,
    realToCanvasMat,
    soldermaskOverCopperColor,
    ml,
    mr,
    mt,
    mb
  } = params;
  const rotation = pad.shape === "rotated_rect" || pad.shape === "rotated_pill" ? pad.ccw_rotation ?? 0 : 0;
  const thicknessL = Math.max(0, -ml);
  const thicknessR = Math.max(0, -mr);
  const thicknessT = Math.max(0, -mt);
  const thicknessB = Math.max(0, -mb);
  ctx.fillStyle = soldermaskOverCopperColor;
  if (pad.shape === "rect" || pad.shape === "rotated_rect") {
    const borderRadius = pad.corner_radius ?? pad.rect_border_radius ?? 0;
    const [cx, cy] = applyToPoint26(realToCanvasMat, [pad.x, pad.y]);
    const scaledWidth = pad.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = pad.height * Math.abs(realToCanvasMat.a);
    const scaledRadius = borderRadius * Math.abs(realToCanvasMat.a);
    const scaledThicknessL = thicknessL * Math.abs(realToCanvasMat.a);
    const scaledThicknessR = thicknessR * Math.abs(realToCanvasMat.a);
    const scaledThicknessT = thicknessT * Math.abs(realToCanvasMat.a);
    const scaledThicknessB = thicknessB * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawRoundedRectPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight,
      radius: scaledRadius
    });
    const innerWidth = scaledWidth - scaledThicknessL - scaledThicknessR;
    const innerHeight = scaledHeight - scaledThicknessT - scaledThicknessB;
    const innerOffsetX = (scaledThicknessL - scaledThicknessR) / 2;
    const innerOffsetY = (scaledThicknessT - scaledThicknessB) / 2;
    const innerRadius = Math.max(
      0,
      scaledRadius - (scaledThicknessL + scaledThicknessR + scaledThicknessT + scaledThicknessB) / 4
    );
    if (innerWidth > 0 && innerHeight > 0) {
      drawRoundedRectPath({
        ctx,
        cx: innerOffsetX,
        cy: innerOffsetY,
        width: innerWidth,
        height: innerHeight,
        radius: innerRadius
      });
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (pad.shape === "circle") {
    const thickness = Math.max(thicknessL, thicknessR, thicknessT, thicknessB);
    const [cx, cy] = applyToPoint26(realToCanvasMat, [pad.x, pad.y]);
    const scaledRadius = pad.radius * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    const innerRadius = Math.max(0, scaledRadius - scaledThickness);
    ctx.save();
    ctx.beginPath();
    ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
    if (innerRadius > 0) {
      ctx.arc(cx, cy, innerRadius, 0, Math.PI * 2, true);
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (pad.shape === "pill" || pad.shape === "rotated_pill") {
    const thickness = Math.max(thicknessL, thicknessR, thicknessT, thicknessB);
    const [cx, cy] = applyToPoint26(realToCanvasMat, [pad.x, pad.y]);
    const scaledWidth = pad.width * Math.abs(realToCanvasMat.a);
    const scaledHeight = pad.height * Math.abs(realToCanvasMat.a);
    const scaledThickness = thickness * Math.abs(realToCanvasMat.a);
    ctx.save();
    ctx.translate(cx, cy);
    if (rotation !== 0) {
      ctx.rotate(-rotation * Math.PI / 180);
    }
    ctx.beginPath();
    drawPillPath({
      ctx,
      cx: 0,
      cy: 0,
      width: scaledWidth,
      height: scaledHeight
    });
    const innerWidth = scaledWidth - scaledThickness * 2;
    const innerHeight = scaledHeight - scaledThickness * 2;
    if (innerWidth > 0 && innerHeight > 0) {
      drawPillPath({
        ctx,
        cx: 0,
        cy: 0,
        width: innerWidth,
        height: innerHeight
      });
    }
    ctx.fill("evenodd");
    ctx.restore();
  } else if (pad.shape === "polygon" && pad.points && pad.points.length >= 3) {
    const thickness = Math.max(thicknessL, thicknessR, thicknessT, thicknessB);
    ctx.save();
    ctx.beginPath();
    const canvasPoints = pad.points.map((p) => {
      const [x, y] = applyToPoint26(realToCanvasMat, [p.x, p.y]);
      return { x, y };
    });
    drawPolygonPath({ ctx, points: canvasPoints });
    const innerPoints = offsetPolygonPoints(pad.points, -thickness);
    if (innerPoints.length >= 3) {
      const innerCanvasPoints = innerPoints.map((p) => {
        const [x, y] = applyToPoint26(realToCanvasMat, [p.x, p.y]);
        return { x, y };
      });
      drawPolygonPath({ ctx, points: innerCanvasPoints });
    }
    ctx.fill("evenodd");
    ctx.restore();
  }
}

// lib/drawer/elements/pcb-trace/normalize-trace-direction.ts
function normalizeTraceDirection(x, y) {
  const length = Math.hypot(x, y);
  if (length <= Number.EPSILON) return { x: 0, y: 0 };
  return { x: x / length, y: y / length };
}

// lib/drawer/elements/pcb-trace/get-trace-direction-at.ts
function getTraceDirectionAt(points, index) {
  const count = points.length;
  if (count < 2) return { x: 1, y: 0 };
  if (index <= 0) {
    const next2 = points[1];
    const current2 = points[0];
    if (!next2 || !current2) return { x: 1, y: 0 };
    const dir = normalizeTraceDirection(next2.x - current2.x, next2.y - current2.y);
    return dir.x === 0 && dir.y === 0 ? { x: 1, y: 0 } : dir;
  }
  if (index >= count - 1) {
    const prev2 = points[count - 2];
    const current2 = points[count - 1];
    if (!prev2 || !current2) return { x: 1, y: 0 };
    const dir = normalizeTraceDirection(current2.x - prev2.x, current2.y - prev2.y);
    return dir.x === 0 && dir.y === 0 ? { x: 1, y: 0 } : dir;
  }
  const prev = points[index - 1];
  const current = points[index];
  const next = points[index + 1];
  if (!prev || !current || !next) return { x: 1, y: 0 };
  const prevDir = normalizeTraceDirection(
    current.x - prev.x,
    current.y - prev.y
  );
  const nextDir = normalizeTraceDirection(
    next.x - current.x,
    next.y - current.y
  );
  const sum = normalizeTraceDirection(
    prevDir.x + nextDir.x,
    prevDir.y + nextDir.y
  );
  if (sum.x === 0 && sum.y === 0) {
    if (prevDir.x !== 0 || prevDir.y !== 0) return prevDir;
    if (nextDir.x !== 0 || nextDir.y !== 0) return nextDir;
    return { x: 1, y: 0 };
  }
  return sum;
}

// lib/drawer/elements/pcb-trace/build-trace-polygon.ts
function buildTracePolygon(points) {
  const left = [];
  const right = [];
  for (let i = 0; i < points.length; i++) {
    const point = points[i];
    if (!point) continue;
    const dir = getTraceDirectionAt(points, i);
    const normal = normalizeTraceDirection(-dir.y, dir.x);
    const offset = point.width / 2;
    left.push({
      x: point.x + normal.x * offset,
      y: point.y + normal.y * offset
    });
    right.push({
      x: point.x - normal.x * offset,
      y: point.y - normal.y * offset
    });
  }
  return left.concat(right.reverse());
}

// lib/drawer/elements/pcb-trace/collect-trace-segments.ts
function collectTraceSegments(route) {
  const segments = [];
  let current = [];
  let currentLayer = null;
  for (const routePoint of route ?? []) {
    if (!routePoint || routePoint.route_type !== "wire") {
      if (current.length >= 2) segments.push(current);
      current = [];
      currentLayer = null;
      continue;
    }
    const layer = routePoint.layer ?? currentLayer;
    if (!layer) continue;
    const x = routePoint.x;
    const y = routePoint.y;
    if (typeof x !== "number" || typeof y !== "number") continue;
    if (typeof routePoint.width !== "number") continue;
    const width = routePoint.width;
    if (currentLayer && layer !== currentLayer) {
      if (current.length >= 2) segments.push(current);
      current = [];
    }
    currentLayer = layer;
    current.push({
      route_type: "wire",
      x,
      y,
      width,
      layer,
      start_pcb_port_id: "start_pcb_port_id" in routePoint ? routePoint.start_pcb_port_id : void 0,
      end_pcb_port_id: "end_pcb_port_id" in routePoint ? routePoint.end_pcb_port_id : void 0
    });
  }
  if (current.length >= 2) segments.push(current);
  return segments;
}

// lib/drawer/elements/pcb-trace/cut-trace-destination-drills.ts
var COORD_EPSILON = 1e-6;
function cutTraceDestinationsAtDrills(params) {
  const { ctx, trace, realToCanvasMat, vias, platedHoles, layer } = params;
  if (!trace.route || trace.route.length < 2) return;
  const segments = collectTraceSegments(trace.route);
  if (segments.length === 0) return;
  const cutouts = /* @__PURE__ */ new Map();
  for (const segment of segments) {
    const segmentLayer = segment[0]?.layer;
    const destination = segment[segment.length - 1];
    if (!segmentLayer || !destination) continue;
    if (layer && segmentLayer !== layer) continue;
    for (const via of vias) {
      if (via.layers && Array.isArray(via.layers) && !via.layers.includes(segmentLayer)) {
        continue;
      }
      if (!isSamePoint(destination.x, destination.y, via.x, via.y)) continue;
      const radius = via.hole_diameter / 2;
      if (radius <= 0) continue;
      const cutout = {
        shape: "circle",
        x: via.x,
        y: via.y,
        radius
      };
      cutouts.set(getCutoutKey(cutout), cutout);
    }
    for (const platedHole of platedHoles) {
      if (platedHole.layers && Array.isArray(platedHole.layers) && !platedHole.layers.includes(segmentLayer)) {
        continue;
      }
      const cutout = getPlatedHoleDrillCutout(platedHole);
      if (!cutout) continue;
      if (!isSamePoint(destination.x, destination.y, cutout.x, cutout.y))
        continue;
      cutouts.set(getCutoutKey(cutout), cutout);
    }
  }
  if (cutouts.size === 0) return;
  ctx.save();
  ctx.globalCompositeOperation = "destination-out";
  for (const cutout of cutouts.values()) {
    if (cutout.shape === "circle") {
      drawCircle({
        ctx,
        center: { x: cutout.x, y: cutout.y },
        radius: cutout.radius,
        fill: "#000",
        realToCanvasMat
      });
      continue;
    }
    if (cutout.shape === "oval") {
      drawOval({
        ctx,
        center: { x: cutout.x, y: cutout.y },
        radius_x: cutout.radiusX,
        radius_y: cutout.radiusY,
        fill: "#000",
        realToCanvasMat,
        rotation: cutout.rotation
      });
      continue;
    }
    drawPill({
      ctx,
      center: { x: cutout.x, y: cutout.y },
      width: cutout.width,
      height: cutout.height,
      fill: "#000",
      realToCanvasMat,
      rotation: cutout.rotation
    });
  }
  ctx.restore();
}
function getPlatedHoleDrillCutout(hole) {
  if (hole.shape === "circle") {
    const radius = hole.hole_diameter / 2;
    if (radius <= 0) return;
    return {
      shape: "circle",
      x: hole.x,
      y: hole.y,
      radius
    };
  }
  if (hole.shape === "oval") {
    const radiusX = hole.hole_width / 2;
    const radiusY = hole.hole_height / 2;
    if (radiusX <= 0 || radiusY <= 0) return;
    return {
      shape: "oval",
      x: hole.x,
      y: hole.y,
      radiusX,
      radiusY,
      rotation: hole.ccw_rotation
    };
  }
  if (hole.shape === "pill") {
    if (hole.hole_width <= 0 || hole.hole_height <= 0) return;
    return {
      shape: "pill",
      x: hole.x,
      y: hole.y,
      width: hole.hole_width,
      height: hole.hole_height,
      rotation: hole.ccw_rotation
    };
  }
  if (hole.shape === "circular_hole_with_rect_pad") {
    const radius = hole.hole_diameter / 2;
    if (radius <= 0) return;
    return {
      shape: "circle",
      x: hole.x + (hole.hole_offset_x ?? 0),
      y: hole.y + (hole.hole_offset_y ?? 0),
      radius
    };
  }
  if (hole.shape === "pill_hole_with_rect_pad") {
    if (hole.hole_width <= 0 || hole.hole_height <= 0) return;
    return {
      shape: "pill",
      x: hole.x + (hole.hole_offset_x ?? 0),
      y: hole.y + (hole.hole_offset_y ?? 0),
      width: hole.hole_width,
      height: hole.hole_height
    };
  }
  if (hole.shape === "rotated_pill_hole_with_rect_pad") {
    if (hole.hole_width <= 0 || hole.hole_height <= 0) return;
    return {
      shape: "pill",
      x: hole.x + (hole.hole_offset_x ?? 0),
      y: hole.y + (hole.hole_offset_y ?? 0),
      width: hole.hole_width,
      height: hole.hole_height,
      rotation: hole.hole_ccw_rotation
    };
  }
  if (hole.shape !== "hole_with_polygon_pad") return;
  const holeX = hole.x + (hole.hole_offset_x ?? 0);
  const holeY = hole.y + (hole.hole_offset_y ?? 0);
  const holeShape = hole.hole_shape;
  if (holeShape === "circle") {
    const radius = (hole.hole_diameter ?? 0) / 2;
    if (radius <= 0) return;
    return {
      shape: "circle",
      x: holeX,
      y: holeY,
      radius
    };
  }
  if (holeShape === "oval") {
    const radiusX = (hole.hole_width ?? 0) / 2;
    const radiusY = (hole.hole_height ?? 0) / 2;
    if (radiusX <= 0 || radiusY <= 0) return;
    return {
      shape: "oval",
      x: holeX,
      y: holeY,
      radiusX,
      radiusY
    };
  }
  if (holeShape === "pill" || holeShape === "rotated_pill") {
    const width = hole.hole_width ?? 0;
    const height = hole.hole_height ?? 0;
    if (width <= 0 || height <= 0) return;
    return {
      shape: "pill",
      x: holeX,
      y: holeY,
      width,
      height,
      rotation: holeShape === "rotated_pill" ? hole.ccw_rotation : 0
    };
  }
}
function isSamePoint(x1, y1, x2, y2) {
  return Math.abs(x1 - x2) <= COORD_EPSILON && Math.abs(y1 - y2) <= COORD_EPSILON;
}
function getCutoutKey(cutout) {
  if (cutout.shape === "circle") {
    return `circle:${cutout.x}:${cutout.y}:${cutout.radius}`;
  }
  if (cutout.shape === "oval") {
    return `oval:${cutout.x}:${cutout.y}:${cutout.radiusX}:${cutout.radiusY}:${cutout.rotation ?? 0}`;
  }
  return `pill:${cutout.x}:${cutout.y}:${cutout.width}:${cutout.height}:${cutout.rotation ?? 0}`;
}

// lib/drawer/elements/pcb-trace/has-variable-width.ts
function hasVariableWidth(points) {
  if (points.length < 2) return false;
  const baseWidth = points[0]?.width;
  if (typeof baseWidth !== "number") return false;
  return points.some(
    (point) => Math.abs(point.width - baseWidth) > Number.EPSILON
  );
}

// lib/drawer/elements/pcb-soldermask/trace.ts
function processTraceSoldermask(params) {
  const { ctx, trace, realToCanvasMat, soldermaskOverCopperColor, layer } = params;
  if (!trace.route || !Array.isArray(trace.route) || trace.route.length < 2) {
    return;
  }
  const segments = collectTraceSegments(trace.route);
  for (const segment of segments) {
    const segmentLayer = segment[0]?.layer;
    if (segmentLayer !== layer) continue;
    if (hasVariableWidth(segment)) {
      const polygonPoints = buildTracePolygon(segment);
      drawPolygon({
        ctx,
        points: polygonPoints,
        fill: soldermaskOverCopperColor,
        realToCanvasMat
      });
      continue;
    }
    for (let i = 0; i < segment.length - 1; i++) {
      const start = segment[i];
      const end = segment[i + 1];
      if (!start || !end) continue;
      drawLine({
        ctx,
        start: { x: start.x, y: start.y },
        end: { x: end.x, y: end.y },
        strokeWidth: start.width,
        stroke: soldermaskOverCopperColor,
        realToCanvasMat,
        lineCap: "round"
      });
    }
  }
  cutTraceDestinationsAtDrills({
    ctx,
    trace,
    realToCanvasMat,
    layer,
    vias: params.vias,
    platedHoles: params.platedHoles
  });
}

// lib/drawer/elements/pcb-soldermask/via.ts
import { applyToPoint as applyToPoint27 } from "transformation-matrix";
function processViaSoldermask(params) {
  const { ctx, via, realToCanvasMat } = params;
  const [cx, cy] = applyToPoint27(realToCanvasMat, [via.x, via.y]);
  const scaledRadius = via.outer_diameter / 2 * Math.abs(realToCanvasMat.a);
  ctx.beginPath();
  ctx.arc(cx, cy, scaledRadius, 0, Math.PI * 2);
  ctx.closePath();
  cutPathFromSoldermask(ctx);
}

// lib/drawer/elements/pcb-soldermask/index.ts
function drawPcbSoldermask(params) {
  const { ctx, elements, realToCanvasMat, colorMap, layer, drawSoldermask } = params;
  if (!drawSoldermask) return;
  if (ctx.canvas.width <= 0 || ctx.canvas.height <= 0) return;
  const boards = elements.filter(
    (el) => el.type === "pcb_board"
  );
  const panel = elements.find((el) => el.type === "pcb_panel");
  if (boards.length === 0 && !panel) return;
  const soldermaskColor = colorMap.soldermask[layer] ?? colorMap.soldermask.top;
  const soldermaskOverCopperColor = colorMap.soldermaskOverCopper[layer] ?? colorMap.soldermaskOverCopper.top;
  const vias = elements.filter((el) => el.type === "pcb_via");
  const platedHoles = elements.filter(
    (el) => el.type === "pcb_plated_hole"
  );
  const soldermaskCtx = createSoldermaskLayerContext(ctx, ctx.canvas.width, ctx.canvas.height) ?? ctx;
  if (panel) {
    drawPanelSoldermask({
      ctx: soldermaskCtx,
      panel,
      realToCanvasMat,
      soldermaskColor
    });
  }
  if (panel && boards.length > 0) {
    soldermaskCtx.save();
    soldermaskCtx.globalCompositeOperation = "destination-out";
    for (const board of boards) {
      drawBoardSoldermask({
        ctx: soldermaskCtx,
        board,
        realToCanvasMat,
        soldermaskColor: "#000"
      });
    }
    soldermaskCtx.restore();
  }
  for (const board of boards) {
    drawBoardSoldermask({
      ctx: soldermaskCtx,
      board,
      realToCanvasMat,
      soldermaskColor
    });
  }
  for (const element of elements) {
    if (element.type === "pcb_trace") {
      processTraceSoldermask({
        ctx: soldermaskCtx,
        trace: element,
        realToCanvasMat,
        soldermaskOverCopperColor,
        layer,
        vias,
        platedHoles
      });
    } else if (element.type === "pcb_copper_pour") {
      processCopperPourSoldermask({
        ctx: soldermaskCtx,
        pour: element,
        realToCanvasMat,
        soldermaskOverCopperColor,
        layer,
        elements
      });
    }
  }
  for (const element of elements) {
    processElementSoldermask({
      ctx: soldermaskCtx,
      element,
      realToCanvasMat,
      soldermaskOverCopperColor,
      layer
    });
  }
  mergeSoldermaskLayer(ctx, soldermaskCtx);
}
function processElementSoldermask(params) {
  const { ctx, element, realToCanvasMat, soldermaskOverCopperColor, layer } = params;
  if (element.type === "pcb_smtpad") {
    processSmtPadSoldermask({
      ctx,
      pad: element,
      realToCanvasMat,
      soldermaskOverCopperColor,
      layer
    });
  } else if (element.type === "pcb_plated_hole") {
    processPlatedHoleSoldermask({
      ctx,
      hole: element,
      realToCanvasMat,
      soldermaskOverCopperColor,
      layer
    });
  } else if (element.type === "pcb_hole") {
    processHoleSoldermask({
      ctx,
      hole: element,
      realToCanvasMat,
      soldermaskOverCopperColor
    });
  } else if (element.type === "pcb_via") {
    processViaSoldermask({
      ctx,
      via: element,
      realToCanvasMat
    });
  } else if (element.type === "pcb_cutout") {
    processCutoutSoldermask({
      ctx,
      cutout: element,
      realToCanvasMat
    });
  } else if (element.type === "pcb_copper_pour") {
    return;
  } else if (element.type === "pcb_trace") {
    return;
  }
}

// lib/drawer/elements/pcb-trace/layer-to-color.ts
function layerToColor4(layer, colorMap) {
  return colorMap.copper[layer] ?? colorMap.copper.top;
}

// lib/drawer/elements/pcb-trace/pcb-trace.ts
function drawPcbTrace(params) {
  const { ctx, trace, realToCanvasMat, colorMap } = params;
  if (!trace.route || !Array.isArray(trace.route) || trace.route.length < 2) {
    return;
  }
  const segments = collectTraceSegments(trace.route);
  for (const segment of segments) {
    const layer = segment[0]?.layer;
    if (!layer) continue;
    const color = layerToColor4(layer, colorMap);
    if (hasVariableWidth(segment)) {
      const polygonPoints = buildTracePolygon(segment);
      drawPolygon({
        ctx,
        points: polygonPoints,
        fill: color,
        realToCanvasMat
      });
      continue;
    }
    for (let i = 0; i < segment.length - 1; i++) {
      const start = segment[i];
      const end = segment[i + 1];
      if (!start || !end) continue;
      drawLine({
        ctx,
        start: { x: start.x, y: start.y },
        end: { x: end.x, y: end.y },
        strokeWidth: start.width,
        stroke: color,
        realToCanvasMat,
        lineCap: "round"
      });
    }
  }
  cutTraceDestinationsAtDrills({
    ctx,
    trace,
    realToCanvasMat,
    vias: params.vias ?? [],
    platedHoles: params.platedHoles ?? []
  });
}

// lib/drawer/elements/pcb-via.ts
function drawPcbVia(params) {
  const { ctx, via, realToCanvasMat, colorMap } = params;
  drawCircle({
    ctx,
    center: { x: via.x, y: via.y },
    radius: via.outer_diameter / 2,
    fill: colorMap.copper.top,
    realToCanvasMat
  });
  ctx.save();
  ctx.globalCompositeOperation = "destination-out";
  drawCircle({
    ctx,
    center: { x: via.x, y: via.y },
    radius: via.hole_diameter / 2,
    fill: "#000",
    realToCanvasMat
  });
  ctx.restore();
  drawCircle({
    ctx,
    center: { x: via.x, y: via.y },
    radius: via.hole_diameter / 2,
    fill: colorMap.drill,
    realToCanvasMat
  });
}

// lib/drawer/pcb-render-layer-filter.ts
import { getElementRenderLayers } from "@tscircuit/circuit-json-util";
function shouldDrawElement(element, options) {
  if (!options.layers || options.layers.length === 0) {
    return true;
  }
  const elementLayers = getElementRenderLayers(element);
  if (elementLayers.length === 0) {
    return true;
  }
  return elementLayers.some((layer) => options.layers.includes(layer));
}

// lib/drawer/types.ts
var DEFAULT_PCB_COLOR_MAP = {
  copper: {
    top: "rgb(200, 52, 52)",
    inner1: "rgb(255, 140, 0)",
    inner2: "rgb(255, 215, 0)",
    inner3: "rgb(50, 205, 50)",
    inner4: "rgb(64, 224, 208)",
    inner5: "rgb(138, 43, 226)",
    inner6: "rgb(255, 105, 180)",
    bottom: "rgb(77, 127, 196)"
  },
  copperPour: {
    top: "rgb(230, 95, 95)",
    bottom: "rgb(122, 162, 222)"
  },
  soldermaskWithCopperUnderneath: {
    top: "rgb(18, 82, 50)",
    bottom: "rgb(77, 127, 196)"
  },
  soldermask: {
    top: "rgb(12, 55, 33)",
    bottom: "rgb(12, 55, 33)"
  },
  soldermaskOverCopper: {
    top: "rgb(52, 135, 73)",
    bottom: "rgb(52, 135, 73)"
  },
  substrate: "rgb(201, 162, 110)",
  drill: "#FF26E2",
  silkscreen: {
    top: "#f2eda1",
    bottom: "#5da9e9"
  },
  boardOutline: "rgba(255, 255, 255, 0.5)",
  courtyard: {
    top: "#FF00FF",
    bottom: "rgb(38, 233, 255)"
  },
  keepout: "#FF6B6B",
  // Red color for keepout zones
  fabricationNote: "rgba(255, 255, 255, 0.5)"
};

// lib/drawer/CircuitToCanvasDrawer.ts
var CircuitToCanvasDrawer = class {
  ctx;
  colorMap;
  realToCanvasMat;
  constructor(canvasOrContext) {
    if ("getContext" in canvasOrContext && typeof canvasOrContext.getContext === "function") {
      const ctx = canvasOrContext.getContext("2d");
      if (!ctx) {
        throw new Error("Failed to get 2D rendering context from canvas");
      }
      this.ctx = ctx;
    } else {
      this.ctx = canvasOrContext;
    }
    this.colorMap = { ...DEFAULT_PCB_COLOR_MAP };
    this.realToCanvasMat = identity();
  }
  configure(config) {
    if (config.colorOverrides) {
      this.colorMap = {
        ...this.colorMap,
        ...config.colorOverrides,
        copper: {
          ...this.colorMap.copper,
          ...config.colorOverrides.copper
        },
        copperPour: {
          ...this.colorMap.copperPour,
          ...config.colorOverrides.copperPour
        },
        silkscreen: {
          ...this.colorMap.silkscreen,
          ...config.colorOverrides.silkscreen
        },
        soldermask: {
          ...this.colorMap.soldermask,
          ...config.colorOverrides.soldermask
        },
        soldermaskWithCopperUnderneath: {
          ...this.colorMap.soldermaskWithCopperUnderneath,
          ...config.colorOverrides.soldermaskWithCopperUnderneath
        },
        soldermaskOverCopper: {
          ...this.colorMap.soldermaskOverCopper,
          ...config.colorOverrides.soldermaskOverCopper
        }
      };
    }
  }
  setCameraBounds(bounds) {
    const canvas = this.ctx.canvas;
    const canvasWidth = canvas.width;
    const canvasHeight = canvas.height;
    const realWidth = bounds.maxX - bounds.minX;
    const realHeight = bounds.maxY - bounds.minY;
    const scaleX = canvasWidth / realWidth;
    const scaleY = canvasHeight / realHeight;
    const uniformScale = Math.min(scaleX, scaleY);
    const offsetX = (canvasWidth - realWidth * uniformScale) / 2;
    const offsetY = (canvasHeight - realHeight * uniformScale) / 2;
    this.realToCanvasMat = compose3(
      translate3(offsetX, offsetY),
      scale(uniformScale, -uniformScale),
      translate3(-bounds.minX, -bounds.maxY)
    );
  }
  drawElements(elements, options = {}) {
    const board = elements.find((el) => el.type === "pcb_board");
    const panel = elements.find((el) => el.type === "pcb_panel");
    if (panel) {
      drawPcbPanelElement({
        ctx: this.ctx,
        panel,
        realToCanvasMat: this.realToCanvasMat,
        colorMap: this.colorMap,
        drawBoardMaterial: false
      });
    }
    const drawBoardMaterial = options.drawBoardMaterial ?? false;
    const drawSoldermask = options.drawSoldermask ?? false;
    const hasExplicitSoldermaskLayers = options.drawSoldermaskTop !== void 0 || options.drawSoldermaskBottom !== void 0;
    const renderTopSoldermask = drawSoldermask && (board !== void 0 || panel !== void 0) && (options.drawSoldermaskTop ?? !hasExplicitSoldermaskLayers);
    const renderBottomSoldermask = drawSoldermask && (board !== void 0 || panel !== void 0) && (options.drawSoldermaskBottom ?? false);
    const drawableVias = elements.filter(
      (el) => shouldDrawElement(el, options) && el.type === "pcb_via"
    );
    const drawablePlatedHoles = elements.filter(
      (el) => shouldDrawElement(el, options) && el.type === "pcb_plated_hole"
    );
    if (board) {
      drawPcbBoard({
        ctx: this.ctx,
        board,
        realToCanvasMat: this.realToCanvasMat,
        colorMap: this.colorMap,
        drawBoardMaterial,
        minBoardOutlineStrokePx: options.minBoardOutlineStrokePx
      });
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_copper_pour") {
        drawPcbCopperPour({
          ctx: this.ctx,
          pour: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_smtpad") {
        drawPcbSmtPad({
          ctx: this.ctx,
          pad: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_copper_text") {
        drawPcbCopperText({
          ctx: this.ctx,
          text: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
    if (renderTopSoldermask) {
      for (const element of elements) {
        if (!shouldDrawElement(element, options)) continue;
        if (element.type === "pcb_trace") {
          drawPcbTrace({
            ctx: this.ctx,
            trace: element,
            realToCanvasMat: this.realToCanvasMat,
            colorMap: this.colorMap,
            vias: drawableVias,
            platedHoles: drawablePlatedHoles
          });
        }
      }
      for (const element of elements) {
        if (!shouldDrawElement(element, options)) continue;
        if (element.type === "pcb_hole") {
          drawPcbHole({
            ctx: this.ctx,
            hole: element,
            realToCanvasMat: this.realToCanvasMat,
            colorMap: this.colorMap,
            soldermaskMargin: element.soldermask_margin,
            drawSoldermask: renderTopSoldermask
          });
        }
        if (element.type === "pcb_plated_hole") {
          drawPcbPlatedHole({
            ctx: this.ctx,
            hole: element,
            realToCanvasMat: this.realToCanvasMat,
            colorMap: this.colorMap,
            soldermaskMargin: element.soldermask_margin,
            drawSoldermask: renderTopSoldermask
          });
        }
        if (element.type === "pcb_via") {
          drawPcbVia({
            ctx: this.ctx,
            via: element,
            realToCanvasMat: this.realToCanvasMat,
            colorMap: this.colorMap
          });
        }
      }
    }
    if (renderTopSoldermask) {
      drawPcbSoldermask({
        ctx: this.ctx,
        elements,
        realToCanvasMat: this.realToCanvasMat,
        colorMap: this.colorMap,
        layer: "top",
        drawSoldermask: true
      });
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_silkscreen_text") {
        drawPcbSilkscreenText({
          ctx: this.ctx,
          text: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_rect") {
        drawPcbSilkscreenRect({
          ctx: this.ctx,
          rect: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_circle") {
        drawPcbSilkscreenCircle({
          ctx: this.ctx,
          circle: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_line") {
        drawPcbSilkscreenLine({
          ctx: this.ctx,
          line: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_path") {
        drawPcbSilkscreenPath({
          ctx: this.ctx,
          path: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_pill") {
        drawPcbSilkscreenPill({
          ctx: this.ctx,
          pill: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_silkscreen_oval") {
        drawPcbSilkscreenOval({
          ctx: this.ctx,
          oval: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_trace" && !renderTopSoldermask) {
        drawPcbTrace({
          ctx: this.ctx,
          trace: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap,
          vias: drawableVias,
          platedHoles: drawablePlatedHoles
        });
      }
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_hole" && !renderTopSoldermask) {
        drawPcbHole({
          ctx: this.ctx,
          hole: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap,
          soldermaskMargin: renderTopSoldermask ? element.soldermask_margin : void 0,
          drawSoldermask: renderTopSoldermask
        });
      }
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_plated_hole" && !renderTopSoldermask) {
        drawPcbPlatedHole({
          ctx: this.ctx,
          hole: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap,
          soldermaskMargin: renderTopSoldermask ? element.soldermask_margin : void 0,
          drawSoldermask: renderTopSoldermask
        });
      }
      if (element.type === "pcb_via" && !renderTopSoldermask) {
        drawPcbVia({
          ctx: this.ctx,
          via: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
    if (renderBottomSoldermask) {
      drawPcbSoldermask({
        ctx: this.ctx,
        elements,
        realToCanvasMat: this.realToCanvasMat,
        colorMap: this.colorMap,
        layer: "bottom",
        drawSoldermask: true
      });
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_cutout") {
        drawPcbCutout({
          ctx: this.ctx,
          cutout: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
    for (const element of elements) {
      if (!shouldDrawElement(element, options)) continue;
      if (element.type === "pcb_keepout") {
        drawPcbKeepout({
          ctx: this.ctx,
          keepout: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_fabrication_note_text") {
        drawPcbFabricationNoteText({
          ctx: this.ctx,
          text: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_fabrication_note_rect") {
        drawPcbFabricationNoteRect({
          ctx: this.ctx,
          rect: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      const showPcbNotes = options.showPcbNotes ?? true;
      if (showPcbNotes && element.type === "pcb_note_rect") {
        drawPcbNoteRect({
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap,
          ctx: this.ctx,
          rect: element
        });
      }
      if (element.type === "pcb_fabrication_note_path") {
        drawPcbFabricationNotePath({
          ctx: this.ctx,
          path: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (showPcbNotes && element.type === "pcb_note_path") {
        drawPcbNotePath({
          ctx: this.ctx,
          path: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (showPcbNotes && element.type === "pcb_note_text") {
        drawPcbNoteText({
          ctx: this.ctx,
          text: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (showPcbNotes && element.type === "pcb_note_line") {
        drawPcbNoteLine({
          ctx: this.ctx,
          line: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (showPcbNotes && element.type === "pcb_note_dimension") {
        drawPcbNoteDimension({
          ctx: this.ctx,
          pcbNoteDimension: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_fabrication_note_dimension") {
        drawPcbFabricationNoteDimension({
          ctx: this.ctx,
          pcbFabricationNoteDimension: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_courtyard_circle") {
        drawPcbCourtyardCircle({
          ctx: this.ctx,
          circle: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_courtyard_rect") {
        drawPcbCourtyardRect({
          ctx: this.ctx,
          rect: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
      if (element.type === "pcb_courtyard_outline") {
        drawPcbCourtyardOutline({
          ctx: this.ctx,
          outline: element,
          realToCanvasMat: this.realToCanvasMat,
          colorMap: this.colorMap
        });
      }
    }
  }
};

// lib/drawer/shapes/arrow.ts
function drawArrow(params) {
  const { ctx, x, y, angle, arrowSize, color, strokeWidth } = params;
  ctx.save();
  ctx.translate(x, y);
  ctx.rotate(angle);
  ctx.beginPath();
  ctx.moveTo(0, 0);
  ctx.lineTo(-arrowSize, -arrowSize / 2);
  ctx.lineTo(-arrowSize, arrowSize / 2);
  ctx.closePath();
  ctx.fillStyle = color;
  ctx.fill();
  ctx.restore();
}
export {
  CircuitToCanvasDrawer,
  DEFAULT_PCB_COLOR_MAP,
  drawArrow,
  drawCircle,
  drawDimensionLine,
  drawLine,
  drawOval,
  drawPath,
  drawPcbBoard,
  drawPcbCopperPour,
  drawPcbCopperText,
  drawPcbCourtyardCircle,
  drawPcbCutout,
  drawPcbFabricationNoteDimension,
  drawPcbFabricationNotePath,
  drawPcbFabricationNoteRect,
  drawPcbFabricationNoteText,
  drawPcbHole,
  drawPcbKeepout,
  drawPcbNoteDimension,
  drawPcbNotePath,
  drawPcbNoteRect,
  drawPcbNoteText,
  drawPcbPanelElement,
  drawPcbPlatedHole,
  drawPcbSilkscreenCircle,
  drawPcbSilkscreenLine,
  drawPcbSilkscreenOval,
  drawPcbSilkscreenPath,
  drawPcbSilkscreenPill,
  drawPcbSilkscreenRect,
  drawPcbSilkscreenText,
  drawPcbSmtPad,
  drawPcbTrace,
  drawPcbVia,
  drawPill,
  drawPolygon,
  drawRect,
  drawSoldermaskRingForCircle,
  drawSoldermaskRingForOval,
  drawSoldermaskRingForPill,
  drawSoldermaskRingForRect,
  drawText,
  getAlphabetLayout,
  getTextStartPosition,
  strokeAlphabetText
};
