// @tscircuit/math-utils/dist/index.js
import {
  areBoundsCompletelyInsidePolygon,
  areBoundsOverlappingPolygon,
  isPointInsideBounds,
  isPointInsidePolygon,
  isRectCompletelyInsidePolygon,
  isRectOverlappingPolygon
} from "./chunk-4V2EVVKE.js";
import {
  range
} from "./chunk-KY53E6CT.js";
import {
  pointToSegmentClosestPoint,
  segmentToBoundsMinDistance,
  segmentToBoxMinDistance,
  segmentToCircleMinDistance,
  segmentToSegmentMinDistance
} from "./chunk-Y7G4VXR7.js";
import {
  getUnitVectorFromDirection,
  getUnitVectorFromPointAToB
} from "./chunk-GIGMPRPV.js";
import {
  grid
} from "./chunk-U45EKA3R.js";
import {
  normalizeDegrees
} from "./chunk-P6GW5PWY.js";
import {
  distSq,
  midpoint,
  pointToBoundsDistance,
  pointToBoxDistance
} from "./chunk-5J3PCV4D.js";
import {
  distance,
  doSegmentsIntersect,
  doesLineIntersectLine,
  doesSegmentIntersectRect,
  getSegmentIntersection,
  onSegment,
  orientation,
  pointToSegmentDistance
} from "./chunk-EFLPMB4J.js";
import {
  clamp,
  computeDistanceBetweenBoxes,
  computeGapBetweenBoxes,
  computeManhattanDistanceBetweenBoxes,
  findNearestPointsBetweenBoxSets,
  getBoundingBox
} from "./chunk-DRDDWFOS.js";
import "./chunk-MTORG67J.js";
import {
  boundsAreaOverlap
} from "./chunk-YA3GC5BB.js";
import {
  boundsDistance
} from "./chunk-KVGAXIWH.js";
import {
  boundsIntersection
} from "./chunk-S7H7MIDL.js";
import {
  doBoundsOverlap
} from "./chunk-CA5ORSO4.js";
import "./chunk-GYQ2KZV6.js";
import {
  getBoundFromCenteredRect
} from "./chunk-MIHKSXJY.js";
import {
  getBoundsCenter
} from "./chunk-3WCUPG5S.js";
import {
  getBoundsFromPoints
} from "./chunk-5N7UJNVK.js";
export {
  areBoundsCompletelyInsidePolygon,
  areBoundsOverlappingPolygon,
  boundsAreaOverlap,
  boundsDistance,
  boundsIntersection,
  clamp,
  computeDistanceBetweenBoxes,
  computeGapBetweenBoxes,
  computeManhattanDistanceBetweenBoxes,
  distSq,
  distance,
  doBoundsOverlap,
  doSegmentsIntersect,
  doesLineIntersectLine,
  doesSegmentIntersectRect,
  findNearestPointsBetweenBoxSets,
  getBoundFromCenteredRect,
  getBoundingBox,
  getBoundsCenter,
  getBoundsFromPoints,
  getSegmentIntersection,
  getUnitVectorFromDirection,
  getUnitVectorFromPointAToB,
  grid,
  isPointInsideBounds,
  isPointInsidePolygon,
  isRectCompletelyInsidePolygon,
  isRectOverlappingPolygon,
  midpoint,
  normalizeDegrees,
  onSegment,
  orientation,
  pointToBoundsDistance,
  pointToBoxDistance,
  pointToSegmentClosestPoint,
  pointToSegmentDistance,
  range,
  segmentToBoundsMinDistance,
  segmentToBoxMinDistance,
  segmentToCircleMinDistance,
  segmentToSegmentMinDistance
};
