// @tscircuit/high-density-a01/dist/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[__getOwnPropNames(cb)[0]])((mod = { exports: {} }).exports, mod), mod.exports;
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// node_modules/is-buffer/index.js
var require_is_buffer = __commonJS({
  "node_modules/is-buffer/index.js"(exports, module) {
    "use strict";
    module.exports = function(obj) {
      return obj != null && (isBuffer(obj) || isSlowBuffer(obj) || !!obj._isBuffer);
    };
    function isBuffer(obj) {
      return !!obj.constructor && typeof obj.constructor.isBuffer === "function" && obj.constructor.isBuffer(obj);
    }
    function isSlowBuffer(obj) {
      return typeof obj.readFloatLE === "function" && typeof obj.slice === "function" && isBuffer(obj.slice(0, 0));
    }
  }
});

// node_modules/kind-of/index.js
var require_kind_of = __commonJS({
  "node_modules/kind-of/index.js"(exports, module) {
    "use strict";
    var isBuffer = require_is_buffer();
    var toString = Object.prototype.toString;
    module.exports = function kindOf(val) {
      if (typeof val === "undefined") {
        return "undefined";
      }
      if (val === null) {
        return "null";
      }
      if (val === true || val === false || val instanceof Boolean) {
        return "boolean";
      }
      if (typeof val === "string" || val instanceof String) {
        return "string";
      }
      if (typeof val === "number" || val instanceof Number) {
        return "number";
      }
      if (typeof val === "function" || val instanceof Function) {
        return "function";
      }
      if (typeof Array.isArray !== "undefined" && Array.isArray(val)) {
        return "array";
      }
      if (val instanceof RegExp) {
        return "regexp";
      }
      if (val instanceof Date) {
        return "date";
      }
      var type = toString.call(val);
      if (type === "[object RegExp]") {
        return "regexp";
      }
      if (type === "[object Date]") {
        return "date";
      }
      if (type === "[object Arguments]") {
        return "arguments";
      }
      if (type === "[object Error]") {
        return "error";
      }
      if (isBuffer(val)) {
        return "buffer";
      }
      if (type === "[object Set]") {
        return "set";
      }
      if (type === "[object WeakSet]") {
        return "weakset";
      }
      if (type === "[object Map]") {
        return "map";
      }
      if (type === "[object WeakMap]") {
        return "weakmap";
      }
      if (type === "[object Symbol]") {
        return "symbol";
      }
      if (type === "[object Int8Array]") {
        return "int8array";
      }
      if (type === "[object Uint8Array]") {
        return "uint8array";
      }
      if (type === "[object Uint8ClampedArray]") {
        return "uint8clampedarray";
      }
      if (type === "[object Int16Array]") {
        return "int16array";
      }
      if (type === "[object Uint16Array]") {
        return "uint16array";
      }
      if (type === "[object Int32Array]") {
        return "int32array";
      }
      if (type === "[object Uint32Array]") {
        return "uint32array";
      }
      if (type === "[object Float32Array]") {
        return "float32array";
      }
      if (type === "[object Float64Array]") {
        return "float64array";
      }
      return "object";
    };
  }
});

// node_modules/rename-keys/index.js
var require_rename_keys = __commonJS({
  "node_modules/rename-keys/index.js"(exports, module) {
    "use strict";
    (function() {
      "use strict";
      function rename2(obj, fn) {
        if (typeof fn !== "function") {
          return obj;
        }
        var res = {};
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            res[fn(key, obj[key]) || key] = obj[key];
          }
        }
        return res;
      }
      if (typeof module !== "undefined" && module.exports) {
        module.exports = rename2;
      } else {
        if (typeof define === "function" && define.amd) {
          define([], function() {
            return rename2;
          });
        } else {
          window.rename = rename2;
        }
      }
    })();
  }
});

// node_modules/deep-rename-keys/index.js
var require_deep_rename_keys = __commonJS({
  "node_modules/deep-rename-keys/index.js"(exports, module) {
    "use strict";
    var typeOf = require_kind_of();
    var rename2 = require_rename_keys();
    module.exports = function renameDeep(obj, cb) {
      var type = typeOf(obj);
      if (type !== "object" && type !== "array") {
        throw new Error("expected an object");
      }
      var res = [];
      if (type === "object") {
        obj = rename2(obj, cb);
        res = {};
      }
      for (var key in obj) {
        if (obj.hasOwnProperty(key)) {
          var val = obj[key];
          if (typeOf(val) === "object" || typeOf(val) === "array") {
            res[key] = renameDeep(val, cb);
          } else {
            res[key] = val;
          }
        }
      }
      return res;
    };
  }
});

// node_modules/eventemitter3/index.js
var require_eventemitter3 = __commonJS({
  "node_modules/eventemitter3/index.js"(exports, module) {
    "use strict";
    var has = Object.prototype.hasOwnProperty;
    var prefix = "~";
    function Events() {
    }
    if (Object.create) {
      Events.prototype = /* @__PURE__ */ Object.create(null);
      if (!new Events().__proto__) prefix = false;
    }
    function EE(fn, context, once) {
      this.fn = fn;
      this.context = context;
      this.once = once || false;
    }
    function EventEmitter() {
      this._events = new Events();
      this._eventsCount = 0;
    }
    EventEmitter.prototype.eventNames = function eventNames() {
      var names = [], events, name;
      if (this._eventsCount === 0) return names;
      for (name in events = this._events) {
        if (has.call(events, name)) names.push(prefix ? name.slice(1) : name);
      }
      if (Object.getOwnPropertySymbols) {
        return names.concat(Object.getOwnPropertySymbols(events));
      }
      return names;
    };
    EventEmitter.prototype.listeners = function listeners(event, exists) {
      var evt = prefix ? prefix + event : event, available = this._events[evt];
      if (exists) return !!available;
      if (!available) return [];
      if (available.fn) return [available.fn];
      for (var i = 0, l = available.length, ee = new Array(l); i < l; i++) {
        ee[i] = available[i].fn;
      }
      return ee;
    };
    EventEmitter.prototype.emit = function emit(event, a1, a2, a3, a4, a5) {
      var evt = prefix ? prefix + event : event;
      if (!this._events[evt]) return false;
      var listeners = this._events[evt], len = arguments.length, args, i;
      if (listeners.fn) {
        if (listeners.once) this.removeListener(event, listeners.fn, void 0, true);
        switch (len) {
          case 1:
            return listeners.fn.call(listeners.context), true;
          case 2:
            return listeners.fn.call(listeners.context, a1), true;
          case 3:
            return listeners.fn.call(listeners.context, a1, a2), true;
          case 4:
            return listeners.fn.call(listeners.context, a1, a2, a3), true;
          case 5:
            return listeners.fn.call(listeners.context, a1, a2, a3, a4), true;
          case 6:
            return listeners.fn.call(listeners.context, a1, a2, a3, a4, a5), true;
        }
        for (i = 1, args = new Array(len - 1); i < len; i++) {
          args[i - 1] = arguments[i];
        }
        listeners.fn.apply(listeners.context, args);
      } else {
        var length = listeners.length, j;
        for (i = 0; i < length; i++) {
          if (listeners[i].once) this.removeListener(event, listeners[i].fn, void 0, true);
          switch (len) {
            case 1:
              listeners[i].fn.call(listeners[i].context);
              break;
            case 2:
              listeners[i].fn.call(listeners[i].context, a1);
              break;
            case 3:
              listeners[i].fn.call(listeners[i].context, a1, a2);
              break;
            case 4:
              listeners[i].fn.call(listeners[i].context, a1, a2, a3);
              break;
            default:
              if (!args) for (j = 1, args = new Array(len - 1); j < len; j++) {
                args[j - 1] = arguments[j];
              }
              listeners[i].fn.apply(listeners[i].context, args);
          }
        }
      }
      return true;
    };
    EventEmitter.prototype.on = function on(event, fn, context) {
      var listener = new EE(fn, context || this), evt = prefix ? prefix + event : event;
      if (!this._events[evt]) this._events[evt] = listener, this._eventsCount++;
      else if (!this._events[evt].fn) this._events[evt].push(listener);
      else this._events[evt] = [this._events[evt], listener];
      return this;
    };
    EventEmitter.prototype.once = function once(event, fn, context) {
      var listener = new EE(fn, context || this, true), evt = prefix ? prefix + event : event;
      if (!this._events[evt]) this._events[evt] = listener, this._eventsCount++;
      else if (!this._events[evt].fn) this._events[evt].push(listener);
      else this._events[evt] = [this._events[evt], listener];
      return this;
    };
    EventEmitter.prototype.removeListener = function removeListener(event, fn, context, once) {
      var evt = prefix ? prefix + event : event;
      if (!this._events[evt]) return this;
      if (!fn) {
        if (--this._eventsCount === 0) this._events = new Events();
        else delete this._events[evt];
        return this;
      }
      var listeners = this._events[evt];
      if (listeners.fn) {
        if (listeners.fn === fn && (!once || listeners.once) && (!context || listeners.context === context)) {
          if (--this._eventsCount === 0) this._events = new Events();
          else delete this._events[evt];
        }
      } else {
        for (var i = 0, events = [], length = listeners.length; i < length; i++) {
          if (listeners[i].fn !== fn || once && !listeners[i].once || context && listeners[i].context !== context) {
            events.push(listeners[i]);
          }
        }
        if (events.length) this._events[evt] = events.length === 1 ? events[0] : events;
        else if (--this._eventsCount === 0) this._events = new Events();
        else delete this._events[evt];
      }
      return this;
    };
    EventEmitter.prototype.removeAllListeners = function removeAllListeners(event) {
      var evt;
      if (event) {
        evt = prefix ? prefix + event : event;
        if (this._events[evt]) {
          if (--this._eventsCount === 0) this._events = new Events();
          else delete this._events[evt];
        }
      } else {
        this._events = new Events();
        this._eventsCount = 0;
      }
      return this;
    };
    EventEmitter.prototype.off = EventEmitter.prototype.removeListener;
    EventEmitter.prototype.addListener = EventEmitter.prototype.on;
    EventEmitter.prototype.setMaxListeners = function setMaxListeners() {
      return this;
    };
    EventEmitter.prefixed = prefix;
    EventEmitter.EventEmitter = EventEmitter;
    if ("undefined" !== typeof module) {
      module.exports = EventEmitter;
    }
  }
});

// node_modules/xml-lexer/dist/lexer.js
var require_lexer = __commonJS({
  "node_modules/xml-lexer/dist/lexer.js"(exports, module) {
    "use strict";
    function _defineProperty(obj, key, value) {
      if (key in obj) {
        Object.defineProperty(obj, key, { value, enumerable: true, configurable: true, writable: true });
      } else {
        obj[key] = value;
      }
      return obj;
    }
    var EventEmitter = require_eventemitter3();
    var noop = function noop2() {
    };
    var State = {
      data: "state-data",
      cdata: "state-cdata",
      tagBegin: "state-tag-begin",
      tagName: "state-tag-name",
      tagEnd: "state-tag-end",
      attributeNameStart: "state-attribute-name-start",
      attributeName: "state-attribute-name",
      attributeNameEnd: "state-attribute-name-end",
      attributeValueBegin: "state-attribute-value-begin",
      attributeValue: "state-attribute-value"
    };
    var Action = {
      lt: "action-lt",
      gt: "action-gt",
      space: "action-space",
      equal: "action-equal",
      quote: "action-quote",
      slash: "action-slash",
      char: "action-char",
      error: "action-error"
    };
    var Type = {
      text: "text",
      openTag: "open-tag",
      closeTag: "close-tag",
      attributeName: "attribute-name",
      attributeValue: "attribute-value"
    };
    var charToAction = {
      " ": Action.space,
      "	": Action.space,
      "\n": Action.space,
      "\r": Action.space,
      "<": Action.lt,
      ">": Action.gt,
      '"': Action.quote,
      "'": Action.quote,
      "=": Action.equal,
      "/": Action.slash
    };
    var getAction = function getAction2(char) {
      return charToAction[char] || Action.char;
    };
    var create = function create2(options) {
      var _State$data, _State$tagBegin, _State$tagName, _State$tagEnd, _State$attributeNameS, _State$attributeName, _State$attributeNameE, _State$attributeValue, _State$attributeValue2, _lexer$stateMachine;
      options = Object.assign({ debug: false }, options);
      var lexer = new EventEmitter();
      var state = State.data;
      var data = "";
      var tagName = "";
      var attrName = "";
      var attrValue = "";
      var isClosing = "";
      var openingQuote = "";
      var emit = function emit2(type, value) {
        if (tagName[0] === "?" || tagName[0] === "!") {
          return;
        }
        var event = { type, value };
        if (options.debug) {
          console.log("emit:", event);
        }
        lexer.emit("data", event);
      };
      lexer.stateMachine = (_lexer$stateMachine = {}, _defineProperty(_lexer$stateMachine, State.data, (_State$data = {}, _defineProperty(_State$data, Action.lt, function() {
        if (data.trim()) {
          emit(Type.text, data);
        }
        tagName = "";
        isClosing = false;
        state = State.tagBegin;
      }), _defineProperty(_State$data, Action.char, function(char) {
        data += char;
      }), _State$data)), _defineProperty(_lexer$stateMachine, State.cdata, _defineProperty({}, Action.char, function(char) {
        data += char;
        if (data.substr(-3) === "]]>") {
          emit(Type.text, data.slice(0, -3));
          data = "";
          state = State.data;
        }
      })), _defineProperty(_lexer$stateMachine, State.tagBegin, (_State$tagBegin = {}, _defineProperty(_State$tagBegin, Action.space, noop), _defineProperty(_State$tagBegin, Action.char, function(char) {
        tagName = char;
        state = State.tagName;
      }), _defineProperty(_State$tagBegin, Action.slash, function() {
        tagName = "";
        isClosing = true;
      }), _State$tagBegin)), _defineProperty(_lexer$stateMachine, State.tagName, (_State$tagName = {}, _defineProperty(_State$tagName, Action.space, function() {
        if (isClosing) {
          state = State.tagEnd;
        } else {
          state = State.attributeNameStart;
          emit(Type.openTag, tagName);
        }
      }), _defineProperty(_State$tagName, Action.gt, function() {
        if (isClosing) {
          emit(Type.closeTag, tagName);
        } else {
          emit(Type.openTag, tagName);
        }
        data = "";
        state = State.data;
      }), _defineProperty(_State$tagName, Action.slash, function() {
        state = State.tagEnd;
        emit(Type.openTag, tagName);
      }), _defineProperty(_State$tagName, Action.char, function(char) {
        tagName += char;
        if (tagName === "![CDATA[") {
          state = State.cdata;
          data = "";
          tagName = "";
        }
      }), _State$tagName)), _defineProperty(_lexer$stateMachine, State.tagEnd, (_State$tagEnd = {}, _defineProperty(_State$tagEnd, Action.gt, function() {
        emit(Type.closeTag, tagName);
        data = "";
        state = State.data;
      }), _defineProperty(_State$tagEnd, Action.char, noop), _State$tagEnd)), _defineProperty(_lexer$stateMachine, State.attributeNameStart, (_State$attributeNameS = {}, _defineProperty(_State$attributeNameS, Action.char, function(char) {
        attrName = char;
        state = State.attributeName;
      }), _defineProperty(_State$attributeNameS, Action.gt, function() {
        data = "";
        state = State.data;
      }), _defineProperty(_State$attributeNameS, Action.space, noop), _defineProperty(_State$attributeNameS, Action.slash, function() {
        isClosing = true;
        state = State.tagEnd;
      }), _State$attributeNameS)), _defineProperty(_lexer$stateMachine, State.attributeName, (_State$attributeName = {}, _defineProperty(_State$attributeName, Action.space, function() {
        state = State.attributeNameEnd;
      }), _defineProperty(_State$attributeName, Action.equal, function() {
        emit(Type.attributeName, attrName);
        state = State.attributeValueBegin;
      }), _defineProperty(_State$attributeName, Action.gt, function() {
        attrValue = "";
        emit(Type.attributeName, attrName);
        emit(Type.attributeValue, attrValue);
        data = "";
        state = State.data;
      }), _defineProperty(_State$attributeName, Action.slash, function() {
        isClosing = true;
        attrValue = "";
        emit(Type.attributeName, attrName);
        emit(Type.attributeValue, attrValue);
        state = State.tagEnd;
      }), _defineProperty(_State$attributeName, Action.char, function(char) {
        attrName += char;
      }), _State$attributeName)), _defineProperty(_lexer$stateMachine, State.attributeNameEnd, (_State$attributeNameE = {}, _defineProperty(_State$attributeNameE, Action.space, noop), _defineProperty(_State$attributeNameE, Action.equal, function() {
        emit(Type.attributeName, attrName);
        state = State.attributeValueBegin;
      }), _defineProperty(_State$attributeNameE, Action.gt, function() {
        attrValue = "";
        emit(Type.attributeName, attrName);
        emit(Type.attributeValue, attrValue);
        data = "";
        state = State.data;
      }), _defineProperty(_State$attributeNameE, Action.char, function(char) {
        attrValue = "";
        emit(Type.attributeName, attrName);
        emit(Type.attributeValue, attrValue);
        attrName = char;
        state = State.attributeName;
      }), _State$attributeNameE)), _defineProperty(_lexer$stateMachine, State.attributeValueBegin, (_State$attributeValue = {}, _defineProperty(_State$attributeValue, Action.space, noop), _defineProperty(_State$attributeValue, Action.quote, function(char) {
        openingQuote = char;
        attrValue = "";
        state = State.attributeValue;
      }), _defineProperty(_State$attributeValue, Action.gt, function() {
        attrValue = "";
        emit(Type.attributeValue, attrValue);
        data = "";
        state = State.data;
      }), _defineProperty(_State$attributeValue, Action.char, function(char) {
        openingQuote = "";
        attrValue = char;
        state = State.attributeValue;
      }), _State$attributeValue)), _defineProperty(_lexer$stateMachine, State.attributeValue, (_State$attributeValue2 = {}, _defineProperty(_State$attributeValue2, Action.space, function(char) {
        if (openingQuote) {
          attrValue += char;
        } else {
          emit(Type.attributeValue, attrValue);
          state = State.attributeNameStart;
        }
      }), _defineProperty(_State$attributeValue2, Action.quote, function(char) {
        if (openingQuote === char) {
          emit(Type.attributeValue, attrValue);
          state = State.attributeNameStart;
        } else {
          attrValue += char;
        }
      }), _defineProperty(_State$attributeValue2, Action.gt, function(char) {
        if (openingQuote) {
          attrValue += char;
        } else {
          emit(Type.attributeValue, attrValue);
          data = "";
          state = State.data;
        }
      }), _defineProperty(_State$attributeValue2, Action.slash, function(char) {
        if (openingQuote) {
          attrValue += char;
        } else {
          emit(Type.attributeValue, attrValue);
          isClosing = true;
          state = State.tagEnd;
        }
      }), _defineProperty(_State$attributeValue2, Action.char, function(char) {
        attrValue += char;
      }), _State$attributeValue2)), _lexer$stateMachine);
      var step = function step2(char) {
        if (options.debug) {
          console.log(state, char);
        }
        var actions = lexer.stateMachine[state];
        var action = actions[getAction(char)] || actions[Action.error] || actions[Action.char];
        action(char);
      };
      lexer.write = function(str) {
        var len = str.length;
        for (var i = 0; i < len; i++) {
          step(str[i]);
        }
      };
      return lexer;
    };
    module.exports = {
      State,
      Action,
      Type,
      create
    };
  }
});

// node_modules/xml-reader/dist/reader.js
var require_reader = __commonJS({
  "node_modules/xml-reader/dist/reader.js"(exports, module) {
    "use strict";
    var EventEmitter = require_eventemitter3();
    var Lexer = require_lexer();
    var Type = Lexer.Type;
    var NodeType = {
      element: "element",
      text: "text"
    };
    var createNode = function createNode2(params) {
      return Object.assign({
        name: "",
        type: NodeType.element,
        value: "",
        parent: null,
        attributes: {},
        children: []
      }, params);
    };
    var create = function create2(options) {
      options = Object.assign({
        stream: false,
        parentNodes: true,
        doneEvent: "done",
        tagPrefix: "tag:",
        emitTopLevelOnly: false,
        debug: false
      }, options);
      var lexer = void 0, rootNode = void 0, current = void 0, attrName = void 0;
      var reader = new EventEmitter();
      var handleLexerData = function handleLexerData2(data) {
        switch (data.type) {
          case Type.openTag:
            if (current === null) {
              current = rootNode;
              current.name = data.value;
            } else {
              var node = createNode({
                name: data.value,
                parent: current
              });
              current.children.push(node);
              current = node;
            }
            break;
          case Type.closeTag:
            var parent = current.parent;
            if (!options.parentNodes) {
              current.parent = null;
            }
            if (current.name !== data.value) {
              break;
            }
            if (options.stream && parent === rootNode) {
              rootNode.children = [];
              current.parent = null;
            }
            if (!options.emitTopLevelOnly || parent === rootNode) {
              reader.emit(options.tagPrefix + current.name, current);
              reader.emit("tag", current.name, current);
            }
            if (current === rootNode) {
              lexer.removeAllListeners("data");
              reader.emit(options.doneEvent, current);
              rootNode = null;
            }
            current = parent;
            break;
          case Type.text:
            if (current) {
              current.children.push(createNode({
                type: NodeType.text,
                value: data.value,
                parent: options.parentNodes ? current : null
              }));
            }
            break;
          case Type.attributeName:
            attrName = data.value;
            current.attributes[attrName] = "";
            break;
          case Type.attributeValue:
            current.attributes[attrName] = data.value;
            break;
        }
      };
      reader.reset = function() {
        lexer = Lexer.create({ debug: options.debug });
        lexer.on("data", handleLexerData);
        rootNode = createNode();
        current = null;
        attrName = "";
        reader.parse = lexer.write;
      };
      reader.reset();
      return reader;
    };
    var parseSync2 = function parseSync3(xml, options) {
      options = Object.assign({}, options, { stream: false, tagPrefix: ":" });
      var reader = create(options);
      var res = void 0;
      reader.on("done", function(ast) {
        res = ast;
      });
      reader.parse(xml);
      return res;
    };
    module.exports = {
      parseSync: parseSync2,
      create,
      NodeType
    };
  }
});

// lib/dataset02/convertDataset02SampleToNodeWithPortPoints.ts
var convertDataset02SampleToNodeWithPortPoints = (sample, options = {}) => {
  const borderPaddingMm = options.borderPaddingMm ?? 0;
  const availableZ = options.availableZ ?? [0, 1];
  const regionById = new Map(
    sample.connectionRegions.map((region) => [region.regionId, region])
  );
  let minX = Number.POSITIVE_INFINITY;
  let maxX = Number.NEGATIVE_INFINITY;
  let minY = Number.POSITIVE_INFINITY;
  let maxY = Number.NEGATIVE_INFINITY;
  for (const connection of sample.connections) {
    const startRegion = regionById.get(connection.startRegionId);
    const endRegion = regionById.get(connection.endRegionId);
    if (!startRegion || !endRegion) {
      throw new Error(
        `Missing region for connection ${connection.connectionId}: ${connection.startRegionId} -> ${connection.endRegionId}`
      );
    }
    minX = Math.min(minX, startRegion.d.center.x, endRegion.d.center.x);
    maxX = Math.max(maxX, startRegion.d.center.x, endRegion.d.center.x);
    minY = Math.min(minY, startRegion.d.center.y, endRegion.d.center.y);
    maxY = Math.max(maxY, startRegion.d.center.y, endRegion.d.center.y);
  }
  const width = maxX - minX + borderPaddingMm * 2;
  const height = maxY - minY + borderPaddingMm * 2;
  const center = {
    x: (minX + maxX) / 2,
    y: (minY + maxY) / 2
  };
  const portPoints = sample.connections.flatMap((connection) => {
    const startRegion = regionById.get(connection.startRegionId);
    const endRegion = regionById.get(connection.endRegionId);
    if (!startRegion || !endRegion) {
      throw new Error(
        `Missing region for connection ${connection.connectionId}: ${connection.startRegionId} -> ${connection.endRegionId}`
      );
    }
    return [
      {
        connectionName: connection.connectionId,
        x: startRegion.d.center.x,
        y: startRegion.d.center.y,
        z: availableZ[0] ?? 0
      },
      {
        connectionName: connection.connectionId,
        x: endRegion.d.center.x,
        y: endRegion.d.center.y,
        z: availableZ[0] ?? 0
      }
    ];
  });
  return {
    capacityMeshNodeId: options.capacityMeshNodeId ?? `dataset02-${sample.config.seed}-${sample.config.rows}x${sample.config.cols}`,
    center,
    width,
    height,
    portPoints,
    availableZ
  };
};

// lib/gridToAffineTransform.ts
function computeGridToAffineTransform(params) {
  const { originX, originY, rows, cols, cellSizeMm, width, height } = params;
  let a;
  let c;
  if (cols > 1) {
    a = width / ((cols - 1) * cellSizeMm);
    c = originX * (1 - a) - 0.5 * cellSizeMm * a;
  } else {
    a = 1;
    c = originX + width / 2 - (originX + 0.5 * cellSizeMm);
  }
  let e;
  let f;
  if (rows > 1) {
    e = height / ((rows - 1) * cellSizeMm);
    f = originY * (1 - e) - 0.5 * cellSizeMm * e;
  } else {
    e = 1;
    f = originY + height / 2 - (originY + 0.5 * cellSizeMm);
  }
  return { a, b: 0, c, d: 0, e, f };
}
function applyAffineTransformToPoint(t, p) {
  return {
    x: t.a * p.x + t.b * p.y + t.c,
    y: t.d * p.x + t.e * p.y + t.f
  };
}

// node_modules/transformation-matrix/src/rotate.js
var { cos, sin, PI } = Math;

// node_modules/transformation-matrix/src/skew.js
var { tan } = Math;

// node_modules/svgson/dist/svgson.esm.js
var import_deep_rename_keys = __toESM(require_deep_rename_keys());
var import_xml_reader = __toESM(require_reader());

// node_modules/@tscircuit/solver-utils/dist/index.js
var BaseSolver = class {
  MAX_ITERATIONS = 1e5;
  solved = false;
  failed = false;
  iterations = 0;
  progress = 0;
  error = null;
  activeSubSolver;
  failedSubSolvers;
  timeToSolve;
  stats = {};
  _setupDone = false;
  getSolverName() {
    return this.constructor.name;
  }
  setup() {
    if (this._setupDone) return;
    this._setup();
    this._setupDone = true;
  }
  /** Override this method to perform setup logic */
  _setup() {
  }
  /** DO NOT OVERRIDE! Override _step() instead */
  step() {
    if (!this._setupDone) {
      this.setup();
    }
    if (this.solved) return;
    if (this.failed) return;
    this.iterations++;
    try {
      this._step();
    } catch (e) {
      this.error = `${this.getSolverName()} error: ${e}`;
      this.failed = true;
      throw e;
    }
    if (!this.solved && this.iterations >= this.MAX_ITERATIONS) {
      this.tryFinalAcceptance();
    }
    if (!this.solved && this.iterations >= this.MAX_ITERATIONS) {
      this.error = `${this.getSolverName()} ran out of iterations`;
      this.failed = true;
    }
    if ("computeProgress" in this) {
      this.progress = this.computeProgress();
    }
  }
  /** Override this method to implement solver logic */
  _step() {
  }
  getConstructorParams() {
    throw new Error("getConstructorParams not implemented");
  }
  /**
   * Override this method to return the standardized output of the solver.
   * This method should only be called after the solver has completed successfully.
   * Returns null by default - solvers with outputs should override this method.
   */
  getOutput() {
    return null;
  }
  solve() {
    const startTime = Date.now();
    while (!this.solved && !this.failed) {
      this.step();
    }
    const endTime = Date.now();
    this.timeToSolve = endTime - startTime;
  }
  visualize() {
    return {
      lines: [],
      points: [],
      rects: [],
      circles: []
    };
  }
  /**
   * Called when the solver is about to fail, but we want to see if we have an
   * "acceptable" or "passable" solution. Mostly used for optimizers that
   * have an aggressive early stopping criterion.
   */
  tryFinalAcceptance() {
  }
  /**
   * A lightweight version of the visualize method that can be used to stream
   * progress
   */
  preview() {
    return {
      lines: [],
      points: [],
      rects: [],
      circles: []
    };
  }
};

// lib/HighDensitySolverA01/HighDensitySolverA01.ts
function rippedContains(r, id) {
  for (let cur = r; cur; cur = cur.prev) if (cur.id === id) return true;
  return false;
}
var MinHeap = class {
  f = [];
  seq = [];
  id = [];
  n = 0;
  push(f, seq, id) {
    let i = this.n++;
    this.f[i] = f;
    this.seq[i] = seq;
    this.id[i] = id;
    while (i > 0) {
      const p = i - 1 >> 1;
      if (this.less(p, i)) break;
      this.swap(i, p);
      i = p;
    }
  }
  pop() {
    const out = this.id[0];
    this.n--;
    if (this.n > 0) {
      this.f[0] = this.f[this.n];
      this.seq[0] = this.seq[this.n];
      this.id[0] = this.id[this.n];
      this.siftDown(0);
    }
    return out;
  }
  get size() {
    return this.n;
  }
  clear() {
    this.n = 0;
  }
  siftDown(i) {
    while (true) {
      const l = i * 2 + 1;
      const r = l + 1;
      if (l >= this.n) return;
      let m = l;
      if (r < this.n && !this.less(l, r)) m = r;
      if (this.less(i, m)) return;
      this.swap(i, m);
      i = m;
    }
  }
  less(i, j) {
    const fi = this.f[i];
    const fj = this.f[j];
    if (fi !== fj) return fi < fj;
    return this.seq[i] < this.seq[j];
  }
  swap(i, j) {
    const tmpF = this.f[i];
    this.f[i] = this.f[j];
    this.f[j] = tmpF;
    const tmpS = this.seq[i];
    this.seq[i] = this.seq[j];
    this.seq[j] = tmpS;
    const tmpI = this.id[i];
    this.id[i] = this.id[j];
    this.id[j] = tmpI;
  }
};
function toRootNetName(connectionName, rootConnectionName) {
  return rootConnectionName ?? connectionName.replace(/_mst\d+$/, "");
}
var DIRS_DR = [-1, -1, -1, 0, 0, 1, 1, 1];
var DIRS_DC = [-1, 0, 1, -1, 1, -1, 0, 1];
var HighDensitySolverA01 = class extends BaseSolver {
  nodeWithPortPoints;
  cellSizeMm;
  viaDiameter;
  traceThickness;
  traceMargin;
  viaMinDistFromBorder;
  showPenaltyMap;
  showUsedCellMap;
  hyperParameters;
  initialPenaltyFn;
  // Grid dimensions
  rows;
  cols;
  layers;
  gridOrigin;
  gridToBoundsTransform;
  // Z-layer mapping
  availableZ;
  zToLayer;
  layerToZ;
  // --- Interned connections ---
  connNameToId;
  connIdToName;
  connIdToRootNet;
  overlapFriendlyRootNets;
  // --- Flat arrays ---
  planeSize;
  // rows * cols
  usedCellsFlat;
  // layers * planeSize; -1 = empty
  portOwnerFlat;
  // layers * planeSize; -1 = none, -2 = shared
  usedDiagFlat;
  // layers * (rows-1) * (cols-1) * 2; -1 = empty
  penalty2d;
  // planeSize
  visitedStamp;
  // layers * planeSize
  sharedCrossRootPortCells;
  stamp = 0;
  // --- Precomputed via footprint offsets ---
  viaOffsetsDr;
  viaOffsetsDc;
  viaOffsetsLen;
  // --- Via zone boundaries (cell coordinates) ---
  minViaRow;
  maxViaRow;
  minViaCol;
  maxViaCol;
  // --- Per-connection used-cell tracking ---
  usedIndicesByConn;
  // connId -> [flatCellIdx, ...]
  usedDiagIndicesByConn;
  // connId -> [flatDiagIdx, ...]
  // --- Connection queues ---
  unsolvedSegs;
  solvedRoutes;
  // --- A* state ---
  activeConnSeg = null;
  activeConnId = -1;
  crossLayerSearch = false;
  nodePool;
  heap;
  seqCounter = 0;
  // --- Reusable scratch for via occupant scan ---
  _viaOccs = [];
  // --- Convergence state ---
  ripCount;
  totalRipEvents = 0;
  searchIterations = 0;
  consecutiveSkips = 0;
  penaltyCap;
  // --- Reusable scratch for computeMoveCostAndRips ---
  _moveCost = 0;
  _moveRipped = null;
  // --- Test/debug compatibility getters ---
  get unsolvedConnections() {
    return this.unsolvedSegs;
  }
  get solvedConnectionsMap() {
    return this.solvedRoutes;
  }
  get activeConnection() {
    if (!this.activeConnSeg) return null;
    const s = this.activeConnSeg;
    return {
      connectionName: this.connIdToName[s.connId] ?? "",
      start: { row: s.startRow, col: s.startCol, z: s.startZ, x: 0, y: 0 },
      end: { row: s.endRow, col: s.endCol, z: s.endZ, x: 0, y: 0 }
    };
  }
  get openSet() {
    return { length: this.heap?.size ?? 0 };
  }
  constructor(props) {
    super();
    this.nodeWithPortPoints = props.nodeWithPortPoints;
    this.cellSizeMm = props.cellSizeMm;
    this.viaDiameter = props.viaDiameter;
    this.traceThickness = props.traceThickness ?? 0.1;
    this.traceMargin = props.traceMargin ?? 0.15;
    this.viaMinDistFromBorder = props.viaMinDistFromBorder ?? 0.15;
    this.showPenaltyMap = props.showPenaltyMap ?? false;
    this.showUsedCellMap = props.showUsedCellMap ?? false;
    this.hyperParameters = {
      shuffleSeed: 0,
      ripCost: 10,
      ripTracePenalty: 0.5,
      ripViaPenalty: 0.75,
      viaBaseCost: 0.1,
      greedyMultiplier: 1.5,
      ...props.hyperParameters
    };
    this.MAX_ITERATIONS = 1e8;
    this.initialPenaltyFn = props.initialPenaltyFn;
  }
  _setup() {
    const { nodeWithPortPoints, cellSizeMm } = this;
    const { width, height, center } = nodeWithPortPoints;
    this.availableZ = nodeWithPortPoints.availableZ ?? [...new Set(nodeWithPortPoints.portPoints.map((pp) => pp.z))].sort(
      (a, b) => a - b
    );
    this.zToLayer = /* @__PURE__ */ new Map();
    this.layerToZ = /* @__PURE__ */ new Map();
    for (let i = 0; i < this.availableZ.length; i++) {
      const z = this.availableZ[i];
      this.zToLayer.set(z, i);
      this.layerToZ.set(i, z);
    }
    this.rows = Math.floor(height / cellSizeMm);
    this.cols = Math.floor(width / cellSizeMm);
    this.layers = this.availableZ.length;
    this.planeSize = this.rows * this.cols;
    const totalCells = this.layers * this.planeSize;
    const totalDiags = this.layers * Math.max(0, this.rows - 1) * Math.max(0, this.cols - 1) * 2;
    this.gridOrigin = {
      x: center.x - width / 2,
      y: center.y - height / 2
    };
    this.gridToBoundsTransform = computeGridToAffineTransform({
      originX: this.gridOrigin.x,
      originY: this.gridOrigin.y,
      rows: this.rows,
      cols: this.cols,
      cellSizeMm,
      width,
      height
    });
    this.connNameToId = /* @__PURE__ */ new Map();
    this.connIdToName = [];
    this.connIdToRootNet = [];
    this.overlapFriendlyRootNets = /* @__PURE__ */ new Set();
    this.penalty2d = new Float64Array(this.planeSize);
    if (this.initialPenaltyFn) {
      for (let row = 0; row < this.rows; row++) {
        const rowBase = row * this.cols;
        for (let col = 0; col < this.cols; col++) {
          const x = this.gridOrigin.x + (col + 0.5) * cellSizeMm;
          const y = this.gridOrigin.y + (row + 0.5) * cellSizeMm;
          const px = (col + 0.5) / this.cols;
          const py = (row + 0.5) / this.rows;
          this.penalty2d[rowBase + col] = this.initialPenaltyFn({
            x,
            y,
            px,
            py,
            row,
            col
          });
        }
      }
    }
    this.usedCellsFlat = new Int32Array(totalCells).fill(-1);
    this.portOwnerFlat = new Int32Array(totalCells).fill(-1);
    this.usedDiagFlat = new Int32Array(totalDiags).fill(-1);
    this.visitedStamp = new Uint32Array(totalCells);
    this.stamp = 0;
    const viaRadiusCells = Math.ceil(this.viaDiameter / 2 / cellSizeMm);
    const r2 = viaRadiusCells * viaRadiusCells;
    const drList = [];
    const dcList = [];
    for (let dr = -viaRadiusCells; dr <= viaRadiusCells; dr++) {
      for (let dc = -viaRadiusCells; dc <= viaRadiusCells; dc++) {
        if (dr * dr + dc * dc <= r2) {
          drList.push(dr);
          dcList.push(dc);
        }
      }
    }
    this.viaOffsetsLen = drList.length;
    this.viaOffsetsDr = new Int32Array(drList);
    this.viaOffsetsDc = new Int32Array(dcList);
    if (this.viaMinDistFromBorder > 0) {
      const borderCells = Math.ceil(this.viaMinDistFromBorder / cellSizeMm);
      this.minViaRow = borderCells;
      this.maxViaRow = this.rows - 1 - borderCells;
      this.minViaCol = borderCells;
      this.maxViaCol = this.cols - 1 - borderCells;
    } else {
      this.minViaRow = 0;
      this.maxViaRow = this.rows - 1;
      this.minViaCol = 0;
      this.maxViaCol = this.cols - 1;
    }
    this.unsolvedSegs = this.buildConnectionSegs();
    this.sharedCrossRootPortCells = /* @__PURE__ */ new Set();
    const rootByPortFlat = /* @__PURE__ */ new Map();
    for (const pp of this.nodeWithPortPoints.portPoints) {
      const connId = this.connNameToId.get(pp.connectionName);
      if (connId === void 0) continue;
      const cell = this.pointToCell(pp);
      const flatIdx = (cell.z * this.rows + cell.row) * this.cols + cell.col;
      const rootNet = this.connIdToRootNet[connId];
      const existingRoot = rootByPortFlat.get(flatIdx);
      if (existingRoot === void 0) {
        rootByPortFlat.set(flatIdx, rootNet);
      } else if (existingRoot !== rootNet) {
        this.sharedCrossRootPortCells.add(flatIdx);
      }
      const existing = this.portOwnerFlat[flatIdx];
      if (existing === -1 || existing === connId) {
        this.portOwnerFlat[flatIdx] = connId;
      } else {
        this.portOwnerFlat[flatIdx] = -2;
      }
    }
    this.solvedRoutes = /* @__PURE__ */ new Map();
    this.usedIndicesByConn = [];
    this.usedDiagIndicesByConn = [];
    this.ripCount = [];
    this.consecutiveSkips = 0;
    this.penaltyCap = this.hyperParameters.ripCost * 0.5;
    this.shuffleConnections();
    this.activeConnSeg = null;
    this.activeConnId = -1;
    this.nodePool = [];
    this.heap = new MinHeap();
    this.seqCounter = 0;
  }
  _step() {
    if (!this.activeConnSeg) {
      if (this.unsolvedSegs.length === 0) {
        this.solved = true;
        return;
      }
      const next = this.unsolvedSegs.shift();
      this.activeConnSeg = next;
      this.activeConnId = next.connId;
      this.crossLayerSearch = next.startZ !== next.endZ;
      this.nodePool = [];
      this.heap.clear();
      this.seqCounter = 0;
      this.searchIterations = 0;
      this.nextStamp();
      const h = this.computeH(
        next.startZ,
        next.startRow,
        next.startCol,
        next.endZ,
        next.endRow,
        next.endCol
      );
      const f = h * this.hyperParameters.greedyMultiplier;
      this.nodePool.push({
        z: next.startZ,
        row: next.startRow,
        col: next.startCol,
        g: 0,
        f,
        parentIdx: -1,
        ripped: null
      });
      this.heap.push(f, this.seqCounter++, 0);
      return;
    }
    this.searchIterations++;
    const connRips = this.ripCount[this.activeConnId] ?? 0;
    const baseBudget = this.planeSize * this.layers * 50;
    const budget = Math.min(
      baseBudget * (1 + connRips * 0.5),
      this.planeSize * this.layers * 500
    );
    if (this.searchIterations > budget) {
      const pen = this.penalty2d;
      for (let i = 0; i < pen.length; i++) {
        pen[i] = pen[i] * 0.9;
      }
      this.unsolvedSegs.push(this.activeConnSeg);
      this.activeConnSeg = null;
      this.activeConnId = -1;
      this.heap.clear();
      this.nodePool = [];
      this.consecutiveSkips++;
      if (this.consecutiveSkips >= this.unsolvedSegs.length * 3) {
        this.error = `Convergence failure: ${this.unsolvedSegs.length} connections stuck`;
        this.failed = true;
      }
      return;
    }
    if (this.heap.size === 0) {
      this.error = `No path found for ${this.connIdToName[this.activeConnId]}`;
      this.failed = true;
      return;
    }
    const nodeIdx = this.heap.pop();
    const node = this.nodePool[nodeIdx];
    const { z, row, col, g, ripped } = node;
    const cellIdx = (z * this.rows + row) * this.cols + col;
    if (this.visitedStamp[cellIdx] === this.stamp) return;
    this.visitedStamp[cellIdx] = this.stamp;
    const seg = this.activeConnSeg;
    if (z === seg.endZ && row === seg.endRow && col === seg.endCol) {
      this.finalizeRoute(nodeIdx);
      this.activeConnSeg = null;
      this.activeConnId = -1;
      return;
    }
    const endZ = seg.endZ;
    const endRow = seg.endRow;
    const endCol = seg.endCol;
    const activeConn = this.activeConnId;
    const rows = this.rows;
    const cols = this.cols;
    const cellSizeMm = this.cellSizeMm;
    const visited = this.visitedStamp;
    const stamp = this.stamp;
    for (let d = 0; d < 8; d++) {
      const nr = row + DIRS_DR[d];
      const nc = col + DIRS_DC[d];
      if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) continue;
      const nIdx = (z * rows + nr) * cols + nc;
      if (visited[nIdx] === stamp) continue;
      this.computeMoveCostAndRips(activeConn, z, row, col, z, nr, nc, ripped);
      if (this._moveCost < 0) continue;
      const g2 = g + this._moveCost;
      const f2 = g2 + this.computeH(z, nr, nc, endZ, endRow, endCol) * this.hyperParameters.greedyMultiplier;
      const newNodeIdx = this.nodePool.length;
      this.nodePool.push({
        z,
        row: nr,
        col: nc,
        g: g2,
        f: f2,
        parentIdx: nodeIdx,
        ripped: this._moveRipped
      });
      this.heap.push(f2, this.seqCounter++, newNodeIdx);
    }
    const canVia = row >= this.minViaRow && row <= this.maxViaRow && col >= this.minViaCol && col <= this.maxViaCol;
    if (canVia) {
      for (let nz = 0; nz < this.layers; nz++) {
        if (nz === z) continue;
        const nIdx = (nz * rows + row) * cols + col;
        if (visited[nIdx] === stamp) continue;
        this.computeMoveCostAndRips(
          activeConn,
          z,
          row,
          col,
          nz,
          row,
          col,
          ripped
        );
        if (this._moveCost < 0) continue;
        const g2 = g + this._moveCost;
        const f2 = g2 + this.computeH(nz, row, col, endZ, endRow, endCol) * this.hyperParameters.greedyMultiplier;
        const newNodeIdx = this.nodePool.length;
        this.nodePool.push({
          z: nz,
          row,
          col,
          g: g2,
          f: f2,
          parentIdx: nodeIdx,
          ripped: this._moveRipped
        });
        this.heap.push(f2, this.seqCounter++, newNodeIdx);
      }
    }
  }
  // --- Merged cost + rip computation (writes to _moveCost/_moveRipped) ---
  computeMoveCostAndRips(activeConn, fromZ, fromRow, fromCol, toZ, toRow, toCol, ripped) {
    let cost = 0;
    let r = ripped;
    const cols = this.cols;
    if (fromZ !== toZ) {
      cost += this.hyperParameters.viaBaseCost;
      cost += Math.min(this.penalty2d[toRow * cols + toCol], this.penaltyCap);
      const toFlatIdx = (toZ * this.rows + toRow) * cols + toCol;
      const fixedOwner = this.portOwnerFlat[toFlatIdx];
      const fixedSameRoot = this.connIdToRootNet[fixedOwner] === this.connIdToRootNet[activeConn];
      const allowFixedOverlap = fixedSameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[activeConn]);
      const seg = this.activeConnSeg;
      const isSegEnd = !!seg && toZ === seg.endZ && toRow === seg.endRow && toCol === seg.endCol;
      if (fixedOwner >= 0 && fixedOwner !== activeConn && !allowFixedOverlap && !isSegEnd) {
        this._moveCost = -1;
        this._moveRipped = r;
        return;
      }
      this.fillViaOccupants(toRow, toCol, activeConn);
      const occs = this._viaOccs;
      for (let i = 0; i < occs.length; i++) {
        const occ = occs[i];
        if (!rippedContains(r, occ)) {
          cost += this.hyperParameters.ripCost;
          r = { id: occ, prev: r };
        }
        cost += this.hyperParameters.ripViaPenalty;
      }
    } else {
      const dr = fromRow > toRow ? fromRow - toRow : toRow - fromRow;
      const dc = fromCol > toCol ? fromCol - toCol : toCol - fromCol;
      cost += (dr + dc > 1 ? Math.SQRT2 : 1) * this.cellSizeMm;
      cost += Math.min(this.penalty2d[toRow * cols + toCol], this.penaltyCap);
      const flatIdx = (toZ * this.rows + toRow) * cols + toCol;
      const fixedOwner = this.portOwnerFlat[flatIdx];
      const fixedSameRoot = this.connIdToRootNet[fixedOwner] === this.connIdToRootNet[activeConn];
      const allowFixedOverlap = fixedSameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[activeConn]);
      const seg = this.activeConnSeg;
      const isSegEnd = !!seg && toZ === seg.endZ && toRow === seg.endRow && toCol === seg.endCol;
      if (fixedOwner >= 0 && fixedOwner !== activeConn && !allowFixedOverlap && !isSegEnd) {
        this._moveCost = -1;
        this._moveRipped = r;
        return;
      }
      const occ = this.usedCellsFlat[flatIdx];
      const sameRoot = this.connIdToRootNet[occ] === this.connIdToRootNet[activeConn];
      const allowSameRootOverlap = sameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[activeConn]);
      if (occ !== -1 && occ !== activeConn && !allowSameRootOverlap) {
        if (!rippedContains(r, occ)) {
          cost += this.hyperParameters.ripCost;
          r = { id: occ, prev: r };
        }
        cost += this.hyperParameters.ripTracePenalty;
      }
      if (dr === 1 && dc === 1) {
        const sqRow = fromRow < toRow ? fromRow : toRow;
        const sqCol = fromCol < toCol ? fromCol : toCol;
        const isBackslash = fromRow < toRow && fromCol < toCol || fromRow > toRow && fromCol > toCol;
        const diagSlot = isBackslash ? 0 : 1;
        const crossingSlot = diagSlot ^ 1;
        const sqCols = this.cols - 1;
        const diagBase = ((toZ * (this.rows - 1) + sqRow) * sqCols + sqCol) * 2;
        const crossingOcc = this.usedDiagFlat[diagBase + crossingSlot];
        const crossingSameRoot = this.connIdToRootNet[crossingOcc] === this.connIdToRootNet[activeConn];
        const allowCrossingOverlap = crossingSameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[activeConn]);
        if (crossingOcc !== -1 && crossingOcc !== activeConn && !allowCrossingOverlap) {
          this._moveCost = -1;
          this._moveRipped = r;
          return;
        }
      }
    }
    this._moveCost = cost;
    this._moveRipped = r;
  }
  // --- Via footprint unique occupants (fills _viaOccs scratch array) ---
  fillViaOccupants(row, col, activeConn) {
    const occs = this._viaOccs;
    occs.length = 0;
    const rows = this.rows;
    const cols = this.cols;
    const offDr = this.viaOffsetsDr;
    const offDc = this.viaOffsetsDc;
    const offLen = this.viaOffsetsLen;
    const used = this.usedCellsFlat;
    for (let z = 0; z < this.layers; z++) {
      const zBase = z * this.planeSize;
      for (let i = 0; i < offLen; i++) {
        const r = row + offDr[i];
        const c = col + offDc[i];
        if (r < 0 || c < 0 || r >= rows || c >= cols) continue;
        const occ = used[zBase + r * cols + c];
        if (occ === -1 || occ === activeConn) continue;
        const sameRoot = this.connIdToRootNet[occ] === this.connIdToRootNet[activeConn];
        if (sameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[activeConn])) {
          continue;
        }
        let seen = false;
        for (let j = 0; j < occs.length; j++) {
          if (occs[j] === occ) {
            seen = true;
            break;
          }
        }
        if (!seen) occs.push(occ);
      }
    }
  }
  // --- Visited stamp management ---
  nextStamp() {
    this.stamp = this.stamp + 1 >>> 0;
    if (this.stamp === 0) {
      this.visitedStamp.fill(0);
      this.stamp = 1;
    }
  }
  // --- Heuristic: Manhattan + via-zone awareness for cross-layer ---
  computeH(z, row, col, toZ, toRow, toCol) {
    const dr = Math.abs(row - toRow);
    const dc = Math.abs(col - toCol);
    const manhattan = dr + dc;
    if (z === toZ) return manhattan * this.cellSizeMm;
    if (!this.crossLayerSearch) {
      return manhattan * this.cellSizeMm + this.hyperParameters.viaBaseCost;
    }
    const vr1 = Math.max(this.minViaRow, Math.min(this.maxViaRow, row));
    const vc1 = Math.max(this.minViaCol, Math.min(this.maxViaCol, col));
    const vr2 = Math.max(this.minViaRow, Math.min(this.maxViaRow, toRow));
    const vc2 = Math.max(this.minViaCol, Math.min(this.maxViaCol, toCol));
    const via1 = Math.abs(row - vr1) + Math.abs(col - vc1) + Math.abs(vr1 - toRow) + Math.abs(vc1 - toCol);
    const via2 = Math.abs(row - vr2) + Math.abs(col - vc2) + Math.abs(vr2 - toRow) + Math.abs(vc2 - toCol);
    return Math.max(Math.min(via1, via2), manhattan) * this.cellSizeMm + this.hyperParameters.viaBaseCost;
  }
  // --- Connection interning ---
  internConn(name, rootNetName) {
    const existing = this.connNameToId.get(name);
    if (existing !== void 0) return existing;
    const id = this.connIdToName.length;
    this.connIdToName.push(name);
    this.connIdToRootNet.push(toRootNetName(name, rootNetName));
    this.connNameToId.set(name, id);
    return id;
  }
  // --- Build connection segments from port points ---
  buildConnectionSegs() {
    const byName = /* @__PURE__ */ new Map();
    for (const pp of this.nodeWithPortPoints.portPoints) {
      const name = pp.connectionName;
      if (!byName.has(name)) {
        byName.set(name, {
          points: [],
          rootConnectionName: pp.rootConnectionName
        });
      }
      byName.get(name).points.push(pp);
    }
    const segs = [];
    const seenSegmentKeys = /* @__PURE__ */ new Set();
    for (const [name, conn] of byName) {
      const pts = conn.points;
      if (pts.length < 2) continue;
      const connId = this.internConn(name, conn.rootConnectionName);
      for (let i = 0; i < pts.length - 1; i++) {
        const s = this.pointToCell(pts[i]);
        const e = this.pointToCell(pts[i + 1]);
        const endpointA = `${s.z}:${s.row}:${s.col}`;
        const endpointB = `${e.z}:${e.row}:${e.col}`;
        const orderedEndpoints = endpointA < endpointB ? `${endpointA}|${endpointB}` : `${endpointB}|${endpointA}`;
        const netName = conn.rootConnectionName ?? name;
        const segKey = `${netName}|${orderedEndpoints}`;
        if (seenSegmentKeys.has(segKey)) {
          this.overlapFriendlyRootNets.add(netName);
          continue;
        }
        seenSegmentKeys.add(segKey);
        segs.push({
          connId,
          startZ: s.z,
          startRow: s.row,
          startCol: s.col,
          endZ: e.z,
          endRow: e.row,
          endCol: e.col
        });
      }
    }
    return segs;
  }
  pointToCell(pt) {
    const col = Math.max(
      0,
      Math.min(
        this.cols - 1,
        Math.round((pt.x - this.gridOrigin.x) / this.cellSizeMm - 0.5)
      )
    );
    const row = Math.max(
      0,
      Math.min(
        this.rows - 1,
        Math.round((pt.y - this.gridOrigin.y) / this.cellSizeMm - 0.5)
      )
    );
    const z = this.zToLayer.get(pt.z) ?? 0;
    return { z, row, col };
  }
  shuffleConnections() {
    const arr = this.unsolvedSegs;
    let s = this.hyperParameters.shuffleSeed;
    const rng = () => {
      s = s * 1664525 + 1013904223 & 4294967295;
      return (s >>> 0) / 4294967295;
    };
    for (let i = arr.length - 1; i > 0; i--) {
      const j = Math.floor(rng() * (i + 1));
      const tmp = arr[i];
      arr[i] = arr[j];
      arr[j] = tmp;
    }
  }
  // --- Finalize a found route ---
  finalizeRoute(goalNodeIdx) {
    this.consecutiveSkips = Math.max(0, this.consecutiveSkips - 1);
    const cells = [];
    let idx = goalNodeIdx;
    while (idx >= 0) {
      const n = this.nodePool[idx];
      cells.push({ z: n.z, row: n.row, col: n.col });
      idx = n.parentIdx;
    }
    cells.reverse();
    while (cells.length > 1) {
      const first = cells[0];
      const firstFlat = (first.z * this.rows + first.row) * this.cols + first.col;
      if (!this.sharedCrossRootPortCells.has(firstFlat)) break;
      cells.shift();
    }
    while (cells.length > 1) {
      const last = cells[cells.length - 1];
      const lastFlat = (last.z * this.rows + last.row) * this.cols + last.col;
      if (!this.sharedCrossRootPortCells.has(lastFlat)) break;
      cells.pop();
    }
    const viaCells = [];
    for (let i = 1; i < cells.length; i++) {
      if (cells[i].z !== cells[i - 1].z) {
        viaCells.push({ row: cells[i].row, col: cells[i].col });
      }
    }
    const connId = this.activeConnId;
    const goalNode = this.nodePool[goalNodeIdx];
    const rippedIds = [];
    for (let cur = goalNode.ripped; cur; cur = cur.prev) {
      rippedIds.push(cur.id);
    }
    for (let i = 0; i < rippedIds.length; i++) {
      this.ripTrace(rippedIds[i]);
    }
    const marginCells = Math.ceil(this.traceMargin / this.cellSizeMm);
    const indices = [];
    const rows = this.rows;
    const cols = this.cols;
    const used = this.usedCellsFlat;
    for (let ci = 0; ci < cells.length; ci++) {
      const cell = cells[ci];
      for (let dr = -marginCells; dr <= marginCells; dr++) {
        for (let dc = -marginCells; dc <= marginCells; dc++) {
          const r = cell.row + dr;
          const c = cell.col + dc;
          if (r < 0 || r >= rows || c < 0 || c >= cols) continue;
          const flatIdx = (cell.z * rows + r) * cols + c;
          const existing = used[flatIdx];
          const sameRoot = this.connIdToRootNet[existing] === this.connIdToRootNet[connId];
          const allowSameRootOverlap = sameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[connId]);
          if (existing !== -1 && existing !== connId && !allowSameRootOverlap) {
            continue;
          }
          used[flatIdx] = connId;
          indices.push(flatIdx);
        }
      }
    }
    const displacedByVias = [];
    const offDr = this.viaOffsetsDr;
    const offDc = this.viaOffsetsDc;
    const offLen = this.viaOffsetsLen;
    for (let vi = 0; vi < viaCells.length; vi++) {
      const via = viaCells[vi];
      for (let z = 0; z < this.layers; z++) {
        const zBase = z * this.planeSize;
        for (let oi = 0; oi < offLen; oi++) {
          const r = via.row + offDr[oi];
          const c = via.col + offDc[oi];
          if (r < 0 || r >= rows || c < 0 || c >= cols) continue;
          const flatIdx = zBase + r * cols + c;
          const existing = used[flatIdx];
          const sameRoot = this.connIdToRootNet[existing] === this.connIdToRootNet[connId];
          const allowSameRootOverlap = sameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[connId]);
          if (existing !== -1 && existing !== connId && !allowSameRootOverlap) {
            let seen = false;
            for (let k = 0; k < displacedByVias.length; k++) {
              if (displacedByVias[k] === existing) {
                seen = true;
                break;
              }
            }
            if (!seen) displacedByVias.push(existing);
          }
          used[flatIdx] = connId;
          indices.push(flatIdx);
        }
      }
    }
    const diagIndices = [];
    const sqCols = this.cols - 1;
    for (let i = 1; i < cells.length; i++) {
      const prev = cells[i - 1];
      const curr = cells[i];
      if (prev.z !== curr.z) continue;
      const dr = prev.row > curr.row ? prev.row - curr.row : curr.row - prev.row;
      const dc = prev.col > curr.col ? prev.col - curr.col : curr.col - prev.col;
      if (dr !== 1 || dc !== 1) continue;
      const sqRow = prev.row < curr.row ? prev.row : curr.row;
      const sqCol = prev.col < curr.col ? prev.col : curr.col;
      const isBackslash = prev.row < curr.row && prev.col < curr.col || prev.row > curr.row && prev.col > curr.col;
      const diagSlot = isBackslash ? 0 : 1;
      const crossingSlot = diagSlot ^ 1;
      const diagBase = ((prev.z * (this.rows - 1) + sqRow) * sqCols + sqCol) * 2;
      const crossingIdx = diagBase + crossingSlot;
      const crossingOcc = this.usedDiagFlat[crossingIdx];
      const crossingSameRoot = this.connIdToRootNet[crossingOcc] === this.connIdToRootNet[connId];
      const allowCrossingOverlap = crossingSameRoot && this.overlapFriendlyRootNets.has(this.connIdToRootNet[connId]);
      if (crossingOcc !== -1 && crossingOcc !== connId && !allowCrossingOverlap) {
        continue;
      }
      const diagIdx = diagBase + diagSlot;
      this.usedDiagFlat[diagIdx] = connId;
      diagIndices.push(diagIdx);
    }
    while (this.usedIndicesByConn.length <= connId) {
      this.usedIndicesByConn.push([]);
    }
    this.usedIndicesByConn[connId] = indices;
    while (this.usedDiagIndicesByConn.length <= connId) {
      this.usedDiagIndicesByConn.push([]);
    }
    this.usedDiagIndicesByConn[connId] = diagIndices;
    this.solvedRoutes.set(connId, { connId, cells, viaCells });
    for (let i = 0; i < displacedByVias.length; i++) {
      this.ripTrace(displacedByVias[i]);
    }
    if (rippedIds.length > 0 || displacedByVias.length > 0) {
      const pen = this.penalty2d;
      const cap = this.penaltyCap;
      if (this.totalRipEvents > 50) {
        for (let i = 0; i < pen.length; i++) {
          pen[i] = pen[i] * 0.99;
        }
      } else {
        for (let i = 0; i < pen.length; i++) {
          if (pen[i] > cap) {
            pen[i] = pen[i] * 0.5;
          }
        }
      }
    }
  }
  // --- Rip a trace ---
  ripTrace(connId) {
    while (this.ripCount.length <= connId) this.ripCount.push(0);
    this.ripCount[connId]++;
    this.totalRipEvents++;
    const route = this.solvedRoutes.get(connId);
    if (route) {
      const cols = this.cols;
      for (let i = 0; i < route.cells.length; i++) {
        const cell = route.cells[i];
        const cellIdx = cell.row * cols + cell.col;
        this.penalty2d[cellIdx] = this.penalty2d[cellIdx] + this.hyperParameters.ripTracePenalty;
      }
      for (let i = 0; i < route.viaCells.length; i++) {
        const via = route.viaCells[i];
        const viaIdx = via.row * cols + via.col;
        this.penalty2d[viaIdx] = this.penalty2d[viaIdx] + this.hyperParameters.ripViaPenalty;
      }
    }
    const indices = this.usedIndicesByConn[connId];
    if (indices) {
      const used = this.usedCellsFlat;
      for (let i = 0; i < indices.length; i++) {
        const flatIdx = indices[i];
        if (used[flatIdx] === connId) {
          used[flatIdx] = -1;
        }
      }
      this.usedIndicesByConn[connId] = [];
    }
    const diagIndices = this.usedDiagIndicesByConn[connId];
    if (diagIndices) {
      const usedDiag = this.usedDiagFlat;
      for (let i = 0; i < diagIndices.length; i++) {
        const flatIdx = diagIndices[i];
        if (usedDiag[flatIdx] === connId) {
          usedDiag[flatIdx] = -1;
        }
      }
      this.usedDiagIndicesByConn[connId] = [];
    }
    if (route) {
      this.solvedRoutes.delete(connId);
      const first = route.cells[0];
      const last = route.cells[route.cells.length - 1];
      this.unsolvedSegs.push({
        connId,
        startZ: first.z,
        startRow: first.row,
        startCol: first.col,
        endZ: last.z,
        endRow: last.row,
        endCol: last.col
      });
    }
  }
  visualize() {
    const LAYER_COLORS = ["red", "blue", "orange", "green"];
    const points = [];
    const lines = [];
    const circles = [];
    const rects = [];
    const { width, height, center } = this.nodeWithPortPoints;
    rects.push({
      center: { x: center.x, y: center.y },
      width,
      height,
      stroke: "gray"
    });
    const vt = this.gridToBoundsTransform;
    if (this.showPenaltyMap && this.penalty2d) {
      let maxPenalty = 0;
      for (let i = 0; i < this.penalty2d.length; i++) {
        if (this.penalty2d[i] > maxPenalty) maxPenalty = this.penalty2d[i];
      }
      if (maxPenalty > 0) {
        for (let row = 0; row < this.rows; row++) {
          for (let col = 0; col < this.cols; col++) {
            const p = this.penalty2d[row * this.cols + col];
            if (p <= 0) continue;
            const alpha = Math.min(0.6, p / maxPenalty * 0.6);
            const tc = applyAffineTransformToPoint(vt, {
              x: this.gridOrigin.x + (col + 0.5) * this.cellSizeMm,
              y: this.gridOrigin.y + (row + 0.5) * this.cellSizeMm
            });
            rects.push({
              center: tc,
              width: this.cellSizeMm * vt.a,
              height: this.cellSizeMm * vt.e,
              fill: `rgba(255,165,0,${alpha.toFixed(3)})`
            });
          }
        }
      }
    }
    if (this.showUsedCellMap && this.usedCellsFlat) {
      for (let z = 0; z < this.layers; z++) {
        for (let row = 0; row < this.rows; row++) {
          for (let col = 0; col < this.cols; col++) {
            const occ = this.usedCellsFlat[(z * this.rows + row) * this.cols + col];
            if (occ === -1) continue;
            const tc = applyAffineTransformToPoint(vt, {
              x: this.gridOrigin.x + (col + 0.5) * this.cellSizeMm,
              y: this.gridOrigin.y + (row + 0.5) * this.cellSizeMm
            });
            rects.push({
              center: tc,
              width: this.cellSizeMm * vt.a,
              height: this.cellSizeMm * vt.e,
              fill: "rgba(0,0,255,0.5)"
            });
          }
        }
      }
    }
    for (const pp of this.nodeWithPortPoints.portPoints) {
      points.push({
        x: pp.x,
        y: pp.y,
        color: LAYER_COLORS[pp.z] ?? "gray",
        label: pp.connectionName
      });
    }
    const TRACE_COLORS = [
      "rgba(255,0,0,0.75)",
      "rgba(0,0,255,0.75)",
      "rgba(255,165,0,0.75)",
      "rgba(0,128,0,0.75)"
    ];
    const transformedRoutes = this.getOutput();
    for (const route of transformedRoutes) {
      if (route.route.length < 2) continue;
      let segStart = 0;
      for (let i = 1; i < route.route.length; i++) {
        const prev = route.route[i - 1];
        const curr = route.route[i];
        if (curr.z !== prev.z) {
          if (i - segStart >= 2) {
            lines.push({
              points: route.route.slice(segStart, i).map((p) => ({ x: p.x, y: p.y })),
              strokeColor: TRACE_COLORS[prev.z] ?? "rgba(128,128,128,0.75)",
              strokeWidth: this.traceThickness
            });
          }
          segStart = i;
        }
      }
      if (route.route.length - segStart >= 2) {
        const lastZ = route.route[segStart].z;
        lines.push({
          points: route.route.slice(segStart).map((p) => ({ x: p.x, y: p.y })),
          strokeColor: TRACE_COLORS[lastZ] ?? "rgba(128,128,128,0.75)",
          strokeWidth: this.traceThickness
        });
      }
    }
    for (const route of transformedRoutes) {
      for (const via of route.vias) {
        circles.push({
          center: { x: via.x, y: via.y },
          radius: this.viaDiameter / 2,
          fill: "rgba(0,0,0,0.3)",
          stroke: "black"
        });
      }
    }
    if (this.activeConnSeg && this.visitedStamp) {
      const currentStamp = this.stamp;
      for (let z = 0; z < this.layers; z++) {
        for (let row = 0; row < this.rows; row++) {
          for (let col = 0; col < this.cols; col++) {
            if (this.visitedStamp[(z * this.rows + row) * this.cols + col] !== currentStamp)
              continue;
            const tc = applyAffineTransformToPoint(vt, {
              x: this.gridOrigin.x + (col + 0.5) * this.cellSizeMm,
              y: this.gridOrigin.y + (row + 0.5) * this.cellSizeMm
            });
            points.push({
              x: tc.x,
              y: tc.y,
              color: "rgba(0,0,255,0.2)"
            });
          }
        }
      }
    }
    return {
      points,
      lines,
      circles,
      rects,
      coordinateSystem: "cartesian",
      title: `HighDensityA01 [${this.solvedRoutes?.size ?? 0} solved, ${this.unsolvedSegs?.length ?? 0} remaining]`
    };
  }
  getOutput() {
    const t = this.gridToBoundsTransform;
    const result = [];
    for (const [connId, route] of this.solvedRoutes) {
      const connName = this.connIdToName[connId];
      result.push({
        connectionName: connName,
        traceThickness: this.traceThickness,
        viaDiameter: this.viaDiameter,
        route: route.cells.map((cell) => {
          const rawX = this.gridOrigin.x + (cell.col + 0.5) * this.cellSizeMm;
          const rawY = this.gridOrigin.y + (cell.row + 0.5) * this.cellSizeMm;
          const tp = applyAffineTransformToPoint(t, { x: rawX, y: rawY });
          return { x: tp.x, y: tp.y, z: this.layerToZ.get(cell.z) ?? cell.z };
        }),
        vias: route.viaCells.map((via) => {
          const rawX = this.gridOrigin.x + (via.col + 0.5) * this.cellSizeMm;
          const rawY = this.gridOrigin.y + (via.row + 0.5) * this.cellSizeMm;
          return applyAffineTransformToPoint(t, { x: rawX, y: rawY });
        })
      });
    }
    return result;
  }
};
export {
  HighDensitySolverA01,
  applyAffineTransformToPoint,
  computeGridToAffineTransform,
  convertDataset02SampleToNodeWithPortPoints
};
/*! Bundled license information:

is-buffer/index.js:
  (*!
   * Determine if an object is a Buffer
   *
   * @author   Feross Aboukhadijeh <https://feross.org>
   * @license  MIT
   *)

deep-rename-keys/index.js:
  (*!
   * deep-rename-keys <https://github.com/jonschlinkert/deep-rename-keys>
   *
   * Copyright (c) 2015 Jon Schlinkert, contributors.
   * Licensed under the MIT license.
   *)
*/
